// bitfab-plugin-lib/dist/flows/assistant.js
import { Flow } from "./schema.js";
import { CODEX_RESOLVE_PLUGIN_DIR_BASH } from "./shared.js";
export const assistantFlow = Flow.parse({
    schemaVersion: 1,
    id: "assistant",
    title: "Bitfab Assistant",
    frontmatter: {
        description: "Iterate on a traced function to improve pass rates using failed traces, labeling, and replay. TRIGGER when: user wants to fix failing AI outputs, improve pass rates, debug LLM behavior, iterate on prompts, label traces, run experiments, or says anything like 'fix my AI', 'improve pass rate', 'why is this failing', 'iterate on traces', 'debug my agent', 'review traces'. SKIP when: user wants to instrument new code or set up tracing (use bitfab:setup instead).",
    },
    tools: {
        bash: { kind: "builtin", name: "Bash" },
        read: { kind: "builtin", name: "Read" },
        glob: { kind: "builtin", name: "Glob" },
        grep: { kind: "builtin", name: "Grep" },
        edit: { kind: "builtin", name: "Edit" },
        write: { kind: "builtin", name: "Write" },
        agent: { kind: "builtin", name: "Agent" },
        ask: { kind: "builtin", name: "AskUserQuestion" },
        monitor: { kind: "builtin", name: "Monitor" },
        skill: { kind: "builtin", name: "Skill" },
        listFunctions: { kind: "mcp", name: "list_trace_functions" },
        searchTraces: { kind: "mcp", name: "search_traces" },
        readTraces: { kind: "mcp", name: "read_traces" },
        updateAgentLabels: { kind: "mcp", name: "update_agent_labels" },
        listDatasets: { kind: "mcp", name: "list_datasets" },
        createDataset: { kind: "mcp", name: "create_dataset" },
        addTracesToDataset: { kind: "mcp", name: "add_traces_to_dataset" },
        removeTracesFromDataset: {
            kind: "mcp",
            name: "remove_traces_from_dataset",
        },
        getTracePlan: { kind: "mcp", name: "get_trace_plan" },
    },
    intro: `Use the local plugin MCP tools ({{tool:listFunctions}}, {{tool:searchTraces}}, {{tool:readTraces}}, {{tool:updateAgentLabels}}, {{tool:listDatasets}}, {{tool:createDataset}}, {{tool:addTracesToDataset}}, {{tool:removeTracesFromDataset}}) to find what's failing in a traced function, build a dataset of labeled traces, and iterate on the code/prompts using replay until pass rates improve.

**MCP tools:** This skill uses \`list_trace_functions\`, \`search_traces\`, \`read_traces\`, \`update_agent_labels\`, \`list_datasets\`, \`create_dataset\`, \`add_traces_to_dataset\`, \`remove_traces_from_dataset\`, and \`get_trace_plan\` from the **local plugin MCP server** (bundled with this plugin){{#claude}}. Do NOT use the remote Bitfab MCP tools (\`mcp__Simforge__*\` or \`mcp__Bitfab__*\`) — use only the \`mcp__plugin_bitfab_Bitfab__*\` variants.{{/claude}}{{^claude}}, exposed under the \`mcp__Bitfab__*\` prefix.{{/claude}}

{{^codex}}**Always use** \`AskUserQuestion\` **when asking questions, reporting results, or presenting choices.** Never print a question as text and wait. Rules:{{/codex}}{{#codex}}**When the flow branches, always present the options clearly and wait for the user's answer before proceeding.** Number or letter the options so the user can pick by reference. Rules:{{/codex}}

- Recommend an option first, explain why in one line
- Present 2-5 concrete options
- One decision per question — never batch

This skill has three invocation modes. \`all\` walks every phase. The two sub-modes do one focused thing each, building a labeled dataset, or running experiments against an existing one, and require the trace function key as the argument because they skip the function picker (Phase 1) and instrumentation/replay verification (Phase 2).

| Invocation | Action |
|---|---|
| \`{{cmd}}assistant\` or \`{{cmd}}assistant all\` | Full flow: pick function → verify instrumentation → pick or create dataset → label → diagnose → iterate → wrap up |
| \`{{cmd}}assistant dataset <key>\` | Build or extend a labeled dataset for one function, then stop. No experiments run. Picks an existing dataset or creates a new one |
| \`{{cmd}}assistant experiment <key> [<dataset-id>]\` | Run experiments to fix failing traces against a labeled dataset, then wrap up. If \`<dataset-id>\` is omitted, you'll be asked to pick one. If the function has no datasets yet, run \`{{cmd}}assistant dataset <key>\` first |

In sub-modes, grep the codebase for \`<key>\` early so labeling and experiments are grounded in the actual instrumented function (the full flow does this in Phase 2; sub-modes skip Phase 2 entirely).

**Studio** is the companion browser surface for the entire assistant flow. It opens automatically at the start and stays open throughout all phases. Individual phases navigate the Studio to the relevant page (dataset review, experiment viewer, etc.).

**Opening a trace plan, when asked.** Opening trace plans is part of this skill, not a separate primitive — but only do it when the user asks (or the context clearly implies it, e.g. they said "show me what's captured"). Never auto-open. When triggered, run two sequential calls (step 2 needs the planId from step 1, so they can't be batched): (1) {{tool:getTracePlan}} with \`{ traceFunctionKey: "<key>" }\` returns the plan id, then (2) \`navigateStudio.js <sessionId> "/studio/trace-plan/<planId>"\` (substituting the id from step 1) routes the already-open Studio tab there in-place. The Studio chrome (header, session indicator, agent activity) stays mounted around the trace plan content — no new tab pops up. No questions, no preamble, no summary up-front. If no plan exists for the key, say so in one line and offer \`{{cmd}}setup modify <key>\` to build one.
{{#codex}}

🚨 **Blocking-process rule (applies to any plugin command described as "blocks until the user does X"):** When you launch a plugin CLI that blocks on a browser handoff (\`login.js\`, etc.), you MUST keep the exec session alive and keep polling it until the process exits on its own.

- Printing the URL is NOT completion. The process is still running, racing a loopback callback and a server-polled ticket — it will exit only after the user's browser action delivers a result (or after the timeout).
- After sending the URL to the user, keep polling the live shell/exec session at least every few seconds with your normal "read more output" tool (\`write_stdin\`, \`read\`, or whatever your runtime's equivalent is for the long-running shell). Do not idle waiting for a user message.
- Do not send a final "waiting for you to click Confirm" text and then stop polling — the user's confirmation does NOT come back to you as a chat message; it comes back as the plugin process exiting with output on stdout.
- Stop polling only when one of: (a) the process exits 0 and prints its completion summary, (b) the process exits non-zero, or (c) the user explicitly cancels.
- When the process exits, immediately continue with the next step in the flow — don't wait for another user message.{{/codex}}`,
    entries: {
        all: {
            default: "studio/open",
            codex: "phase-0/resolve-plugin-dir",
        },
        dataset: {
            default: "studio/open",
            codex: "phase-0/resolve-plugin-dir",
        },
        experiment: {
            default: "studio/open",
            codex: "phase-0/resolve-plugin-dir",
        },
    },
    defaultMode: "all",
    variables: {
        /**
         * How a question is presented to the user. Claude / Cursor have the
         * `AskUserQuestion` tool ("via AskUserQuestion"); Codex doesn't and
         * presents options inline ("as a numbered choice"). Used wherever the
         * body says "Present the … to the user {{presentMode}}: …".
         */
        presentMode: {
            default: "via `AskUserQuestion`",
            codex: "as a numbered choice",
        },
    },
    phases: [
        {
            id: "phase-0",
            title: "Phase 0: Status + Update Check",
            stepStyle: "list",
            steps: [
                {
                    id: "resolve-plugin-dir",
                    kind: "action",
                    preamble: true,
                    toolCalls: ["bash"],
                    title: "Resolve BITFAB_PLUGIN_DIR",
                    body: `First, resolve \`BITFAB_PLUGIN_DIR\` if it isn't already exported in this shell. Run this block verbatim — it auto-detects dev / prod / custom-\`CODEX_HOME\` installs:

${CODEX_RESOLVE_PLUGIN_DIR_BASH}`,
                    next: "phase-0/run-status",
                },
                {
                    id: "run-status",
                    kind: "branch",
                    toolCalls: ["bash"],
                    title: "Run the status command",
                    body: `Then run the status command:

\`\`\`bash
node "{{pluginRoot}}/dist/commands/status.js"
\`\`\`

Watch the output and route on it:

{{whens}}`,
                    branches: [
                        {
                            when: "not authenticated",
                            description: "stop the flow immediately. Tell the user to run `{{cmd}}setup login` first",
                            next: null,
                        },
                        {
                            when: "authenticated (with or without a `v<X> available` upgrade notice)",
                            description: "continue to the Studio gateway. If an upgrade notice appeared, pass it through to the user verbatim, but don't block on it; surface the notice once and move on",
                            next: "studio/open",
                        },
                    ],
                },
            ],
        },
        {
            id: "studio",
            title: "Studio Lifecycle",
            intro: `The Studio is the companion browser surface for the entire assistant flow. It opens once at the start and stays open throughout all phases. Individual phases navigate the Studio to the relevant page (dataset review, experiment viewer, etc.) using \`navigateStudio.js\`. If the Studio background process outputs \`{"event":"session-ended",...}\` at any point, the user has closed the Studio early. This is not an error: continue the flow normally, but skip any \`navigateStudio.js\` calls for the rest of the session (the session is gone). Do **not** attempt to reopen the Studio.`,
            stepStyle: "list",
            steps: [
                {
                    id: "open",
                    kind: "action",
                    toolCalls: ["bash"],
                    title: "Open the Studio",
                    body: `Start the Studio as a long-running background process. The command accepts an optional initial path argument so Studio opens directly at the relevant page:

- **\`all\` mode:** no path argument (opens at \`/studio\` root)
- **\`dataset <key>\` mode:** pass \`/studio/trace-functions/<key>/datasets\`
- **\`experiment <key>\` mode:** pass \`/studio/trace-functions/<key>/experiments\`

\`\`\`bash
node "{{pluginRoot}}/dist/commands/openStudio.js" [initialPath]
\`\`\`

{{#claude}}Run it with \`run_in_background: true\` on the Bash tool.{{/claude}}{{#cursor}}Run it as a long-running terminal process.{{/cursor}}{{#codex}}Run it via your runtime's "long-running exec session" mechanism.{{/codex}}

The script outputs JSON lines on stdout:

- \`{"event":"started","sessionId":"..."}\` on startup: capture the \`sessionId\` and hold it in working context for the rest of the flow. Every \`navigateStudio.js\` call in later phases uses it.
- \`{"event":"auth-required","sessionId":"..."}\` if the user is not signed in to Bitfab in their browser: tell the user to sign in to Bitfab in the Studio window that just opened, then wait for the \`authenticated\` event before continuing.
- \`{"event":"authenticated","sessionId":"..."}\` the user has signed in. Continue the flow.
- \`{"event":"session-ended","sessionId":"..."}\` if the user closes the Studio: the process exits. See the lifecycle note above.

Status messages (e.g. "Opening Studio: ...") go to stderr, not stdout. Filter to JSON lines only.`,
                    next: {
                        byMode: {
                            all: "phase-1/list-or-arg",
                            dataset: "phase-3/list-datasets",
                            experiment: "phase-5/pick-dataset",
                        },
                    },
                },
            ],
        },
        {
            id: "phase-1",
            title: "Phase 1: Identify the Trace Function",
            intro: "If a `traceFunctionKey` was provided as an argument, skip the listing and the user prompt — but still cross-check the provided key against the local codebase before moving on. Otherwise, work through all four steps below:",
            stepStyle: "list",
            steps: [
                {
                    id: "list-or-arg",
                    kind: "action",
                    toolCalls: ["listFunctions"],
                    title: "List trace functions",
                    body: `**Skip this step if a \`traceFunctionKey\` argument was provided** — use the argument directly and continue to cross-check. Otherwise, call {{tool:listFunctions}} to list all available trace functions. Use **only** the keys and metadata returned (trace counts, last activity) — do NOT invent or infer descriptions of what each function does from its key name. Key names are often ambiguous or misleading, and guessing produces hallucinated descriptions that confuse the user.`,
                    next: "phase-1/cross-check",
                },
                {
                    id: "cross-check",
                    kind: "action",
                    toolCalls: ["grep", "glob"],
                    title: "Cross-check against the local codebase",
                    body: `**Cross-check each key against the local codebase** before presenting. For each returned key, \`grep\` the repo for string-literal uses of that exact key (across \`*.ts\`, \`*.tsx\`, \`*.py\`, \`*.rb\`, \`*.go\`, \`*.baml\`). Mark each function in the presented list as:

- **✅ instrumented here** — found in this repo, with the file path
- **⚠️ not found in this repo** — traces exist on Bitfab but the key isn't in this codebase (likely another repo or a renamed key)`,
                    next: "phase-1/present-list",
                },
                {
                    id: "present-list",
                    kind: "action",
                    title: "Present the list",
                    body: `**Skip this step if a \`traceFunctionKey\` argument was provided** — there's no list to present. Otherwise, present the full list in the question text showing ONLY: \`<key>\` · \`<trace count>\` · \`<last activity>\` · \`<instrumented-here marker + path, or not-found marker>\`. No invented summaries.`,
                    next: "phase-1/ask-function",
                },
                {
                    id: "ask-function",
                    kind: "action",
                    toolCalls: ["ask"],
                    title: "Ask which function to improve",
                    body: `**Skip this step if a \`traceFunctionKey\` argument was provided** — the function is already chosen. Otherwise, {{askVerb}} with 2 options: the recommended function (prefer one that is ✅ instrumented here AND has recent activity) and a free-text "Type a function key" option. If nothing is instrumented here, say so explicitly in the question — don't hide it.`,
                    next: "phase-2/check-instrumentation",
                },
            ],
        },
        {
            id: "phase-2",
            title: "Phase 2: Verify Instrumentation & Replay",
            intro: "Check that this trace function has both instrumentation and a replay script.",
            stepStyle: "list",
            steps: [
                {
                    id: "check-instrumentation",
                    kind: "branch",
                    toolCalls: ["bash", "grep", "ask", "skill"],
                    title: "Check instrumentation",
                    body: `Search the codebase for the trace function key to {{^codex}}find where the SDK is used{{/codex}}{{#codex}}confirm SDK usage{{/codex}}:

- TypeScript: \`grep -r "<traceFunctionKey>" --include="*.ts" --include="*.tsx"\`
- Python: \`grep -r "<traceFunctionKey>" --include="*.py"\`
- Ruby: \`grep -r "<traceFunctionKey>" --include="*.rb"\`
- Go: \`grep -r "<traceFunctionKey>" --include="*.go"\`

If the key is found, note the file location — this is the code you'll iterate on in later phases.

If the key is NOT found in the codebase, the function is instrumented elsewhere (the traces exist on Bitfab). {{^codex}}Use \`AskUserQuestion\` to ask:{{/codex}}{{#codex}}Ask:{{/codex}}

> "I can't find \`<traceFunctionKey>\` in this codebase — it may be instrumented in another repo or under a different key."
>
{{branches}}

If the user chooses **"Instrument now"**, invoke \`{{cmd}}setup instrument\`{{#claude}} using the Skill tool{{/claude}}, then verify whether a replay script exists for this function. If **"Continue anyway"**, skip the replay-script check and start building the dataset — there's no local code to iterate on yet.`,
                    branches: [
                        {
                            when: "key found in codebase",
                            next: "phase-2/check-replay-script",
                        },
                        {
                            option: {
                                letter: "A",
                                label: "Instrument now",
                                description: "set up tracing in this codebase",
                            },
                            recommended: true,
                            next: "phase-2/check-replay-script",
                        },
                        {
                            option: {
                                letter: "B",
                                label: "Continue anyway",
                                description: "work with the traces even without local code",
                            },
                            next: "phase-3/list-datasets",
                        },
                        {
                            option: {
                                letter: "C",
                                label: "Pick a different function",
                            },
                            next: "phase-1/list-or-arg",
                        },
                        {
                            option: { letter: "D", label: "Stop" },
                            next: null,
                        },
                    ],
                },
                {
                    id: "check-replay-script",
                    kind: "branch",
                    toolCalls: ["read", "glob", "ask", "skill"],
                    title: "Check replay script",
                    body: `Search for a replay script that covers this trace function:

- Look for files matching \`scripts/replay.*\`, \`scripts/*replay*\`, or any file that imports \`bitfab.replay\` / \`client.replay\`
- Read the script and check that it maps the target trace function key

If a replay script exists but targets a different function key, do NOT modify the existing script or suggest changing the code's function key. Instead, treat it as "no replay script for this function" and offer to create a new one.

If no replay script exists or it doesn't cover this function, {{askVerb}}:

> "No replay script found for \`<traceFunctionKey>\`."
>
{{branches}}

If the user chooses **"Create replay now"**, invoke \`{{cmd}}setup replay\`{{#claude}} using the Skill tool{{/claude}}, then start building the dataset.`,
                    branches: [
                        {
                            when: "replay script covers this function",
                            next: "phase-3/list-datasets",
                        },
                        {
                            option: {
                                letter: "A",
                                label: "Create replay now",
                                description: "create the replay script inline",
                            },
                            recommended: true,
                            next: "phase-3/list-datasets",
                        },
                        {
                            option: {
                                letter: "B",
                                label: "Pick a different function",
                            },
                            next: "phase-1/list-or-arg",
                        },
                        {
                            option: { letter: "C", label: "Stop" },
                            next: null,
                        },
                    ],
                },
            ],
        },
        {
            id: "phase-3",
            title: "Phase 3: Pick a Dataset and Label Traces",
            intro: `A **dataset** is the named bucket of labeled traces an experiment replays against. This phase picks (or creates) one for the trace function, labels candidate traces, attaches them to the dataset, then hands off to the per-dataset review page where the user approves labels and can ask the agent to add or remove traces.

In \`dataset\` mode this phase is the entry point — Phase 1 (function picker) and Phase 2 (instrumentation/replay verification) are skipped, so the trace function key comes from the argument. Before calling any MCP tools, grep the codebase for the key (e.g. \`grep -r "<traceFunctionKey>" --include="*.ts" --include="*.tsx" --include="*.py" --include="*.rb" --include="*.go" --include="*.baml"\`) and note the file path — every later step ("Label them yourself", and Phase 4 "Read the code" in \`all\` mode) needs it.`,
            stepStyle: "list",
            steps: [
                {
                    id: "list-datasets",
                    kind: "branch",
                    toolCalls: ["listDatasets", "createDataset", "ask"],
                    title: "Pick or create a dataset",
                    body: `**Pick or create a dataset** — Call {{tool:listDatasets}} with the trace function key. Then branch on whether any exist. Hold the chosen \`datasetId\` in working context — every step from here on uses it.

{{whens}}`,
                    branches: [
                        {
                            when: "no datasets exist for this function (list_datasets returned empty)",
                            description: `**don't ask** — silently call {{tool:createDataset}} with \`traceFunctionKey: <key>\` and \`name: <key>\` (just the trace function key as the name; the user can rename it later in the UI if they want). Hold the returned \`datasetId\` and continue. The first-time user shouldn't have to answer a name prompt before they've even seen the dataset.`,
                            next: "phase-3/ask-search-mode",
                        },
                        {
                            when: "one or more datasets already exist",
                            description: `present them to the user {{presentMode}}, with one option per existing dataset (name · id · current trace count) plus a "Create new" option. Recommend the most recently used dataset that has traces. If the user picks an existing dataset, hold its id and continue. If the user picks "Create new", silently call {{tool:createDataset}} with \`name: "<key> #N"\` where N is one more than the number of existing datasets (e.g. \`eval-assistant #2\`) — don't ask for a name. Hold the new id and continue.`,
                            next: "phase-3/ask-search-mode",
                        },
                    ],
                },
                {
                    id: "ask-search-mode",
                    kind: "branch",
                    toolCalls: ["searchTraces", "ask"],
                    title: "Ask how to source candidate traces",
                    body: `**Ask how to source candidate traces.** Before searching, decide *where* the candidate traces come from. Three real options:

1. **Define new criteria** — agent searches unlabeled traces shaped by what the user wants to surface. Best when the user has a hypothesis or a specific failure pattern in mind.
2. **Reuse existing labels for this function** — pull traces that already have a validated human or approved-agent label (from any prior dataset on this function) and seed the new dataset with them. Best when the user wants to hill-climb off prior labeling work — same labels, different cut, add more later.
3. **Open / you decide** — agent samples broadly with no hypothesis, ignoring prior labels for the search shape. Best for discovery passes.

**Probe for prior label volume first** so the recommendation is grounded. Call {{tool:searchTraces}} with \`traceFunctionKey: <key>\`, \`validated: true\`, \`limit: 50\` to see roughly how many validated labels already exist for this function. Note the count — you'll need it for the recommendation and for option 2.

Then {{askVerb}} with the three options below. Recommend:
- Option **2 (Reuse)** if the function has 5+ validated labels AND the picked dataset is freshly created or empty (the user is starting a new cut and prior work is the right baseline)
- Option **1 (Define)** if the user has a hypothesis or the function has < 5 validated labels (not enough prior signal to reuse)
- Option **3 (Open)** if the user explicitly says they don't have a hypothesis yet and there's not much prior labeling

Hold the chosen mode in working context — the next steps branch on it.

{{branches}}`,
                    branches: [
                        {
                            option: {
                                letter: "A",
                                label: "Define new criteria",
                                description: "tell me what to find (failure pattern, customer reports, etc.) and I search unlabeled traces",
                            },
                            next: "phase-3/ask-search-intent",
                        },
                        {
                            option: {
                                letter: "B",
                                label: "Reuse existing labels for this function",
                                description: "seed the dataset with traces that already have validated labels, then optionally add more",
                            },
                            recommended: true,
                            next: "phase-3/seed-from-existing-labels",
                        },
                        {
                            option: {
                                letter: "C",
                                label: "Open — you decide",
                                description: "broad sample with no hypothesis; ignore prior labels for the search shape",
                            },
                            next: "phase-3/find-unlabeled",
                        },
                    ],
                },
                {
                    id: "seed-from-existing-labels",
                    kind: "branch",
                    toolCalls: [
                        "searchTraces",
                        "readTraces",
                        "addTracesToDataset",
                        "ask",
                    ],
                    title: "Seed dataset from existing validated labels",
                    body: `**Seed dataset from existing validated labels.** Reachable only when the user picked Option B in \`ask-search-mode\`. Pull traces that already have a validated label (human-authored, or agent-authored and human-approved) for this function, attach them to the picked dataset, and route on whether the user also wants to add more.

1. Call {{tool:searchTraces}} with \`traceFunctionKey: <key>\`, \`validated: true\`, and a generous \`limit\` (50 is the cap). Both \`labelResult: true\` and \`labelResult: false\` matter — failures are the hill-climbing signal, but passes anchor the regression boundary. If 50 isn't enough to cover the function's labeled history, run a second call with \`labelResult: false\` only to bias toward fails first, then a third with \`labelResult: true\`. De-dupe trace IDs across calls.
2. Call {{tool:readTraces}} with \`scope: "summary"\` on the resulting trace IDs so the labels + annotations are in working context. Don't re-label them — these are already validated.
3. Call {{tool:addTracesToDataset}} once with \`datasetId\` (the one picked in \`list-datasets\`) and the full deduped trace ID array. The call is idempotent, so re-attaching IDs already in the dataset is a safe no-op.
4. Briefly summarize for the user: "Seeded the dataset with N reused labels (M fails, K passes). Want me to find more candidates to label, or is this set enough to move on?"

{{branches}}`,
                    branches: [
                        {
                            option: {
                                letter: "A",
                                label: "Find more candidates to label",
                                description: "go through the regular intent + search + label flow on top of the reused set",
                            },
                            next: "phase-3/ask-search-intent",
                        },
                        {
                            option: {
                                letter: "B",
                                label: "Move on with just the reused set",
                                description: "skip further labeling; go straight to dataset review",
                            },
                            recommended: true,
                            next: "phase-3/open-page",
                        },
                    ],
                },
                {
                    id: "ask-search-intent",
                    kind: "action",
                    toolCalls: ["ask"],
                    title: "Ask what kinds of traces to find",
                    body: `**Ask what kinds of traces to find** — The user picked "Define new criteria" (or arrived here from the reuse path wanting more). Find out what they're actually trying to surface. The trace function may have thousands of traces; "what should I label?" is the question that makes the rest of this phase useful.

When asking, {{askVerb}} with these options (and a free-text fallback so the user can describe something specific):

- **A — Failures of a certain kind** *(recommended when the user already has a hypothesis)* — they tell you the pattern (empty outputs, hallucinated tool args, regressions on a specific input shape, etc.) and you search for matching traces
- **B — Recent customer complaints / reports** — they paste or describe specific incidents and you find the matching traces by user, session, or time window
- **C — Open-ended, you decide** — no hypothesis yet; you sample broadly across recent traces, look for diversity, and surface anything that looks like a candidate failure or interesting edge case

Hold the user's answer (the chosen option **and** any free-text detail) in working context — the next step uses it to shape the {{tool:searchTraces}} filters and which traces to prioritise reading. If they pick C, default to recent + diverse + non-empty outputs.`,
                    next: "phase-3/find-unlabeled",
                },
                {
                    id: "find-unlabeled",
                    kind: "action",
                    toolCalls: ["searchTraces", "readTraces"],
                    title: "Find unlabeled traces",
                    body: `**Find unlabeled traces** — Search without label filters to find unlabeled traces for the trace function. **Shape the search by the intent captured in the previous step** (or by the prior dataset's existing labels, if any): Option A = filter to traces matching the user's described failure pattern; Option B = filter by the user, session, or time window of the reported incidents; Option C = default sweep (recent, diverse inputs, non-empty outputs). Use {{tool:searchTraces}} with the relevant filters, then {{tool:readTraces}} with \`scope: "summary"\` to read candidates and identify which are worth labeling — look for diverse inputs, traces that produced output (not empty), and traces that cover different scenarios under the chosen intent. Filter out near-duplicates and uninteresting traces. If every trace is already labeled and attached to this dataset, you can move straight on with no new candidates.`,
                    next: "phase-3/ask-labeling-mode",
                },
                {
                    id: "ask-labeling-mode",
                    kind: "branch",
                    toolCalls: ["ask"],
                    title: "Ask how the user wants to label",
                    body: `**Ask how the user wants to label** — Before any verdicts go on these candidate traces, {{askVerb}} how the user wants to label them. There are exactly two modes, and the answer determines whether you call {{tool:updateAgentLabels}} at all:

{{branches}}

Recommend Option A — an agent first pass turns the labeling page into a quick approve/edit review. But respect the user's choice: if they pick B, do **not** call {{tool:updateAgentLabels}} for any of these candidates. They want to label from scratch in the labeling page, with no agent verdicts pre-filled. If no new candidate traces were found in the previous step, skip this question and continue.`,
                    branches: [
                        {
                            option: {
                                letter: "A",
                                label: "Agent labels first, I approve / edit",
                                description: "agent makes a first pass; you approve or edit each verdict in the labeling page",
                            },
                            recommended: true,
                            next: "phase-3/label-self",
                        },
                        {
                            option: {
                                letter: "B",
                                label: "I'll label them manually",
                                description: "no agent verdicts; you label every trace from scratch in the labeling page",
                            },
                            next: "phase-3/attach-to-dataset",
                        },
                    ],
                },
                {
                    id: "label-self",
                    kind: "action",
                    toolCalls: ["read", "readTraces", "updateAgentLabels"],
                    title: "Agent first pass: label them yourself before opening the labeling page",
                    body: `**Agent first pass: label them yourself before opening the labeling page** — Reachable only when the user picked Option A in the previous step. **You** label the approved candidate traces so the labeling page becomes an approve/edit review instead of a blank labeling session. Call {{tool:readTraces}} with \`scope: "full"\` on the approved trace IDs (batch them — up to 10 per call), read each trace's inputs / output / spans yourself, and decide for each one whether it looks like a PASS or a FAIL. **Ground your judgment in the codebase, not just the trace text.** Before you start labeling, read the instrumented function in the user's source (located in Phase 2 in \`all\` mode, or via the grep step in this phase's intro in \`dataset\` mode) and any nearby code that explains intent — comments, docstrings, README sections, related tests, BAML files — so you know what the function is *supposed* to do and what "good" looks like for it. Apply the same context to every trace: does this output achieve the function's goal as expressed in the code? Does it match the patterns in the already-validated traces? Then call {{tool:updateAgentLabels}} once with an array of \`{ traceId, label, annotation }\` objects — **both \`label\` (true for pass, false for fail) and \`annotation\` (a one-or-two-sentence explanation written for the human reviewer, ideally referencing what the code is trying to do) are required for every trace**. Commit to a verdict — if you genuinely cannot decide, you didn't read the trace or the code carefully enough. The labels you save here start unapproved; they only become part of the validated dataset once a human approves them in the labeling page.

> 🚨 **HARD RULE — DO NOT SKIP (agent-first mode only):** When the user picked Option A, you MUST call {{tool:updateAgentLabels}} with verdicts for every approved trace BEFORE navigating Studio to the labeling page. Sending the user into an agent-first review with no pre-labeled verdicts is a process violation. (In manual mode this step is unreachable, and the rule does not apply.)

> **Made a mistake?** If you realize a verdict was wrong (e.g., you mislabeled a trace or want to re-evaluate), call {{tool:updateAgentLabels}} again with \`{ traceId, archive: true }\` for those traces. The previous label is hidden (kept for audit), and you can re-label the trace from scratch with another \`update_agent_labels\` call.`,
                    next: "phase-3/attach-to-dataset",
                },
                {
                    id: "attach-to-dataset",
                    kind: "action",
                    toolCalls: ["addTracesToDataset"],
                    title: "Attach labeled traces to the dataset",
                    body: `**Attach candidate traces to the dataset** — Call {{tool:addTracesToDataset}} with the \`datasetId\` chosen earlier and the array of approved candidate trace IDs (in agent-first mode, the ones you just labeled; in manual mode, the candidates the user approved in find-unlabeled). The call is idempotent — re-adding traces already in the dataset is a no-op, so it's safe to include the full set. If no new candidate traces were approved (the dataset was already populated), skip this step.

> 🚨 **HARD RULE — DO NOT SKIP:** All approved candidate trace IDs MUST be attached to the dataset before opening the page. The page reviews the dataset's contents, not the trace function's label table. An empty dataset means an empty review.`,
                    next: "phase-3/open-page",
                },
                {
                    id: "open-page",
                    kind: "action",
                    toolCalls: ["bash"],
                    title: "Open the dataset review page",
                    body: `Open the dataset review page for the user. Use \`navigateStudio.js\` to route the already-open Studio to the dataset review page using the \`sessionId\` captured in the \`studio/open\` step:

\`\`\`bash
node "{{pluginRoot}}/node_modules/bitfab-plugin-lib/dist/commands/navigateStudio.js" <sessionId> "/studio/trace-functions/<functionKey>/datasets/<datasetId>"
\`\`\`

The command sends a navigate event and exits immediately. The path must stay within the \`/studio/\` route tree so the Studio shell (header, session management) stays mounted. The \`?session=\` param is appended automatically by the shell's navigate handler.

If the Studio was closed early (\`session-ended\` event from the background process), skip this step and continue directly to \`build-dataset\`.

After opening, tell the user you've opened the dataset page and are waiting for them to finish reviewing. Then proceed to \`await-event\`.`,
                    next: "phase-3/await-event",
                },
                {
                    id: "await-event",
                    kind: "branch",
                    toolCalls: ["bash", "monitor"],
                    title: "Wait for user to finish dataset review",
                    body: `**Wait for user to finish dataset review.** Use the **Monitor tool** to watch for the next JSON event from the Studio background process (\`openStudio.js\`, started in \`studio/open\`). The output file path was returned when you started the background process. Set up a monitor that tails only NEW lines (skip lines already read) and filters for JSON event lines:

\`\`\`bash
tail -f -n +<NEXT_LINE> <output-file> | grep -E --line-buffered '"event"'
\`\`\`

Where \`<NEXT_LINE>\` is the line number after the last line you already processed (e.g. if you read line 1 with the \`started\` event, use \`-n +2\`). Track which line you last processed so you don't re-fire on old events.

The monitor will notify you when a new event arrives. Do NOT just read the output file once and wait, that will cause the flow to stall. The Studio process emits these events relevant to dataset review:

- \`{"event":"edit-with-agent","sessionId":"...","datasetId":"..."}\` — the user clicked **Edit with agent**. Go to the modify loop, then come back here.
- \`{"event":"return-to-agent","sessionId":"..."}\` — the user clicked **Done**, which triggered \`returnToStudio()\` and navigated back to \`/studio\`. Dataset review is complete.
- \`{"event":"session-ended","sessionId":"..."}\` — the user closed Studio entirely.

Filter to JSON lines only (skip status text). Route on the \`event\` field:

{{whens}}`,
                    branches: [
                        {
                            when: "`event: edit-with-agent`",
                            description: "user clicked Edit with agent on the dataset page. Go to the modify loop, then come back here to read the next event",
                            next: "phase-3/modify-loop",
                        },
                        {
                            when: "`event: return-to-agent`",
                            description: "user clicked Done on the dataset page. Dataset review is complete, move on to build + confirm the dataset",
                            next: "phase-3/build-dataset",
                        },
                        {
                            when: "`event: session-ended`",
                            description: "user closed Studio. Stop the flow",
                            next: null,
                        },
                    ],
                },
                {
                    id: "modify-loop",
                    kind: "action",
                    toolCalls: [
                        "searchTraces",
                        "readTraces",
                        "updateAgentLabels",
                        "addTracesToDataset",
                        "removeTracesFromDataset",
                    ],
                    title: "Modify loop: add or remove traces in chat",
                    body: `**Modify loop: add or remove traces in chat** — The dataset page is still open in Studio and the user wants you to add or remove traces. Ask in plain chat:

> What would you like to add or remove? You can describe by criteria (e.g. "drop empty-output traces", "add 5 more from last week with errors") or paste explicit trace IDs.

Then wait for the user's next message. It will contain their answer. Do NOT {{askVerb}} here (the answer is free-form and options would just add an extra step before the user can type).

Then act on it:

- **Adding traces:** find candidates with {{tool:searchTraces}} / {{tool:readTraces}}, then respect the labeling mode the user chose earlier in this phase (the ask-labeling-mode step). In **agent-first mode (Option A)**, label them yourself with {{tool:updateAgentLabels}} (same rigor as label-self: every trace gets a verdict + annotation, grounded in the code) before attaching. In **manual mode (Option B)**, do NOT call {{tool:updateAgentLabels}}. **If no labeling mode was selected** (the user took the Reuse → Move-on path that bypasses ask-labeling-mode, or find-unlabeled returned no candidates so ask-labeling-mode self-skipped), default to **agent-first mode (Option A)** — match the recommended default and label new candidates yourself before attaching. Either way, call {{tool:addTracesToDataset}} to attach.
- **Removing traces:** call {{tool:removeTracesFromDataset}} with the trace IDs to remove. The traces themselves aren't deleted, only their membership in the dataset.

The dataset page reflects each add/remove live (SSE), so the user sees changes flow in as you make them. When you're done, summarize what changed in chat and **return to the await-event step to read the next event**. The user can click Edit with agent again for another modify round, or Done to finalize.`,
                    next: "phase-3/await-event",
                },
                {
                    id: "build-dataset",
                    kind: "action",
                    toolCalls: ["readTraces"],
                    title: "Build the dataset",
                    body: `**Build the dataset** — You already know the trace IDs in this dataset (you attached them in earlier steps and tracked any add/remove from modify rounds). Call {{tool:readTraces}} with all of them and \`scope: "full"\` to load the labels + annotations into context. This is the working set for confirm + every Phase 5 experiment.`,
                    next: "phase-3/confirm-dataset",
                },
                {
                    id: "confirm-dataset",
                    kind: "branch",
                    toolCalls: ["searchTraces", "ask"],
                    title: "Confirm the dataset",
                    body: `**Confirm the dataset** — Present the dataset {{presentMode}}: each entry showing (trace ID, label, annotation summary). The dataset must contain at least one **validated failing label** — i.e. at least one trace where a human either authored or approved a \`false\` label. To check, call {{tool:searchTraces}} restricted to the dataset trace IDs with \`validated: true\` and \`labelResult: false\`. Two outcomes:

{{whens}}

Unapproved agent labels do **not** satisfy this gate by design — \`validated: true\` excludes them.`,
                    branches: [
                        {
                            when: "gate fails (no validated failing label — search returns nothing)",
                            description: "tell the user and loop back to find or label more unlabeled traces",
                            next: "phase-3/find-unlabeled",
                        },
                        {
                            when: "gate passes (at least one validated failing label)",
                            description: "get explicit approval, then continue",
                            next: "phase-3/hold-in-context",
                        },
                    ],
                },
                {
                    id: "hold-in-context",
                    kind: "action",
                    title: "Hold in-context",
                    body: `**Hold in-context** — This approved dataset is the benchmark for all experiments in Phase 5. Keep both the \`datasetId\` and the trace IDs in your working context throughout. In \`dataset\` mode, stop here: kill the Studio background process (send SIGINT or abort the background task). Surface the dataset summary (including the id) and exit so they can pick up later with \`{{cmd}}assistant experiment <key> <datasetId>\`.`,
                    next: {
                        byMode: {
                            all: "phase-4/understand-failures",
                            dataset: null,
                            experiment: null,
                        },
                    },
                },
            ],
        },
        {
            id: "phase-4",
            title: "Phase 4: Diagnose & Plan",
            stepStyle: "list",
            steps: [
                {
                    id: "understand-failures",
                    kind: "action",
                    toolCalls: ["readTraces"],
                    title: "Understand failures",
                    body: `**Understand failures.** Using the failed traces you read in Phase 3 (or read them now if you haven't):

- Call {{tool:readTraces}} on 3–5 failed traces with \`scope: "full"\`

Synthesize the failure patterns — what's going wrong, what the common threads are.`,
                    next: "phase-4/read-code",
                },
                {
                    id: "read-code",
                    kind: "action",
                    toolCalls: ["read"],
                    title: "Read the code",
                    body: `**Read the code.**

- Find the instrumented function in the codebase (in \`all\` mode you found it in Phase 2; this step is unreachable in \`dataset\` / \`experiment\` modes)
- Read the full implementation — follow the call chain to understand the logic
- Identify **iteration targets**: prompts, system messages, parameters, preprocessing, postprocessing
- If BAML files are involved, read the relevant \`.baml\` files`,
                    next: "phase-4/categorize-fixes",
                },
                {
                    id: "categorize-fixes",
                    kind: "action",
                    toolCalls: ["ask"],
                    title: "Categorize fixes based on failure annotations",
                    body: `**Categorize fixes based on failure annotations.** Based on the failure patterns, the code, and the labeled dataset from Phase 3, categorize proposed changes into three buckets:

**Bucket 1 — Code fixes**: Deterministic bugs (off-by-one, type mismatch, missing null check, wrong variable). These won't recur once fixed. Bundle all code fixes into a single experiment unless they are large feature changes. These are applied first as a foundation that all subsequent experiments build on.

**Bucket 2 — Judgment-based fixes**: Prompt changes, context truncation, search tuning, output formatting, etc. These require the user's judgment to evaluate correctness. Each gets its own experiment.

**Bucket 3 — Infrastructure proposals**: Larger changes that require new infrastructure, architectural changes, or significant feature work. These are separated out because experiments become harder to compare when some include large infra changes and others don't — apples-to-apples comparison requires a consistent baseline. Do not run experiments for these. Instead, if the user has integrations (Linear, Notion, Jira), propose creating a task with a clear writeup for future work.

Present the categorized plan {{presentMode}}:

> "Based on the N traces in the dataset, here's what I see:
>
> **Code fixes** (experiment #1 — bundled):
>
> - [Fix]: [What and why, which traces it addresses]
>
> **Judgment-based experiments** (#2, #3, ...):
>
> - [Experiment]: [What change, which traces it targets, hypothesis]
>
> **Future infrastructure** (not experiments):
>
> - [Proposal]: [What it would require, which traces it would help]
>
> I'll replay each experiment against the labeled dataset and evaluate using the annotations as acceptance criteria."

Get the user's confirmation before proceeding.`,
                    next: {
                        byMode: {
                            all: "phase-5/make-change",
                            dataset: null,
                            experiment: null,
                        },
                    },
                },
            ],
        },
        {
            id: "phase-5",
            title: "Phase 5: Iterate with Replay",
            intro: `Run an iterative improvement loop. Each iteration:

The Studio is already open (launched in the \`studio/open\` step at the start of the flow). Use the \`sessionId\` captured there for all \`navigateStudio.js\` calls. If the Studio was closed early (\`session-ended\` event), skip navigation calls but continue the improve loop normally.`,
            stepStyle: "list",
            steps: [
                {
                    id: "pick-dataset",
                    kind: "branch",
                    toolCalls: ["bash", "grep", "listDatasets", "readTraces", "ask"],
                    title: "Pick a dataset and locate the code",
                    body: `The trace function key comes from the argument and no prior phase has run. Pick the dataset to iterate against, then locate the code:

1. **Grep the codebase** for the trace function key (e.g. \`grep -r "<traceFunctionKey>" --include="*.ts" --include="*.tsx" --include="*.py" --include="*.rb" --include="*.go" --include="*.baml"\`) and note the file path. This is the code you'll iterate on.
2. **Pick the dataset.** If a \`<dataset-id>\` argument was provided, use it directly. Otherwise call {{tool:listDatasets}} with the trace function key, present the result to the user {{presentMode}}, and use their choice. Hold the chosen \`datasetId\` in working context.
3. **Load it.** Call {{tool:readTraces}} with the dataset's trace IDs and \`scope: "full"\` so labels + annotations are in context.
4. **Branch on the result:**

{{whens}}`,
                    branches: [
                        {
                            when: "no datasets exist for this function (`list_datasets` returned empty), or the picked dataset has no validated failing labels",
                            description: `tell the user the function has no usable dataset yet and recommend running \`{{cmd}}assistant dataset <key>\` first; kill the Studio background process; then stop the flow`,
                            next: null,
                        },
                        {
                            when: "dataset loaded (≥1 validated failing label)",
                            description: `summarize the dataset for the user (counts of pass/fail) and the failure annotations. Pick a first experiment from the failure patterns and continue`,
                            next: "phase-5/pick-execution-mode",
                        },
                    ],
                },
                {
                    id: "pick-execution-mode",
                    kind: "branch",
                    toolCalls: ["bash"],
                    title: "Pick parallel or serial execution",
                    body: `**Decide once: parallel worktree subagents, or serial in this main agent.** The check is whether subagent worktree sessions would inherit bypass permissions.

{{#claude}}\`.claude/settings.local.json\` is gitignored and does NOT propagate into subagent worktrees, so it can't grant bypass. The two locations that DO propagate are committed \`.claude/settings.json\` and user-global \`~/.claude/settings.json\`. Run:

\`\`\`bash
python3 -c "
import json, os
def has_bypass(p):
    if not os.path.exists(p): return False
    try: d = json.load(open(p))
    except Exception: return False
    return (d.get('permissions') or {}).get('defaultMode') == 'bypassPermissions'
p = has_bypass('.claude/settings.json')
g = has_bypass(os.path.expanduser('~/.claude/settings.json'))
print('parallel' if (p or g) else 'serial')
"
\`\`\`{{/claude}}{{^claude}}This editor doesn't support worktree-isolated subagents. Skip the bash check and take the serial branch.{{/claude}}

Hold the chosen mode in working context. Every iteration below (\`make-change\`, \`replay-against-dataset\`, \`evaluate-results\`) honors it.

{{whens}}`,
                    branches: [
                        {
                            when: "{{#claude}}bash output is `parallel` (bypass found in committed or user-global settings){{/claude}}{{^claude}}(unreachable on this editor){{/claude}}",
                            description: `**Parallel mode.** For each independent experiment, fork to a subagent using the Agent tool with \`isolation: "worktree"\` and \`subagent_type: "general-purpose"\`. The subagent edits its worktree, runs replay, returns its scored items + \`testRunId\` to this main agent`,
                            next: "phase-5/make-change",
                        },
                        {
                            when: "{{#claude}}bash output is `serial` (no bypass found){{/claude}}{{^claude}}always{{/claude}}",
                            description: `**Serial mode.** Iterate experiments one at a time in this main agent. Subagent worktrees wouldn't inherit bypass permissions, so their Edit tool would be denied`,
                            next: "phase-5/make-change",
                        },
                    ],
                },
                {
                    id: "make-change",
                    kind: "action",
                    toolCalls: ["ask", "edit", "agent"],
                    title: "Make the change",
                    body: `**Make the change.**

- {{^codex}}Use \`AskUserQuestion\` to explain{{/codex}}{{#codex}}Explain to the user{{/codex}} what you're changing and why, and confirm before editing
- Edit the iteration target (prompt, code, tools, parameters)`,
                    next: "phase-5/replay-against-dataset",
                },
                {
                    id: "replay-against-dataset",
                    kind: "action",
                    toolCalls: ["bash"],
                    title: "Replay against the dataset",
                    body: `**Replay against the dataset.** Collect the trace IDs from the labeled dataset (built in Phase 3 in \`all\` mode, or rehydrated at the start of this phase in \`experiment\` mode). Run the replay script with those specific traces.

\`\`\`bash
# The exact command depends on the replay script — adapt to what exists
# Example for TypeScript:
cd <project-dir> && npx tsx scripts/replay.ts <pipeline-name> --trace-ids <id1>,<id2>,<id3>,...
\`\`\`

**Before running: verify the replay script prints the full original and new output values to stdout for every item** (not just lengths, counts, hashes, or truncated previews). If it doesn't, fix the script first — the Replay Output Contract and example script live in the SDK reference at \`https://docs.bitfab.ai/<language>-sdk#replay\`. Subagents can't evaluate an improvement from \`5 → 7 (+2)\`.

**Capture the \`testRunId\` from the replay output** — the SDK prints it (alongside \`testRunUrl\`) when the run completes. Track every \`testRunId\` produced across all iterations of this phase: you'll feed them to \`open-experiments\` so the user can review every experiment side-by-side in one viewer.

**If a child span fails during replay, tag it with \`mockOnReplay\` instead of debugging it.** When a non-root span throws (missing API key for a paid call, flaky external service, deleted/moved dependency, env not reproducible), it usually blocks the whole trace from completing, even though the failure is environmental, not a bug in the function you're iterating on. The short-term fix is to mark that span as replayable from its recorded output:

1. Find the failing span's call site in the codebase (\`withSpan("<spanName>", ...)\` in TS, \`@bitfab.span\` / \`bitfab.span\` equivalents in other SDKs).
2. Add the flag to its span declaration (TypeScript and Python today; Ruby and Go as they land):
   \`\`\`ts
   // TypeScript: SpanOptions.mockOnReplay
   bitfab.withSpan("expensive-llm-call", { mockOnReplay: true }, async () => { ... })
   \`\`\`
   \`\`\`python
   # Python: mock_on_replay kwarg on @client.span(...)
   @client.span("expensive-llm-call", mock_on_replay=True)
   def expensive_llm_call(...):
       ...
   \`\`\`
3. Re-run the replay script passing \`mock: "marked"\` to \`client.replay(...)\` (or \`mock="marked"\` in Python). That child will return its historical output; the root function still runs real code.
4. Flag the tag to the user: it's a replay-only escape hatch, has no effect on prod execution, and is worth removing once the underlying issue is fixed.

Use this when the goal is to unblock iteration on the root function, not when the child itself is what you're trying to improve.

**After the run, classify items before evaluating.** A failed item means one of two things: the new code produced a bad output (real signal), or the wrapped fn threw on infra (missing DB row, stale FK, rejected write, missing env). Infra failures are not regressions.

From the JSON compute:

- \`completed\` — \`item.error\` unset
- \`infraErrored\` — \`item.error\` set
- \`total\` — \`result.items.length\`; \`0\` or non-zero exit code = whole-replay crash

If \`completed === 0\`, do not score pass/fail on an empty set — branch to \`check-replay-health\`.`,
                    next: "phase-5/check-replay-health",
                },
                {
                    id: "check-replay-health",
                    kind: "branch",
                    toolCalls: ["bash", "read", "edit", "ask"],
                    title: "Check replay health",
                    body: `**Route on the counts and exit code.** Goal: keep infra noise out of evaluation. Read a sample of \`item.error\` strings (and stderr on crash) first to identify the DB-shaped pattern (missing record, FK / unique constraint, write rejected, connection refused, missing env).

**🚨 Do not silently work around DB issues.** Do not drop affected trace IDs, stub the read in the script, gate writes behind a script-only flag, wrap the function in a rollback transaction, or edit the instrumented function to skip DB calls. Those all hide infra problems as fake passing or fake failing results and corrupt the experiment.

**Instead: tell the user what's wrong and offer exactly two workarounds.** Use {{askVerb}} to surface a clear summary first — the failing trace ID(s), the error pattern, the function and span where it happens — then present the two options below. Pick a representative failing trace and call {{tool:readTraces}} with \`scope: "summary"\` to read its \`environment\` field (production / staging / development), so option B can name the source environment concretely.

- **Workaround A: \`mockOnReplay\`** *(recommended for spans whose side effects shouldn't run during experimentation)* — apply the \`mockOnReplay\` recipe from step \`replay-against-dataset\` above (find the failing span, add \`mockOnReplay: true\` to its \`SpanOptions\`, re-run with \`{ mock: "marked" }\`). Edit only the span options, never the function body. Use this when the span is a DB read/write the experiment isn't testing and the captured output can stand in for it.
- **Workaround B: Point replay at the trace's source database** — the trace's \`environment\` field names where it was captured (e.g. \`production\`). Tell the user that's the only environment whose DB has the rows the trace references, then offer to (i) update the replay env to point at that environment's DB (env vars, connection string) or (ii) ask which environment they want to use if multiple are valid. Apply the change to env / config, not to the function under test.

After whichever workaround the user picks, re-run \`replay-against-dataset\` and re-check health. If the user can't or won't do either, stop and report — don't fabricate a workaround on your own.

{{whens}}`,
                    branches: [
                        {
                            when: "whole replay crashed (non-zero exit, total is 0, or unparseable stdout)",
                            description: `show stderr / exit code, diagnose, confirm a script fix with the user, apply, loop back to \`replay-against-dataset\``,
                            next: "phase-5/replay-against-dataset",
                        },
                        {
                            when: "every item errored (completed is 0 but total is non-zero)",
                            description: `systemic infra failure (usually env mismatch). Diagnose, confirm a script fix with the user, loop back`,
                            next: "phase-5/replay-against-dataset",
                        },
                        {
                            when: "high infra error rate (over half of items errored)",
                            description: `signal is noisy. Flag the rate and ask the user whether to fix the env and retry, or proceed with the partial signal`,
                            next: "phase-5/evaluate-results",
                        },
                        {
                            when: "healthy or mixed run (at least one completed item, infra errors at most half of total)",
                            description: `proceed. Carry \`infraErrored\` forward — surface as its own bucket in share-results, never folded into pass/fail`,
                            next: "phase-5/evaluate-results",
                        },
                    ],
                },
                {
                    id: "evaluate-results",
                    kind: "action",
                    toolCalls: ["read", "write", "bash"],
                    title: "Evaluate against labels & annotations",
                    body: `**Evaluate against labels & annotations.** Score only items where \`item.error\` is unset. Items with \`item.error\` set are unreplayable (already classified) and go in their own bucket — never pass, fail, or regression.

For each completed (non-errored) replay item, derive a verdict by comparing the replay's new output against the original trace's label and annotation (from Phase 3, or rehydrated in \`experiment\` mode):

- **fail**-labeled original: does the replay's new output address the annotation? If yes → \`label: true\` (PASS). If no → \`label: false\` (FAIL). Use the annotation as the acceptance criterion.
- **pass**-labeled original: preserved → \`label: true\` (PASS). regressed → \`label: false\` (FAIL).
- Cannot judge from the output alone (genuinely ambiguous, not laziness): \`skip: true\` instead of guessing. Skips are recorded explicitly so the verify step knows you intentionally did not verdict.
- Unreplayable items (\`item.error\` set) are NOT verdicted here — keep their list (trace ID + error string) for \`share-results\`.

**The verdict you produce here is persisted onto the REPLAY trace IDs (not the originals).** That's what makes "did this fix actually pass on replay?" queryable across iterations.

**Persist via \`persistReplayLabels.js\`.** Write the verdicts to a tmp JSON file then run the script — one Bash call, one batched MCP call server-side, file is auto-deleted on success:

1. Pick a tmp path. Recommended: \`.bitfab/tmp/verdicts-<testRunId>.json\` (create the dir if missing). Falls back to \`os.tmpdir()\` if the project root isn't writable.
2. Use the {{tool:write}} tool to write JSON of this exact shape:

\`\`\`json
{
  "expectedTraceIds": ["<replayTraceId1>", "<replayTraceId2>", "..."],
  "verdicts": [
    { "traceId": "<replayTraceId1>", "label": true, "annotation": "Now returns the missing field; original annotation said it was empty.", "confidence": "High" },
    { "traceId": "<replayTraceId2>", "label": false, "annotation": "Output still hallucinates a tool argument.", "confidence": "VeryHigh" },
    { "traceId": "<replayTraceId3>", "skip": true }
  ]
}
\`\`\`

\`expectedTraceIds\` MUST be the full set of REPLAY trace IDs you committed to verdict (every completed item from the run). \`verdicts\` MUST have one entry per ID — either a \`{label, annotation, confidence?}\` verdict or a \`{skip: true}\` explicit skip. \`confidence\` is optional but recommended (\`VeryLow|Low|Medium|High|VeryHigh\`); it surfaces in the labeling UI so reviewers can prioritize low-confidence verdicts. If verdict counts don't match \`expectedTraceIds\`, the script returns \`status: "missing-coverage"\` and the verify step routes you back to fill the gaps.

3. Run the script:

\`\`\`bash
node "{{pluginRoot}}/dist/commands/persistReplayLabels.js" .bitfab/tmp/verdicts-<testRunId>.json
\`\`\`

4. Read its single JSON line on stdout. Hold the parsed result for the next step.

**Spill working notes to a separate tmp file if context gets big.** Don't conflate working notes with the verdicts file — the script deletes the verdicts file on success.{{#claude}}

**If you're a worktree subagent** (parallel mode from \`pick-execution-mode\`): after the script returns, hand the parsed result + \`testRunId\` + unreplayable list back to the main agent and exit. The main agent collects results from all parallel experiments before \`open-experiments\`.{{/claude}}`,
                    next: "phase-5/verify-replay-labels",
                },
                {
                    id: "verify-replay-labels",
                    kind: "branch",
                    toolCalls: ["read", "write", "bash", "readTraces"],
                    title: "Verify replay labels persisted",
                    body: `**Verify replay labels persisted.** Route on the \`status\` field of the JSON the script printed in \`evaluate-results\`. The script is the deterministic gate — if it didn't return \`ok\`, the agent's verdicts are NOT yet on the replay traces and the experiment delta will be wrong on the next iteration.

{{whens}}`,
                    branches: [
                        {
                            when: '`status: "ok"` (every replay trace has a verdict or explicit skip persisted)',
                            description: "labels are persisted on the replay traces and the verdicts file is gone. Continue to open-experiments",
                            next: "phase-5/open-experiments",
                        },
                        {
                            when: '`status: "missing-coverage"` (script returned a non-empty `missingTraceIds` array)',
                            description: 'you under-verdicted. Read the missing replay trace IDs (use {{tool:readTraces}} with `scope: "summary"` or `"full"` if you didn\'t already), decide each one (PASS / FAIL with annotation, or `skip: true` if genuinely ambiguous), write a NEW verdicts file at the same path covering ALL the originally expected IDs (the script needs the full `expectedTraceIds` list each call, not just the gaps), and re-run the script. Loop back here with the new result',
                            next: "phase-5/verify-replay-labels",
                        },
                        {
                            when: '`status: "invalid-input"` (malformed verdicts JSON or missing fields)',
                            description: "the verdicts file you wrote doesn't match the schema. Read the script's `message` field, fix the JSON (most common: missing annotation on a non-skip entry, missing traceId, expectedTraceIds empty), and re-run the script. Loop back here",
                            next: "phase-5/verify-replay-labels",
                        },
                        {
                            when: '`status: "mcp-error"` (MCP call to update_agent_labels failed mid-batch)',
                            description: "network or auth error. The script's `partialTraceIds` lists which IDs were already persisted. Tell the user, recommend re-running the script (it's idempotent — already-persisted labels just upsert), and loop back here. If it keeps failing, stop and surface the error",
                            next: "phase-5/verify-replay-labels",
                        },
                    ],
                },
                {
                    id: "open-experiments",
                    kind: "action",
                    toolCalls: ["bash"],
                    title: "Open experiment viewer",
                    body: `**Open experiment viewer.** If no \`testRunId\`s were captured (e.g. the replay script didn't print them), skip this step and continue, but flag it to the user in \`share-results\` so the script can be fixed before the next iteration.

Navigate the already-open Studio to the experiments page using the \`sessionId\` captured in the \`studio/open\` step. Build the path with **every** \`testRunId\` you've collected across iterations of this phase (comma-separated):

\`\`\`bash
node "{{pluginRoot}}/node_modules/bitfab-plugin-lib/dist/commands/navigateStudio.js" <sessionId> "/studio/experiments?testRunIds=<testRunId1>,<testRunId2>,<testRunId3>"
\`\`\`

The command sends a navigate event and exits immediately. If the Studio was closed early (the background process exited with a \`session-ended\` event), skip this step entirely.

The user reviews the viewer alongside your \`share-results\` summary before deciding whether to iterate.`,
                    next: "phase-5/share-results",
                },
                {
                    id: "share-results",
                    kind: "branch",
                    toolCalls: ["ask"],
                    title: "Share results to the user",
                    body: `**Share results to the user.**

> "After N experiments these are the results: X/Y traces now pass (Z unreplayable, excluded from pass/fail).
>
> - ✅ Trace \`abc123\`: Now passes — [how the annotation's issue was resolved]
> - ❌ Trace \`def456\`: Still failing — annotation said [X], output still [Y]
> - ❌⚠️ Trace \`ghi789\`: Was passing, now failing (regression)
> - ⚠️ Trace \`jkl012\`: Unreplayable — [DB record not found / FK violation / write rejected]"

Keep \`unreplayable\` out of the pass-rate denominator. If \`unreplayable > 0\`, name the cause (missing record, write blocked, env mismatch) and note that fixing the env or trimming those trace IDs will clean up the next iteration. If \`check-replay-health\` fired in the previous iteration too, flag that infra has now blocked two runs and recommend fixing it before another experiment.

Show this across the full data set, and highlight the best outcome concisely. Explain why it worked best with references to code, docs, and/or research if needed. For the best outcome:

- **If pass rate improved and no regressions**: {{askVerb}} to confirm whether they want to keep iterating or stop
- **If pass rate improved but regressions exist or no improvement**: tell the user and propose to create a plan for new experiments and continue iterating.

Ensure your question includes your recommended next step.

{{branches}}`,
                    branches: [
                        {
                            option: {
                                letter: "A",
                                label: "Keep iterating",
                                description: "run another experiment from the plan",
                            },
                            recommended: true,
                            next: "phase-5/make-change",
                        },
                        {
                            option: {
                                letter: "B",
                                label: "Stop and wrap up",
                                description: "move to the final summary",
                            },
                            next: "phase-6/summary",
                        },
                    ],
                },
            ],
        },
        {
            id: "phase-6",
            title: "Phase 6: Validate & Wrap Up",
            stepStyle: "list",
            steps: [
                {
                    id: "summary",
                    kind: "action",
                    toolCalls: ["ask"],
                    title: "Summary",
                    body: `**Summary.** {{^codex}}Use \`AskUserQuestion\` to present{{/codex}}{{#codex}}Present{{/codex}} the final results similar to this. You may expand where appropriate based on context from the user:

> "**Improvement summary for** \`<traceFunctionKey>\`:
>
> - Failed traces fixed: X/Y (from N% → M% pass rate on labeled failures)
> - Full replay pass rate: A/B (Z unreplayable, excluded)
> - Changes made:
>   - [File]: [Description of change]
>   - [File]: [Description of change]
>
> The changes are in your working tree (not committed). Review the diffs and commit when ready."

Kill the Studio background process (send SIGINT or abort the background task).

If \`Z > 0\`, add one line naming the infra cause (e.g. "Z traces unreplayable — missing DB rows; refresh the dataset or scope to a snapshot next pass") so the user has a next step beyond the code.`,
                    next: null,
                },
            ],
        },
    ],
});
