// bitfab-plugin-lib/dist/flows/improve.js
import { Flow } from "./schema.js";
import { CODEX_RESOLVE_PLUGIN_DIR_BASH } from "./shared.js";
export const improveFlow = Flow.parse({
    schemaVersion: 1,
    id: "improve",
    title: "Bitfab Improve",
    frontmatter: {
        description: {
            claude: "Iterate on a traced function to improve pass rates using failed traces, labeling, and replay",
            cursor: "Improve a traced function with Bitfab",
            codex: "Iterate on a traced Bitfab function to improve its pass rate. Use when the user wants to review failing traces, label examples, and replay an LLM function until it passes",
        },
        argumentHint: "<trace-function-key>",
    },
    tools: {
        bash: { kind: "builtin", name: "Bash" },
        read: { kind: "builtin", name: "Read" },
        glob: { kind: "builtin", name: "Glob" },
        grep: { kind: "builtin", name: "Grep" },
        edit: { kind: "builtin", name: "Edit" },
        write: { kind: "builtin", name: "Write" },
        agent: { kind: "builtin", name: "Agent" },
        ask: { kind: "builtin", name: "AskUserQuestion" },
        skill: { kind: "builtin", name: "Skill" },
        listFunctions: { kind: "mcp", name: "list_trace_functions" },
        searchTraces: { kind: "mcp", name: "search_traces" },
        readTraces: { kind: "mcp", name: "read_traces" },
        updateAgentLabels: { kind: "mcp", name: "update_agent_labels" },
    },
    intro: `Use the local plugin MCP tools ({{tool:listFunctions}}, {{tool:searchTraces}}, {{tool:readTraces}}, {{tool:updateAgentLabels}}) to find what's failing in a traced function, gather labeled failed traces, then iterate on the code/prompts using replay until pass rates improve.

**MCP tools:** This skill uses \`list_trace_functions\`, \`search_traces\`, \`read_traces\`, and \`update_agent_labels\` from the **local plugin MCP server** (bundled with this plugin){{#claude}}. Do NOT use the remote Bitfab MCP tools (\`mcp__Simforge__*\` or \`mcp__Bitfab__*\`) — use only the \`mcp__plugin_bitfab_Bitfab__*\` variants.{{/claude}}{{^claude}}, exposed under the \`mcp__Bitfab__*\` prefix.{{/claude}}

{{^codex}}**Always use** \`AskUserQuestion\` **when asking questions, reporting results, or presenting choices.** Never print a question as text and wait. Rules:{{/codex}}{{#codex}}**When the flow branches, always present the options clearly and wait for the user's answer before proceeding.** Number or letter the options so the user can pick by reference. Rules:{{/codex}}

- Recommend an option first, explain why in one line
- Present 2-5 concrete options
- One decision per question — never batch{{#codex}}

🚨 **Blocking-process rule (applies to any plugin command described as "blocks until the user does X"):** When you launch a plugin CLI that blocks on a browser handoff (\`startDataset.js\`, \`login.js\`, etc.), you MUST keep the exec session alive and keep polling it until the process exits on its own.

- Printing the URL is NOT completion. The process is still running, racing a loopback callback and a server-polled ticket — it will exit only after the user's browser action delivers a result (or after the timeout).
- After sending the URL to the user, keep polling the live shell/exec session at least every few seconds with your normal "read more output" tool (\`write_stdin\`, \`read\`, or whatever your runtime's equivalent is for the long-running shell). Do not idle waiting for a user message.
- Do not send a final "waiting for you to click Confirm" text and then stop polling — the user's confirmation does NOT come back to you as a chat message; it comes back as the plugin process exiting with output on stdout.
- Stop polling only when one of: (a) the process exits 0 and prints its completion summary, (b) the process exits non-zero, or (c) the user explicitly cancels.
- When the process exits, immediately continue with the next step in the flow — don't wait for another user message.{{/codex}}`,
    entries: {
        default: {
            default: "phase-1/list-or-arg",
            codex: "phase-0/resolve-plugin-dir",
        },
    },
    defaultMode: "default",
    variables: {
        /**
         * How a question is presented to the user. Claude / Cursor have the
         * `AskUserQuestion` tool ("via AskUserQuestion"); Codex doesn't and
         * presents options inline ("as a numbered choice"). Used wherever the
         * body says "Present the … to the user {{presentMode}}: …".
         */
        presentMode: {
            default: "via `AskUserQuestion`",
            codex: "as a numbered choice",
        },
    },
    phases: [
        {
            id: "phase-0",
            title: "Phase 0: Status + Update Check",
            stepStyle: "list",
            steps: [
                {
                    id: "resolve-plugin-dir",
                    kind: "action",
                    preamble: true,
                    toolCalls: ["bash"],
                    title: "Resolve BITFAB_PLUGIN_DIR",
                    body: `First, resolve \`BITFAB_PLUGIN_DIR\` if it isn't already exported in this shell. Run this block verbatim — it auto-detects dev / prod / custom-\`CODEX_HOME\` installs:

${CODEX_RESOLVE_PLUGIN_DIR_BASH}`,
                    next: "phase-0/run-status",
                },
                {
                    id: "run-status",
                    kind: "branch",
                    toolCalls: ["bash"],
                    title: "Run the status command",
                    body: `Then run the status command:

\`\`\`bash
node "{{pluginRoot}}/dist/commands/status.js"
\`\`\`

Watch the output and route on it:

{{whens}}`,
                    branches: [
                        {
                            when: "not authenticated",
                            description: "stop the flow immediately. Tell the user to run `{{cmd}}setup login` first",
                            next: null,
                        },
                        {
                            when: "authenticated (with or without a `v<X> available` upgrade notice)",
                            description: "continue to Phase 1. If an upgrade notice appeared, pass it through to the user verbatim — but don't block on it; surface the notice once and move on",
                            next: "phase-1/list-or-arg",
                        },
                    ],
                },
            ],
        },
        {
            id: "phase-1",
            title: "Phase 1: Identify the Trace Function",
            intro: "If a `traceFunctionKey` was provided as an argument, skip the listing and the user prompt — but still cross-check the provided key against the local codebase before moving on. Otherwise, work through all four steps below:",
            stepStyle: "list",
            steps: [
                {
                    id: "list-or-arg",
                    kind: "action",
                    toolCalls: ["listFunctions"],
                    title: "List trace functions",
                    body: `**Skip this step if a \`traceFunctionKey\` argument was provided** — use the argument directly and continue to cross-check. Otherwise, call {{tool:listFunctions}} to list all available trace functions. Use **only** the keys and metadata returned (trace counts, last activity) — do NOT invent or infer descriptions of what each function does from its key name. Key names are often ambiguous or misleading, and guessing produces hallucinated descriptions that confuse the user.`,
                    next: "phase-1/cross-check",
                },
                {
                    id: "cross-check",
                    kind: "action",
                    toolCalls: ["grep", "glob"],
                    title: "Cross-check against the local codebase",
                    body: `**Cross-check each key against the local codebase** before presenting. For each returned key, \`grep\` the repo for string-literal uses of that exact key (across \`*.ts\`, \`*.tsx\`, \`*.py\`, \`*.rb\`, \`*.go\`, \`*.baml\`). Mark each function in the presented list as:

- **✅ instrumented here** — found in this repo, with the file path
- **⚠️ not found in this repo** — traces exist on Bitfab but the key isn't in this codebase (likely another repo or a renamed key)`,
                    next: "phase-1/present-list",
                },
                {
                    id: "present-list",
                    kind: "action",
                    title: "Present the list",
                    body: `**Skip this step if a \`traceFunctionKey\` argument was provided** — there's no list to present. Otherwise, present the full list in the question text showing ONLY: \`<key>\` · \`<trace count>\` · \`<last activity>\` · \`<instrumented-here marker + path, or not-found marker>\`. No invented summaries.`,
                    next: "phase-1/ask-function",
                },
                {
                    id: "ask-function",
                    kind: "action",
                    toolCalls: ["ask"],
                    title: "Ask which function to improve",
                    body: `**Skip this step if a \`traceFunctionKey\` argument was provided** — the function is already chosen. Otherwise, {{askVerb}} with 2 options: the recommended function (prefer one that is ✅ instrumented here AND has recent activity) and a free-text "Type a function key" option. If nothing is instrumented here, say so explicitly in the question — don't hide it.`,
                    next: "phase-2/check-instrumentation",
                },
            ],
        },
        {
            id: "phase-2",
            title: "Phase 2: Verify Instrumentation & Replay",
            intro: "Check that this trace function has both instrumentation and a replay script.",
            stepStyle: "list",
            steps: [
                {
                    id: "check-instrumentation",
                    kind: "branch",
                    toolCalls: ["bash", "grep", "ask", "skill"],
                    title: "Check instrumentation",
                    body: `Search the codebase for the trace function key to {{^codex}}find where the SDK is used{{/codex}}{{#codex}}confirm SDK usage{{/codex}}:

- TypeScript: \`grep -r "<traceFunctionKey>" --include="*.ts" --include="*.tsx"\`
- Python: \`grep -r "<traceFunctionKey>" --include="*.py"\`
- Ruby: \`grep -r "<traceFunctionKey>" --include="*.rb"\`
- Go: \`grep -r "<traceFunctionKey>" --include="*.go"\`

If the key is found, note the file location — this is the code you'll iterate on in later phases.

If the key is NOT found in the codebase, the function is instrumented elsewhere (the traces exist on Bitfab). {{^codex}}Use \`AskUserQuestion\` to ask:{{/codex}}{{#codex}}Ask:{{/codex}}

> "I can't find \`<traceFunctionKey>\` in this codebase — it may be instrumented in another repo or under a different key."
>
{{branches}}

If the user chooses **"Instrument now"**, invoke \`{{cmd}}setup instrument\`{{#claude}} using the Skill tool{{/claude}}, then verify whether a replay script exists for this function. If **"Continue anyway"**, skip the replay-script check and start building the dataset — there's no local code to iterate on yet.`,
                    branches: [
                        {
                            when: "key found in codebase",
                            next: "phase-2/check-replay-script",
                        },
                        {
                            option: {
                                letter: "A",
                                label: "Instrument now",
                                description: "set up tracing in this codebase",
                            },
                            recommended: true,
                            next: "phase-2/check-replay-script",
                        },
                        {
                            option: {
                                letter: "B",
                                label: "Continue anyway",
                                description: "work with the traces even without local code",
                            },
                            next: "phase-3/check-validated",
                        },
                        {
                            option: {
                                letter: "C",
                                label: "Pick a different function",
                            },
                            next: "phase-1/list-or-arg",
                        },
                        {
                            option: { letter: "D", label: "Stop" },
                            next: null,
                        },
                    ],
                },
                {
                    id: "check-replay-script",
                    kind: "branch",
                    toolCalls: ["read", "glob", "ask", "skill"],
                    title: "Check replay script",
                    body: `Search for a replay script that covers this trace function:

- Look for files matching \`scripts/replay.*\`, \`scripts/*replay*\`, or any file that imports \`bitfab.replay\` / \`client.replay\`
- Read the script and check that it maps the target trace function key

If a replay script exists but targets a different function key, do NOT modify the existing script or suggest changing the code's function key. Instead, treat it as "no replay script for this function" and offer to create a new one.

If no replay script exists or it doesn't cover this function, {{askVerb}}:

> "No replay script found for \`<traceFunctionKey>\`."
>
{{branches}}

If the user chooses **"Create replay now"**, invoke \`{{cmd}}setup replay\`{{#claude}} using the Skill tool{{/claude}}, then start building the dataset.`,
                    branches: [
                        {
                            when: "replay script covers this function",
                            next: "phase-3/check-validated",
                        },
                        {
                            option: {
                                letter: "A",
                                label: "Create replay now",
                                description: "create the replay script inline",
                            },
                            recommended: true,
                            next: "phase-3/check-validated",
                        },
                        {
                            option: {
                                letter: "B",
                                label: "Pick a different function",
                            },
                            next: "phase-1/list-or-arg",
                        },
                        {
                            option: { letter: "C", label: "Stop" },
                            next: null,
                        },
                    ],
                },
            ],
        },
        {
            id: "phase-3",
            title: "Phase 3: Build Dataset via Labeling",
            intro: `Build a dataset of labeled traces. If there are already enough validated traces, use them directly. Otherwise, the agent labels candidate traces and opens the labeling page for the user to approve or correct them.`,
            stepStyle: "list",
            steps: [
                {
                    id: "check-validated",
                    kind: "branch",
                    toolCalls: ["searchTraces", "ask"],
                    title: "Check existing validated traces",
                    body: `**Check existing validated traces** — Use {{tool:searchTraces}} with \`validated: true\` to find traces with validated labels (human-authored, or agent-authored and approved by a human). Count them, and check whether at least one is a failing label. Present the summary to the user {{presentMode}} (e.g., "Found 8 validated traces (3 pass, 5 fail). Do you want to label more, or proceed with this dataset?"). Note: \`search_traces\` excludes replayed traces (traces from test runs) by default — leave \`includeReplays\` off when labeling, since replays have new inputs (not the original production inputs) and shouldn't be mixed in with the traces you're labeling.

{{branches}}`,
                    branches: [
                        {
                            option: {
                                letter: "A",
                                label: "Use these validated traces",
                                description: "proceed with the existing dataset",
                            },
                            next: "phase-3/build-dataset",
                        },
                        {
                            option: {
                                letter: "B",
                                label: "Label more traces",
                                description: "add more labels before proceeding",
                            },
                            recommended: true,
                            next: "phase-3/find-unlabeled",
                        },
                    ],
                },
                {
                    id: "find-unlabeled",
                    kind: "action",
                    toolCalls: ["searchTraces", "readTraces"],
                    title: "Find unlabeled traces",
                    body: `**Find unlabeled traces** — If more labels are needed, search again without label filters to find unlabeled traces. Use {{tool:readTraces}} with \`scope: "summary"\` to read them and identify which are worth labeling — look for diverse inputs, traces that produced output (not empty), and traces that cover different scenarios. Filter out near-duplicates and uninteresting traces.`,
                    next: "phase-3/present-candidates",
                },
                {
                    id: "present-candidates",
                    kind: "action",
                    toolCalls: ["ask"],
                    title: "Present candidates",
                    body: `**Present candidates** — {{^codex}}Use \`AskUserQuestion\` to show{{/codex}}{{#codex}}Show{{/codex}} which unlabeled traces you recommend labeling and why. Include the already-labeled trace count for context (e.g., "4 traces already labeled, recommending 5 more for labeling"). Let the user approve, adjust, or skip.`,
                    next: "phase-3/label-self",
                },
                {
                    id: "label-self",
                    kind: "action",
                    toolCalls: ["read", "readTraces", "updateAgentLabels"],
                    title: "Label them yourself FIRST (mandatory before opening the labeling page)",
                    body: `**Label them yourself FIRST (mandatory before opening the labeling page)** — Once the user approves the candidate traces, **you** label them. Call {{tool:readTraces}} with \`scope: "full"\` on the approved trace IDs (batch them — up to 10 per call), read each trace's inputs / output / spans yourself, and decide for each one whether it looks like a PASS or a FAIL. **Ground your judgment in the codebase, not just the trace text.** Before you start labeling, read the instrumented function in the user's source (you found it in Phase 2) and any nearby code that explains intent — comments, docstrings, README sections, related tests, BAML files — so you know what the function is *supposed* to do and what "good" looks like for it. Apply the same context to every trace: does this output achieve the function's goal as expressed in the code? Does it match the patterns in the already-validated traces? Then call {{tool:updateAgentLabels}} once with an array of \`{ traceId, label, annotation }\` objects — **both \`label\` (true for pass, false for fail) and \`annotation\` (a one-or-two-sentence explanation written for the human reviewer, ideally referencing what the code is trying to do) are required for every trace**. Commit to a verdict — if you genuinely cannot decide, you didn't read the trace or the code carefully enough. The labels you save here start unapproved; they only become part of the validated dataset once a human approves them in the labeling page.

> 🚨 **HARD RULE — DO NOT SKIP:** You MUST call {{tool:updateAgentLabels}} with verdicts for every approved trace BEFORE running \`startDataset.js\` to open the labeling page. Sending the user into the labeling page without pre-labeled verdicts is a process violation. This is non-negotiable.

> **Made a mistake?** If you realize a verdict was wrong (e.g., you mislabeled a trace or want to re-evaluate), call {{tool:updateAgentLabels}} again with \`{ traceId, archive: true }\` for those traces. The previous label is hidden (kept for audit), and you can re-label the trace from scratch with another \`update_agent_labels\` call.`,
                    next: "phase-3/open-labeling-page",
                },
                {
                    id: "open-labeling-page",
                    kind: "branch",
                    toolCalls: ["bash"],
                    title: "Open the labeling page",
                    body: `**Open the labeling page** — Run the label script to open the labeling page in the browser:

\`\`\`bash
node "{{pluginRoot}}/dist/commands/startDataset.js" <functionKey>
\`\`\`

(\`{{pluginRoot}}\` resolves to the plugin directory; \`<functionKey>\` is the trace function key.) This opens the labeling page showing agent-labeled traces awaiting approval and already-labeled traces. The user approves agent labels, relabels traces, or skips. The script blocks until the user clicks "Confirm dataset"{{#codex}} (up to 30 minutes){{/codex}}.{{#codex}}

The script prints a handoff URL (with \`callbackPort=\` and \`ticket=\` query params). Surface that URL to the user verbatim so they can paste it into a browser if the auto-launch didn't open one.

**Polling (mandatory — see the Blocking-process rule at the top of this skill):** after surfacing the URL, keep polling the live exec session until the process exits. Do NOT stop after printing the URL. Do NOT wait for a chat message from the user — their confirmation arrives as stdout on the already-running process, not as a new prompt. Poll every few seconds until one of these terminal conditions, then route accordingly:

{{whens}}{{/codex}}`,
                    branches: [
                        {
                            when: "user explicitly cancels",
                            description: "stop the flow",
                            next: null,
                        },
                        {
                            when: "process exits (success or failure)",
                            description: "continue to the next step. Exit 0 prints `Labeling complete: ...`; non-zero will surface the error",
                            next: "phase-3/wait-for-labeling",
                        },
                    ],
                },
                {
                    id: "wait-for-labeling",
                    kind: "action",
                    title: "Wait for labeling to complete",
                    body: `**Wait for labeling to complete** — The label script {{^codex}}blocks until the user finishes and clicks "Confirm dataset". It prints a summary when done (e.g., "Labeling complete: 8/10 traces labeled").{{/codex}}{{#codex}}prints a summary on exit, e.g. \`Labeling complete: 8/10 traces labeled (last: pass) [via ticket]\`. This step is complete only when you have observed that final line on the polled exec session. As soon as you see it, move directly to building the dataset without waiting for another user message.{{/codex}}`,
                    next: "phase-3/build-dataset",
                },
                {
                    id: "build-dataset",
                    kind: "action",
                    toolCalls: ["searchTraces", "readTraces"],
                    title: "Build the dataset",
                    body: `**Build the dataset** — Call {{tool:searchTraces}} with \`validated: true\` to get the final set of validated traces. Call {{tool:readTraces}} with all validated trace IDs and \`scope: "full"\` to get the full dataset with labels and annotations.`,
                    next: "phase-3/confirm-dataset",
                },
                {
                    id: "confirm-dataset",
                    kind: "branch",
                    toolCalls: ["searchTraces", "ask"],
                    title: "Confirm the dataset",
                    body: `**Confirm the dataset** — Present the dataset {{presentMode}}: each entry showing (trace ID, label, annotation summary). The dataset must contain at least one **validated failing label** — i.e. at least one trace where a human either authored or approved a \`false\` label. To check, call {{tool:searchTraces}} restricted to the dataset trace IDs with \`validated: true\` and \`labelResult: false\`. Two outcomes:

{{whens}}

Unapproved agent labels do **not** satisfy this gate by design — \`validated: true\` excludes them.`,
                    branches: [
                        {
                            when: "gate fails (no validated failing label — search returns nothing)",
                            description: "tell the user and loop back to find or label more unlabeled traces",
                            next: "phase-3/find-unlabeled",
                        },
                        {
                            when: "gate passes (at least one validated failing label)",
                            description: "get explicit approval, then continue",
                            next: "phase-3/hold-in-context",
                        },
                    ],
                },
                {
                    id: "hold-in-context",
                    kind: "action",
                    title: "Hold in-context",
                    body: `**Hold in-context** — This approved dataset is the benchmark for all experiments in Phase 5. Keep it in your working context throughout.`,
                    next: "phase-4/understand-failures",
                },
            ],
        },
        {
            id: "phase-4",
            title: "Phase 4: Diagnose & Plan",
            stepStyle: "list",
            steps: [
                {
                    id: "understand-failures",
                    kind: "action",
                    toolCalls: ["readTraces"],
                    title: "Understand failures",
                    body: `**Understand failures.** Using the failed traces you read in Phase 3 (or read them now if you haven't):

- Call {{tool:readTraces}} on 3–5 failed traces with \`scope: "full"\`

Synthesize the failure patterns — what's going wrong, what the common threads are.`,
                    next: "phase-4/read-code",
                },
                {
                    id: "read-code",
                    kind: "action",
                    toolCalls: ["read"],
                    title: "Read the code",
                    body: `**Read the code.**

- Find the instrumented function in the codebase (you found it in Phase 2)
- Read the full implementation — follow the call chain to understand the logic
- Identify **iteration targets**: prompts, system messages, parameters, preprocessing, postprocessing
- If BAML files are involved, read the relevant \`.baml\` files`,
                    next: "phase-4/categorize-fixes",
                },
                {
                    id: "categorize-fixes",
                    kind: "action",
                    toolCalls: ["ask"],
                    title: "Categorize fixes based on failure annotations",
                    body: `**Categorize fixes based on failure annotations.** Based on the failure patterns, the code, and the labeled dataset from Phase 3, categorize proposed changes into three buckets:

**Bucket 1 — Code fixes**: Deterministic bugs (off-by-one, type mismatch, missing null check, wrong variable). These won't recur once fixed. Bundle all code fixes into a single experiment unless they are large feature changes. These are applied first as a foundation that all subsequent experiments build on.

**Bucket 2 — Judgment-based fixes**: Prompt changes, context truncation, search tuning, output formatting, etc. These require the user's judgment to evaluate correctness. Each gets its own experiment.

**Bucket 3 — Infrastructure proposals**: Larger changes that require new infrastructure, architectural changes, or significant feature work. These are separated out because experiments become harder to compare when some include large infra changes and others don't — apples-to-apples comparison requires a consistent baseline. Do not run experiments for these. Instead, if the user has integrations (Linear, Notion, Jira), propose creating a task with a clear writeup for future work.

Present the categorized plan {{presentMode}}:

> "Based on the N traces in the dataset, here's what I see:
>
> **Code fixes** (experiment #1 — bundled):
>
> - [Fix]: [What and why, which traces it addresses]
>
> **Judgment-based experiments** (#2, #3, ...):
>
> - [Experiment]: [What change, which traces it targets, hypothesis]
>
> **Future infrastructure** (not experiments):
>
> - [Proposal]: [What it would require, which traces it would help]
>
> I'll replay each experiment against the labeled dataset and evaluate using the annotations as acceptance criteria."

Get the user's confirmation before proceeding.`,
                    next: "phase-5/make-change",
                },
            ],
        },
        {
            id: "phase-5",
            title: "Phase 5: Iterate with Replay",
            intro: `Run an iterative improvement loop.{{#claude}} If experiments are independent, fork them to subagents in parallel using the **Agent tool** with \`isolation: "worktree"\`; otherwise run sequentially.{{/claude}} Each iteration:`,
            stepStyle: "list",
            steps: [
                {
                    id: "make-change",
                    kind: "action",
                    toolCalls: ["ask", "edit", "agent"],
                    title: "Make the change",
                    body: `**Make the change.**

- {{^codex}}Use \`AskUserQuestion\` to explain{{/codex}}{{#codex}}Explain to the user{{/codex}} what you're changing and why, and confirm before editing
- Edit the iteration target (prompt, code, tools, parameters)`,
                    next: "phase-5/replay-against-dataset",
                },
                {
                    id: "replay-against-dataset",
                    kind: "action",
                    toolCalls: ["bash"],
                    title: "Replay against the dataset",
                    body: `**Replay against the dataset.** Collect the trace IDs from the labeled dataset (Phase 3). Run the replay script with those specific traces.

\`\`\`bash
# The exact command depends on the replay script — adapt to what exists
# Example for TypeScript:
cd <project-dir> && npx tsx scripts/replay.ts <pipeline-name> --trace-ids <id1>,<id2>,<id3>,...
\`\`\`

**Before running: verify the replay script prints the full original and new output values to stdout for every item** (not just lengths, counts, hashes, or truncated previews). If it doesn't, fix the script first — the Replay Output Contract and example script live in the SDK reference at \`https://docs.bitfab.ai/<language>-sdk#replay\`. Subagents can't evaluate an improvement from \`5 → 7 (+2)\`.`,
                    next: "phase-5/evaluate-results",
                },
                {
                    id: "evaluate-results",
                    kind: "action",
                    toolCalls: ["read", "write"],
                    title: "Evaluate against labels & annotations",
                    body: `**Evaluate against labels & annotations.** Read the replay output. For each trace in the dataset, use the label (pass/fail) and annotation from Phase 3 to judge whether the new output is an improvement:

- For traces labeled **fail**: Does the new output address the issue described in the annotation? The annotation explains what went wrong — use it as the acceptance criteria.
- For traces labeled **pass**: Did the replay preserve the correct behavior, or did it regress?
- Record the results into a tmp file if the dataset/context is too big so you can recall it later easily.
- Return the results of the sub agent if you are in one to the main agent.`,
                    next: "phase-5/share-results",
                },
                {
                    id: "share-results",
                    kind: "branch",
                    toolCalls: ["ask"],
                    title: "Share results to the user",
                    body: `**Share results to the user.**

> "After N experiments these are the results: X/Y traces now pass.
>
> - ✅ Trace \`abc123\`: Now passes — [how the annotation's issue was resolved]
> - ❌ Trace \`def456\`: Still failing — annotation said [X], output still [Y]
> - ❌⚠️ Trace \`ghi789\`: Was passing, now failing (regression)"

Show this across the full data set, and highlight the best outcome concisely. Explain why it worked best with references to code, docs, and/or research if needed. For the best outcome:

- **If pass rate improved and no regressions**: {{askVerb}} to confirm whether they want to keep iterating or stop
- **If pass rate improved but regressions exist or no improvement**: tell the user and propose to create a plan for new experiments and continue iterating.

Ensure your question includes your recommended next step.

{{branches}}`,
                    branches: [
                        {
                            option: {
                                letter: "A",
                                label: "Keep iterating",
                                description: "run another experiment from the plan",
                            },
                            recommended: true,
                            next: "phase-5/make-change",
                        },
                        {
                            option: {
                                letter: "B",
                                label: "Stop and wrap up",
                                description: "move to the final summary",
                            },
                            next: "phase-6/summary",
                        },
                    ],
                },
            ],
        },
        {
            id: "phase-6",
            title: "Phase 6: Validate & Wrap Up",
            stepStyle: "list",
            steps: [
                {
                    id: "summary",
                    kind: "action",
                    toolCalls: ["ask"],
                    title: "Summary",
                    body: `**Summary.** {{^codex}}Use \`AskUserQuestion\` to present{{/codex}}{{#codex}}Present{{/codex}} the final results similar to this. You may expand where appropriate based on context from the user:

> "**Improvement summary for** \`<traceFunctionKey>\`:
>
> - Failed traces fixed: X/Y (from N% → M% pass rate on labeled failures)
> - Full replay pass rate: A/B
> - Changes made:
>   - [File]: [Description of change]
>   - [File]: [Description of change]
>
> The changes are in your working tree (not committed). Review the diffs and commit when ready."`,
                    next: null,
                },
            ],
        },
    ],
});
