// bitfab-plugin-lib/dist/flows/openLastPlan.js
import { Flow } from "./schema.js";
import { CODEX_RESOLVE_PLUGIN_DIR_BASH } from "./shared.js";
export const openLastPlanFlow = Flow.parse({
    schemaVersion: 1,
    id: "plan",
    title: "Bitfab Plan",
    frontmatter: {
        description: "Open the most recent Bitfab trace plan in the browser",
    },
    tools: {
        bash: { kind: "builtin", name: "Bash" },
    },
    intro: `Open the user's most recent trace plan in the browser. This is the fast path: no codebase scan, no LLM calls, no instrumentation work — just look up the latest plan for the current Bitfab organization and open it in the studio.

Use this when the user wants to revisit, share, or re-confirm a plan they already created with \`{{cmd}}setup\`. If they have not created any trace plans yet, the page will show a small empty state pointing them at \`{{cmd}}setup\`.

A fresh agent session is created so the resulting \`tracePlan:latestOpened\` event lands in the Studio live activity stream for whoever is watching the home page.`,
    entries: {
        all: {
            default: "main/open",
            codex: "main/resolve-plugin-dir",
        },
    },
    defaultMode: "all",
    phases: [
        {
            id: "main",
            title: "Open last trace plan",
            stepStyle: "headers",
            steps: [
                {
                    id: "resolve-plugin-dir",
                    kind: "action",
                    preamble: true,
                    toolCalls: ["bash"],
                    title: "Resolve `BITFAB_PLUGIN_DIR`",
                    body: `Codex does not inject a plugin-root env var. Resolve it first (same block used by \`{{cmd}}setup\` and \`{{cmd}}update\`):

${CODEX_RESOLVE_PLUGIN_DIR_BASH}`,
                    next: "main/open",
                },
                {
                    id: "open",
                    kind: "action",
                    toolCalls: ["bash"],
                    title: "Open the latest trace plan",
                    body: `Run the plugin's view-last-plan helper. It opens \`/trace-plans/latest\` in the browser, which the bitfab-web server resolves to the most recent trace plan for the user's organization and redirects to it. If no plans exist, the page shows an empty state.

\`\`\`bash
node "{{pluginRoot}}/dist/commands/viewLastTracePlan.js"
\`\`\`

The command exits as soon as the browser tab is launched — it does not wait for the user to confirm or cancel. After the command exits, simply acknowledge: "Opened your most recent trace plan." Do not poll or run any other commands.`,
                    next: null,
                },
            ],
        },
    ],
});
