// bitfab-plugin-lib/dist/flows/render.js
import { collectCalledToolIds, computeModesPerStep, computeReachableSteps, MUSTACHE_SECTION_NAMES, resolvePerTarget, resolveTargetableString, } from "./schema.js";
/** Shorthand: resolve a flow string field for the current target. */
function s(value, target, label) {
    return resolveTargetableString(value, target, label);
}
export function renderMarkdown(flow, target) {
    const sections = [];
    const description = s(flow.frontmatter.description, target.name, `flow.frontmatter.description`);
    const argumentHint = flow.frontmatter.argumentHint !== undefined
        ? s(flow.frontmatter.argumentHint, target.name, "frontmatter.argumentHint")
        : undefined;
    sections.push(target.frontmatter({
        flowId: flow.id,
        description,
        argumentHint,
        allowedTools: resolveAllowedTools(flow, target),
    }));
    sections.push(`# ${s(flow.title, target.name, "flow.title")}`);
    if (flow.intro !== undefined) {
        const introStr = s(flow.intro, target.name, "flow.intro");
        if (introStr.length > 0) {
            sections.push(introStr);
        }
    }
    // Flow-wide invariants — rules that apply throughout, rendered once near
    // the top so the agent reads them before working through the phases. Each
    // invariant is a one-line bullet under `## Invariants`. Standard
    // placeholders (`{{tool:…}}`, `{{cmd}}`, Mustache sections) work because
    // the substitution pipeline runs over the joined output below.
    if (flow.invariants.length > 0) {
        const items = flow.invariants.map((inv) => `- ${inv}`).join("\n");
        sections.push(`## Invariants\n\n${items}`);
    }
    const allModes = Object.keys(flow.entries);
    const modesPerStep = computeModesPerStep(flow, target.name);
    const reachable = computeReachableSteps(flow, target.name);
    for (const phase of flow.phases) {
        const rendered = renderPhase(phase, target.name, allModes, modesPerStep, reachable);
        // Skip phases with no reachable steps — pushing an empty string here would
        // create a double `\n\n` separator (triple newline) between adjacent
        // visible sections, which shows up as an extra blank line in the output.
        if (rendered.length > 0) {
            sections.push(rendered);
        }
    }
    if (flow.appendices.length > 0) {
        for (const appendix of flow.appendices) {
            const title = s(appendix.title, target.name, `appendix[${appendix.id}].title`);
            const body = s(appendix.body, target.name, `appendix[${appendix.id}].body`);
            sections.push(`## ${title}\n\n${body}`);
        }
    }
    const joined = sections.join("\n\n");
    // Order matters: strip non-matching Mustache sections FIRST. Otherwise
    // placeholders inside (e.g.) `{{#codex}}{{step:codex-only-step}}{{/codex}}`
    // get resolved against the wrong target's reachability set and throw.
    const afterSections = applyMustacheSections(joined, target.name);
    const withFlowData = applyFlowPlaceholders(afterSections, flow, target);
    const substituted = applyTokens(withFlowData, target, flow);
    return ensureTrailingNewline(substituted);
}
/**
 * Substitute flow-level data placeholders.
 *
 * - `{{step:phase-id/step-id}}` — renders the step's number for the current
 *   target (preamble steps render as "0"). Catches stale step references
 *   when the flow's steps are reordered or split.
 * - `{{tool:id}}` — backtick-wrapped per-target tool name from `flow.tools`
 *   (with the target's `mcpPrefix` for MCP tools)
 * - `{{languages:FIELD}}` — table with the named field (and `name`)
 * - `{{languages:F1,F2,...}}` — table with multiple fields (and `name`)
 * - `{{languages:imports}}` is special-cased as a bullet list because it's
 *   the only field rendered as `displayName: value` rather than a table row.
 *
 * Output still contains target tokens (`{{cmd}}`, `{{editor}}`); those are
 * resolved by `applyTemplate` after this pass.
 */
function applyFlowPlaceholders(text, flow, target) {
    let result = text;
    const stepNumbers = computeStepNumbers(flow, target.name);
    result = result.replace(/\{\{step:([a-z0-9-]+\/[a-z0-9-]+)\}\}/g, (_match, ref) => {
        const num = stepNumbers.get(ref);
        if (num === undefined) {
            throw new Error(`Step reference {{step:${ref}}} not found in flow "${flow.id}" for target "${target.name}"`);
        }
        return String(num);
    });
    result = result.replace(/\{\{tool:([a-zA-Z][a-zA-Z0-9_-]*)\}\}/g, (_match, toolId) => `\`${fullToolName(flow, toolId, target)}\``);
    result = result.replace(/\{\{languages:([a-z,]+)\}\}/g, (_match, columnsRaw) => renderLanguagesBlock(flow, columnsRaw.split(",")));
    return result;
}
/**
 * Map of `phase-id/step-id` → step number for the given target. Mirrors the
 * numbering logic in `renderPhaseAsHeaders` / `renderPhaseAsList`: preamble
 * steps render as 0, everything else 1-indexed among visible steps.
 */
function computeStepNumbers(flow, target) {
    const result = new Map();
    const reachable = computeReachableSteps(flow, target);
    for (const phase of flow.phases) {
        let counter = 1;
        for (const step of phase.steps) {
            if (!reachable.has(`${phase.id}/${step.id}`)) {
                continue;
            }
            const num = step.preamble === true ? 0 : counter;
            if (step.preamble !== true) {
                counter += 1;
            }
            result.set(`${phase.id}/${step.id}`, num);
        }
    }
    return result;
}
/**
 * Auto-derive `allowed-tools` from the union of every step's `toolCalls`,
 * using registry order. When the flow declares no tools the explicit
 * `frontmatter.allowedTools` is used as-is.
 */
function resolveAllowedTools(flow, target) {
    if (Object.keys(flow.tools).length === 0) {
        return flow.frontmatter.allowedTools;
    }
    const calledIds = collectCalledToolIds(flow);
    return calledIds.map((id) => fullToolName(flow, id, target));
}
function fullToolName(flow, toolId, target) {
    const def = flow.tools[toolId];
    if (def === undefined) {
        throw new Error(`Tool "${toolId}" not declared in flow.tools`);
    }
    const name = resolvePerTarget(def.name, target.name);
    if (name === undefined) {
        throw new Error(`Tool "${toolId}" has no name for target "${target.name}" (and no default)`);
    }
    return def.kind === "mcp" ? `${target.mcpPrefix}${name}` : name;
}
function renderLanguagesBlock(flow, columns) {
    if (flow.languages.length === 0) {
        return "";
    }
    // `imports` renders as a bullet list using displayName, the only field
    // styled this way today. Everything else is a table.
    if (columns.length === 1 && columns[0] === "imports") {
        const lines = [];
        for (const lang of flow.languages) {
            if (lang.imports !== undefined) {
                lines.push(`  - ${lang.displayName}: ${lang.imports}`);
            }
        }
        return lines.join("\n");
    }
    const headers = ["Language", ...columns.map((c) => columnHeader(flow, c))];
    const lines = [
        `| ${headers.join(" | ")} |`,
        `|${headers.map(() => "---").join("|")}|`,
    ];
    for (const lang of flow.languages) {
        const cells = columns.map((col) => readLanguageField(lang, col));
        if (cells.every((c) => c !== undefined)) {
            lines.push(`| ${lang.name} | ${cells.join(" | ")} |`);
        }
    }
    return lines.join("\n");
}
function columnHeader(flow, column) {
    return flow.languageColumnHeaders[column] ?? titleCase(column);
}
function titleCase(s) {
    return s.charAt(0).toUpperCase() + s.slice(1);
}
function readLanguageField(lang, column) {
    switch (column) {
        case "imports":
            return lang.imports;
        case "detection":
            return lang.detection;
        case "outdated":
            return lang.outdated;
        case "update":
            return lang.update;
        default:
            throw new Error(`Unknown languages field "${column}"`);
    }
}
function renderPhase(phase, target, allModes, modesPerStep, reachable) {
    const visibleSteps = phase.steps
        .filter((step) => reachable.has(`${phase.id}/${step.id}`))
        .map((step) => ({
        step,
        stepKey: `${phase.id}/${step.id}`,
    }));
    if (visibleSteps.length === 0) {
        return "";
    }
    if (phase.stepStyle === "headers") {
        return renderPhaseAsHeaders(phase, visibleSteps, allModes, modesPerStep, target);
    }
    return renderPhaseAsList(phase, visibleSteps, allModes, modesPerStep, target);
}
function modesForStep(stepKey, modesPerStep, allModes) {
    return modesPerStep.get(stepKey) ?? new Set(allModes);
}
/**
 * Format a positive mode-gate sentence: "Run only when mode is `X`." for one
 * mode, "Run only when mode is `X` or `Y`." for two, "Run only when mode is
 * `X`, `Y` or `Z`." for three or more.
 */
function formatPositiveGate(inModes) {
    const phrases = inModes.map((m) => `\`${m}\``);
    const list = phrases.length === 1
        ? phrases[0]
        : `${phrases.slice(0, -1).join(", ")} or ${phrases[phrases.length - 1]}`;
    return `**Run only when mode is ${list}.**`;
}
/**
 * Build mode-gating prose for a phase, in the positive form: "Run only when
 * mode is `X` [or `Y`]". Two output channels:
 *
 *   - `phaseGate`: a single line emitted just under the phase heading when
 *     every visible step in the phase is reachable in the same proper subset
 *     of modes. One line gates the whole phase; no per-step prose needed.
 *
 *   - `perStepGates`: per-step gates emitted in the step body when the
 *     phase has mixed reachability (different steps reachable in different
 *     mode subsets — e.g. update.ts where step 2 runs in `all`/`plugin` but
 *     step 3 runs in `all`/`sdk`). Each affected step gets its own positive
 *     gate; gates do not cascade or inherit.
 *
 * Steps reachable in every declared mode get no gate. Phases where every
 * visible step is reachable in every mode also get no `phaseGate`.
 *
 * The positive form is unambiguous on scope (it states when the step runs,
 * full stop) and is usually shorter than the negative "Skip if X" form when
 * there are more excluded modes than included modes.
 */
function precomputePhaseAndStepGates(stepKeys, modesPerStep, allModes) {
    if (stepKeys.length === 0) {
        return { phaseGate: null, perStepGates: new Map() };
    }
    const stepModeSets = stepKeys.map((key) => modesForStep(key, modesPerStep, allModes));
    const allModesCount = allModes.length;
    const firstSet = stepModeSets[0];
    const uniform = stepModeSets.every((s) => s.size === firstSet.size && [...s].every((mode) => firstSet.has(mode)));
    if (uniform) {
        if (firstSet.size === allModesCount) {
            // Phase reachable in every mode: no gate at all.
            return { phaseGate: null, perStepGates: new Map() };
        }
        const inModes = allModes.filter((m) => firstSet.has(m));
        return {
            phaseGate: formatPositiveGate(inModes),
            perStepGates: new Map(),
        };
    }
    // Mixed reachability: each affected step gets its own positive gate.
    const perStepGates = new Map();
    for (let i = 0; i < stepKeys.length; i++) {
        const set = stepModeSets[i];
        if (set.size === allModesCount) {
            continue;
        }
        const inModes = allModes.filter((m) => set.has(m));
        perStepGates.set(stepKeys[i], formatPositiveGate(inModes));
    }
    return { phaseGate: null, perStepGates };
}
function bodyWithModeGate(step, stepKey, perStepGates, target) {
    const withBranches = expandBranchesPlaceholders(step, target);
    const gate = perStepGates.get(stepKey);
    if (gate === undefined) {
        return withBranches;
    }
    return `${gate}\n\n${withBranches}`;
}
/**
 * Replace branch-related placeholders in a branch step's body:
 *
 * - `{{branches}}` / `{{branches:prompt-name}}` — formatted AskUserQuestion
 *   option block. Only branches with an `option` field are rendered.
 * - `{{whens}}` — formatted runtime-branch routing block. Only branches
 *   with a `when` field (no `option`) are rendered. Each line shows the
 *   runtime condition and the action (the branch's `description`, falling
 *   back to "stop the flow" for `next: null` or "continue" otherwise).
 *
 * Runtime `when` and user-facing `option` branches are intentionally
 * separate placeholders — the agent presents `{{branches}}` to the user
 * via AskUserQuestion, and reads `{{whens}}` to itself when routing on
 * observed state.
 */
function expandBranchesPlaceholders(step, target) {
    const body = s(step.body, target, `step[${step.id}].body`);
    if (step.kind !== "branch") {
        return body;
    }
    return body
        .replace(/\{\{branches(?::([a-z0-9-]+))?\}\}/g, (_match, promptFilter) => {
        const lines = [];
        for (const branch of step.branches) {
            if (branch.option === undefined) {
                continue;
            }
            if (promptFilter !== undefined && branch.prompt !== promptFilter) {
                continue;
            }
            const { letter, label, description } = branch.option;
            const recommended = branch.recommended === true ? " *(recommended)*" : "";
            const desc = description !== undefined ? ` — ${description}` : "";
            lines.push(`> ${letter}) **${label}**${desc}${recommended}`);
        }
        return lines.join("\n");
    })
        .replace(/\{\{whens\}\}/g, () => {
        const lines = [];
        for (const branch of step.branches) {
            if (branch.when === undefined || branch.option !== undefined) {
                continue;
            }
            const action = branch.description ??
                (branch.next === null ? "stop the flow" : "continue");
            lines.push(`- **${branch.when}** — ${action}`);
        }
        return lines.join("\n");
    });
}
function renderPhaseAsHeaders(phase, visibleSteps, allModes, modesPerStep, target) {
    const blocks = [];
    const stepKeys = visibleSteps.map((v) => v.stepKey);
    const { phaseGate, perStepGates } = precomputePhaseAndStepGates(stepKeys, modesPerStep, allModes);
    if (phaseGate !== null) {
        blocks.push(phaseGate);
    }
    if (phase.intro !== undefined) {
        const intro = s(phase.intro, target, `phase[${phase.id}].intro`);
        if (intro.length > 0) {
            blocks.push(intro);
        }
    }
    let counter = 1;
    for (const { step, stepKey } of visibleSteps) {
        const number = step.preamble === true ? 0 : counter;
        if (step.preamble !== true) {
            counter += 1;
        }
        const title = s(step.title, target, `step[${step.id}].title`);
        const body = bodyWithModeGate(step, stepKey, perStepGates, target);
        blocks.push(`## ${number}. ${title}\n\n${body}`);
    }
    if (phase.outro !== undefined) {
        const outro = s(phase.outro, target, `phase[${phase.id}].outro`);
        if (outro.length > 0) {
            blocks.push(outro);
        }
    }
    return blocks.join("\n\n");
}
function renderPhaseAsList(phase, visibleSteps, allModes, modesPerStep, target) {
    const phaseTitle = s(phase.title, target, `phase[${phase.id}].title`);
    const blocks = [`## ${phaseTitle}`];
    const stepKeys = visibleSteps.map((v) => v.stepKey);
    const { phaseGate, perStepGates } = precomputePhaseAndStepGates(stepKeys, modesPerStep, allModes);
    if (phaseGate !== null) {
        blocks.push(phaseGate);
    }
    if (phase.intro !== undefined) {
        const intro = s(phase.intro, target, `phase[${phase.id}].intro`);
        if (intro.length > 0) {
            blocks.push(intro);
        }
    }
    const items = [];
    let counter = 1;
    for (const { step, stepKey } of visibleSteps) {
        const number = step.preamble === true ? 0 : counter;
        if (step.preamble !== true) {
            counter += 1;
        }
        const body = bodyWithModeGate(step, stepKey, perStepGates, target);
        items.push(`${number}. ${indentBody(body)}`);
    }
    blocks.push(items.join("\n"));
    if (phase.outro !== undefined) {
        const outro = s(phase.outro, target, `phase[${phase.id}].outro`);
        if (outro.length > 0) {
            blocks.push(outro);
        }
    }
    return blocks.join("\n\n");
}
function indentBody(body) {
    const lines = body.split("\n");
    return lines
        .map((line, i) => (i === 0 ? line : line.length > 0 ? `   ${line}` : line))
        .join("\n");
}
/**
 * Strip Mustache conditional sections (`{{#target}}...{{/target}}` and the
 * inverse `{{^target}}...{{/target}}`) leaving only content that matches the
 * current target. Exported so the dev-server payload builder can apply the
 * same stripping to step / appendix bodies before they reach the UI; without
 * it the StepDetail "Body" pane would still show `{{#codex}}…{{/codex}}`
 * blocks even when target=claude.
 *
 * In `renderMarkdown` this runs BEFORE flow-data placeholder resolution so
 * that placeholders inside non-matching sections (which would throw if
 * evaluated against the wrong target's reachability set) never get evaluated.
 */
export function applyMustacheSections(text, targetName) {
    const knownSectionNames = new Set(MUSTACHE_SECTION_NAMES);
    const sectionRegex = /\{\{([#^])([a-z]+)\}\}([\s\S]*?)\{\{\/\2\}\}/g;
    return text.replace(sectionRegex, (_match, kind, name, inner) => {
        if (!knownSectionNames.has(name)) {
            throw new Error(`Unknown target name "${name}" in Mustache section. Valid names: ${[...knownSectionNames].join(", ")}.`);
        }
        const isMatch = name === targetName;
        const include = kind === "#" ? isMatch : !isMatch;
        return include ? inner : "";
    });
}
/**
 * Resolve target tokens (cmd, editor, etc.) and flow variables, then sanity-
 * check that no `{{...}}` placeholder slipped through. Runs AFTER Mustache
 * sections and flow-data placeholders.
 */
function applyTokens(text, target, flow) {
    // Build a unified substitution pool: target tokens + flow variables.
    // Variables that collide with target tokens are caught at parse time by
    // the schema's superRefine, so this loop never throws in practice.
    const pool = { ...target.tokens };
    for (const [name, value] of Object.entries(flow.variables)) {
        pool[name] = s(value, target.name, `flow.variables.${name}`);
    }
    const tokenRegex = /\{\{([a-zA-Z][a-zA-Z0-9_]*)\}\}/g;
    const result = text.replace(tokenRegex, (_match, key) => {
        if (!Object.hasOwn(pool, key)) {
            throw new Error(`Unresolved placeholder {{${key}}} — not a target token or flow variable. Define it in flow.variables, fix the typo, or remove the placeholder.`);
        }
        return pool[key];
    });
    // Final sanity scan — any leftover `{{...}}` is a malformed or unknown
    // namespaced placeholder that no earlier pass recognized.
    const leftovers = result.match(/\{\{[^}]+\}\}/g);
    if (leftovers) {
        throw new Error(`Unresolved placeholder(s): ${leftovers.join(", ")}. Check spelling and namespace (step:, tool:, languages:, branches, branches:).`);
    }
    return result;
}
function ensureTrailingNewline(text) {
    return text.endsWith("\n") ? text : `${text}\n`;
}
