// bitfab-plugin-lib/dist/flows/schema.js
import { z } from "zod";
const StepIdRef = z.string().regex(/^[a-z0-9-]+\/[a-z0-9-]+$/, {
    message: "Step references must use the form 'phase-id/step-id'",
});
const StepLocalId = z.string().regex(/^[a-z0-9-]+$/);
const PhaseId = z.string().regex(/^[a-z0-9-]+$/);
/**
 * Generic shape for per-target overrides. Any string field can be either a
 * bare value (applies to all targets) or an object with target-specific
 * overrides. Per-target objects must include `default` OR define all three
 * targets explicitly — half-specified shapes are rejected at parse time.
 *
 *   body: "shared text"                                    // applies to all
 *   body: { default: "shared", codex: "codex variant" }    // codex overrides
 *   body: { claude: "...", cursor: "...", codex: "..." }  // exhaustive
 *
 * `resolvePerTarget(value, target)` reads the value for a given target.
 */
function PerTarget(inner) {
    const objectForm = z
        .object({
        default: inner.optional(),
        claude: inner.optional(),
        cursor: inner.optional(),
        codex: inner.optional(),
    })
        .refine((v) => v.default !== undefined ||
        (v.claude !== undefined &&
            v.cursor !== undefined &&
            v.codex !== undefined), {
        message: "Per-target object must have `default` or define all of claude/cursor/codex.",
    });
    return z.union([inner, objectForm]);
}
export function resolvePerTarget(value, target) {
    if (typeof value !== "object" || value === null) {
        return value;
    }
    const map = value;
    return map[target] ?? map.default;
}
/** Shorthand for `string | PerTarget<string>` — used everywhere strings appear. */
const TargetableString = PerTarget(z.string());
const TargetableNonEmptyString = PerTarget(z.string().min(1));
/**
 * Token names that targets always provide via `Target.tokens`. Flow variables
 * cannot reuse these names (parse-time check on `flow.variables`).
 */
export const RESERVED_TOKEN_NAMES = [
    "cmd",
    "editor",
    "pluginRoot",
    "askUser",
    "askVerb",
    "askVerbCap",
    "askNoun",
];
/** Names that may appear as the body of a Mustache section, e.g. `{{#codex}}…{{/codex}}`. */
export const MUSTACHE_SECTION_NAMES = ["claude", "cursor", "codex"];
/**
 * Resolve a `Targetable<string>` into a plain string for the given target,
 * throwing when the value is a per-target object with no entry for the
 * target and no `default`.
 */
export function resolveTargetableString(value, target, label) {
    const resolved = resolvePerTarget(value, target);
    if (resolved === undefined) {
        throw new Error(`Per-target value missing for ${target}: ${label}`);
    }
    return resolved;
}
const NextRef = StepIdRef.nullable();
/**
 * Mode-aware dispatch. Each declared mode must have an entry; missing modes
 * are caught by `superRefine` so the schema enforces exhaustive routing.
 *
 * Use the linear `NextRef` form (a step id string or null) when every mode
 * goes to the same place — that's the common case and keeps the data clean.
 * Switch to `byMode` only at the points where the modes actually diverge.
 */
const ByModeNext = z.object({
    byMode: z.record(z.string(), NextRef),
});
const Next = z.union([NextRef, ByModeNext]);
const Base = z.object({
    id: StepLocalId,
    title: TargetableNonEmptyString,
    body: TargetableString,
    /**
     * Logical ids of tools this step calls (must exist in `flow.tools`). Used
     * to auto-derive `frontmatter.allowedTools` and to render `{{tool:id}}`
     * placeholders in the body.
     */
    toolCalls: z.array(z.string()).optional(),
    /**
     * If true (in a "headers" phase), the step is numbered as "0" instead of
     * 1-indexed among non-preamble visible steps. Used for setup-style preamble
     * sections like Codex's "Resolve BITFAB_PLUGIN_DIR" that conceptually sit
     * before step 1 even though they execute first.
     */
    preamble: z.boolean().optional(),
    /**
     * Optional list of appendix ids this step consults. Available on every step
     * kind so that an action or branch step can document its references to
     * appendices (validated against `flow.appendices` at parse time). The
     * `kind: "reference"` discriminant carries a stricter `min(1)` requirement
     * since consulting an appendix is its primary purpose.
     */
    refs: z.array(z.string().min(1)).optional(),
    meta: z.record(z.string(), z.unknown()).optional(),
});
const Action = Base.extend({
    kind: z.literal("action"),
    next: Next,
});
const BranchOption = z.object({
    /** Letter shown in the AskUserQuestion prompt, e.g. "A", "B", "C". */
    letter: z.string().min(1),
    /** Bolded option name, e.g. "Update all N outdated workspaces". */
    label: z.string().min(1),
    /** Optional explanation that follows the label after an em dash. */
    description: z.string().optional(),
});
const Branch = Base.extend({
    kind: z.literal("branch"),
    branches: z
        .array(z
        .object({
        /**
         * Edge label in the schema diagram. Required for runtime branches
         * (exit conditions, not user choices). Optional for user-choice
         * branches — when omitted, the diagram label defaults to
         * `${option.letter} — ${option.label}`.
         */
        when: z.string().optional(),
        /**
         * Where this branch goes after the user picks it (or after the
         * runtime condition resolves). Linear `NextRef` form for the common
         * case where every mode goes to the same place; `byMode` dispatch
         * for branches that need per-mode targets — e.g. an "exit" option
         * that hands off to the next phase in `all` mode but terminates in
         * sub-modes.
         */
        next: Next,
        /** Marks the option the agent should default to. */
        recommended: z.boolean().optional(),
        /**
         * Tags the prompt this option belongs to when a single step exposes
         * multiple AskUserQuestion forks. Lets the renderer pick a subset
         * with `{{branches:prompt-name}}`.
         */
        prompt: z.string().optional(),
        /**
         * Present on user-choice branches — drives the rendered prompt
         * block via `{{branches}}` and supersedes `when` as the source of
         * truth for option text. Absent on runtime branches.
         */
        option: BranchOption.optional(),
        /**
         * Optional action description for runtime (`when`) branches — the
         * "what to do" text that appears after the em-dash in the
         * `{{whens}}` rendering. When omitted, falls back to "stop the
         * flow" (for `next: null`) or "continue" (for any step ref). Has
         * no effect on user-choice (`option`) branches; those use
         * `option.description`.
         */
        description: z.string().optional(),
    })
        .refine((b) => b.when !== undefined || b.option !== undefined, {
        message: "branch must define `when` or `option`",
    }))
        .min(2),
});
const Loop = Base.extend({
    kind: z.literal("loop"),
    backTo: StepIdRef,
    exitWhen: z.string().min(1),
    onExit: StepIdRef,
});
const Reference = Base.extend({
    kind: z.literal("reference"),
    refs: z.array(z.string().min(1)).min(1),
    next: Next,
});
const Parallel = Base.extend({
    kind: z.literal("parallel"),
    branches: z
        .array(z.object({
        label: z.string().min(1),
        body: z.string(),
    }))
        .min(2),
    next: Next,
});
const Step = z.discriminatedUnion("kind", [
    Action,
    Branch,
    Loop,
    Reference,
    Parallel,
]);
const Phase = z.object({
    id: PhaseId,
    title: TargetableNonEmptyString,
    intro: TargetableString.optional(),
    steps: z.array(Step).min(1),
    outro: TargetableString.optional(),
    /**
     * How steps inside this phase are rendered.
     * - "list": phase renders as `## {title}`, intro paragraph, then steps as a `1. ... 2. ...` numbered markdown list.
     * - "headers": phase title is hidden; each step renders as its own `## {N}. {title}` H2 section with body below.
     *
     * In "headers" mode, steps marked `preamble: true` are numbered "0" instead of joining the 1-indexed sequence.
     */
    stepStyle: z.enum(["list", "headers"]).default("list"),
});
const Appendix = z.object({
    id: z.string().min(1),
    title: TargetableNonEmptyString,
    body: TargetableString,
});
const Frontmatter = z.object({
    description: TargetableNonEmptyString,
    argumentHint: TargetableString.optional(),
    /**
     * Optional manual override. When omitted and `flow.tools` is non-empty,
     * the renderer auto-derives the list from the union of step `toolCalls`.
     */
    allowedTools: z.array(z.string()).optional(),
});
/**
 * One tool the flow can call. Steps reference tools by their logical id
 * (the registry key) via `step.toolCalls` and `{{tool:id}}` placeholders.
 *
 * Built-in tools render as their `name` directly. MCP tools have the
 * target's `mcpPrefix` prepended at render time, so the same tool spec
 * produces e.g. `mcp__plugin_bitfab_Bitfab__get_bitfab_api_key` for
 * Claude and `mcp__Bitfab__get_bitfab_api_key` for Cursor / Codex.
 */
const ToolDef = z.object({
    name: PerTarget(z.string()),
    kind: z.enum(["builtin", "mcp"]),
});
/**
 * One language the flow operates on (typescript, python, ruby, go, ...).
 *
 * Field names map directly to `{{languages:FIELD}}` placeholder names — no
 * suffix-stripping or other implicit conventions. Multi-column tables use
 * comma-separated placeholders, e.g. `{{languages:detection,outdated}}`.
 */
const Language = z.object({
    /** Lowercase id used in tables (typescript, python, ruby, go). */
    name: z.string().min(1),
    /** Capitalized name used in bullet lists (TypeScript, Python, Ruby, Go). */
    displayName: z.string().min(1),
    /** Markdown snippet used by `{{languages:imports}}`. */
    imports: z.string().optional(),
    /** Markdown snippet used by `{{languages:detection}}`. */
    detection: z.string().optional(),
    /** Markdown snippet used by `{{languages:outdated}}`. */
    outdated: z.string().optional(),
    /** Markdown snippet used by `{{languages:update}}`. */
    update: z.string().optional(),
});
export const Flow = z
    .object({
    schemaVersion: z.literal(1),
    id: z.string().min(1),
    title: TargetableNonEmptyString,
    frontmatter: Frontmatter,
    intro: TargetableString.optional(),
    /**
     * Entry step per invocation mode. Each entry is either a bare step id
     * (every target enters at the same step) or a per-target object so a
     * target can start at a different step than the others — for example,
     * Codex enters at the `resolve-plugin-dir` preamble while Claude and
     * Cursor enter directly at `run-update-script`.
     */
    entries: z.record(z.string(), PerTarget(StepIdRef)),
    /** Which mode is invoked when the user passes no argument. Must be a key in `entries`. */
    defaultMode: z.string().optional(),
    phases: z.array(Phase).min(1),
    appendices: z.array(Appendix).default([]),
    invariants: z.array(z.string()).default([]),
    languages: z.array(Language).default([]),
    /**
     * Headers for the columns rendered by `{{languages:FIELD}}` placeholders.
     * Keyed by Language field name (`imports`, `detection`, `outdated`,
     * `update`). When unset for a column, the renderer falls back to the
     * field's title-cased name.
     */
    languageColumnHeaders: z.record(z.string(), z.string()).default({}),
    /**
     * Tool registry keyed by logical id. `step.toolCalls` references entries
     * here; `{{tool:id}}` placeholders render the per-target name.
     */
    tools: z.record(z.string(), ToolDef).default({}),
    /**
     * User-defined variables. Each value is either a bare string (applies to
     * all targets) or a per-target map. Referenced via `{{name}}` placeholders
     * in any string field. Must not collide with target-level tokens (cmd,
     * editor, pluginRoot, askUser).
     */
    variables: z.record(z.string(), TargetableString).default({}),
})
    .superRefine((flow, ctx) => {
    const allStepIds = new Set();
    const phaseIds = new Set();
    for (const phase of flow.phases) {
        if (phaseIds.has(phase.id)) {
            ctx.addIssue({
                code: "custom",
                path: ["phases", phase.id],
                message: `duplicate phase id "${phase.id}"`,
            });
        }
        phaseIds.add(phase.id);
        const localIds = new Set();
        for (const step of phase.steps) {
            if (localIds.has(step.id)) {
                ctx.addIssue({
                    code: "custom",
                    path: ["phases", phase.id, "steps"],
                    message: `duplicate step id "${step.id}" in phase "${phase.id}"`,
                });
            }
            localIds.add(step.id);
            allStepIds.add(`${phase.id}/${step.id}`);
        }
    }
    const declaredModes = Object.keys(flow.entries);
    const declaredModeSet = new Set(declaredModes);
    if (flow.defaultMode !== undefined &&
        !declaredModeSet.has(flow.defaultMode)) {
        ctx.addIssue({
            code: "custom",
            path: ["defaultMode"],
            message: `defaultMode "${flow.defaultMode}" not declared in entries`,
        });
    }
    const refsFromNext = (next) => {
        if (next === null) {
            return [];
        }
        if (typeof next === "string") {
            return [next];
        }
        return Object.values(next.byMode).filter((v) => typeof v === "string");
    };
    const refsFromStep = (s) => {
        switch (s.kind) {
            case "action":
            case "reference":
            case "parallel":
                return refsFromNext(s.next);
            case "branch":
                return s.branches.flatMap((b) => refsFromNext(b.next));
            case "loop":
                return [s.backTo, s.onExit];
        }
    };
    for (const phase of flow.phases) {
        for (const step of phase.steps) {
            for (const ref of refsFromStep(step)) {
                if (!allStepIds.has(ref)) {
                    ctx.addIssue({
                        code: "custom",
                        path: ["phases", phase.id, "steps", step.id],
                        message: `edge → unknown step "${ref}"`,
                    });
                }
            }
            const validateByMode = (byMode, path, label) => {
                for (const mode of Object.keys(byMode)) {
                    if (!declaredModeSet.has(mode)) {
                        ctx.addIssue({
                            code: "custom",
                            path,
                            message: `${label} lists mode "${mode}" not declared in flow.entries`,
                        });
                    }
                }
                const missingModes = declaredModes.filter((m) => !(m in byMode));
                if (missingModes.length > 0) {
                    ctx.addIssue({
                        code: "custom",
                        path,
                        message: `${label} byMode is missing entries for mode(s): ${missingModes.join(", ")}`,
                    });
                }
            };
            if ((step.kind === "action" ||
                step.kind === "reference" ||
                step.kind === "parallel") &&
                typeof step.next === "object" &&
                step.next !== null) {
                validateByMode(step.next.byMode, ["phases", phase.id, "steps", step.id, "next", "byMode"], `step "${step.id}" next`);
            }
            if (step.kind === "branch") {
                for (const [i, branch] of step.branches.entries()) {
                    if (typeof branch.next === "object" && branch.next !== null) {
                        const branchLabelText = branch.option !== undefined
                            ? `option "${branch.option.letter}"`
                            : (branch.when ?? `branch[${i}]`);
                        validateByMode(branch.next.byMode, [
                            "phases",
                            phase.id,
                            "steps",
                            step.id,
                            "branches",
                            i,
                            "next",
                            "byMode",
                        ], `step "${step.id}" branch ${branchLabelText}`);
                    }
                }
            }
        }
    }
    // At most one preamble step per phase — multiple would all render as
    // `## 0.` and produce duplicate headers in the markdown.
    for (const phase of flow.phases) {
        const preambleCount = phase.steps.filter((s) => s.preamble === true).length;
        if (preambleCount > 1) {
            ctx.addIssue({
                code: "custom",
                path: ["phases", phase.id, "steps"],
                message: `phase "${phase.id}" has ${preambleCount} preamble steps; only one is allowed per phase`,
            });
        }
    }
    const appendixIds = new Set(flow.appendices.map((a) => a.id));
    for (const phase of flow.phases) {
        for (const step of phase.steps) {
            const refs = step.refs ?? [];
            for (const r of refs) {
                if (!appendixIds.has(r)) {
                    ctx.addIssue({
                        code: "custom",
                        path: ["phases", phase.id, "steps", step.id, "refs"],
                        message: `unknown appendix "${r}"`,
                    });
                }
            }
        }
    }
    const targetNames = ["claude", "cursor", "codex"];
    for (const [mode, entry] of Object.entries(flow.entries)) {
        for (const target of targetNames) {
            const resolved = resolvePerTarget(entry, target);
            if (resolved === undefined) {
                ctx.addIssue({
                    code: "custom",
                    path: ["entries", mode],
                    message: `entry for mode "${mode}" has no value for target "${target}" (and no default)`,
                });
                continue;
            }
            if (!allStepIds.has(resolved)) {
                ctx.addIssue({
                    code: "custom",
                    path: ["entries", mode],
                    message: `entry "${mode}" → "${target}" → unknown step "${resolved}"`,
                });
            }
        }
    }
    // Per-mode-per-target reachability: each (mode, target) walks from its
    // resolved entry, following the mode-specific next pointer where one
    // exists. A step is OK if reachable in at least one (mode, target).
    const reachableUnion = new Set();
    for (const [mode, entry] of Object.entries(flow.entries)) {
        for (const target of targetNames) {
            const start = resolvePerTarget(entry, target);
            if (start === undefined) {
                continue;
            }
            const queue = [start];
            const seen = new Set();
            while (queue.length > 0) {
                const id = queue.shift();
                if (id === undefined || seen.has(id)) {
                    continue;
                }
                seen.add(id);
                reachableUnion.add(id);
                const [phaseId, stepId] = id.split("/");
                const step = flow.phases
                    .find((p) => p.id === phaseId)
                    ?.steps.find((s) => s.id === stepId);
                if (step === undefined) {
                    continue;
                }
                for (const n of nextRefsForMode(step, mode)) {
                    if (n !== null) {
                        queue.push(n);
                    }
                }
            }
        }
    }
    for (const id of allStepIds) {
        if (reachableUnion.has(id)) {
            continue;
        }
        ctx.addIssue({
            code: "custom",
            path: ["phases"],
            message: `unreachable step "${id}" in any (mode, target) combination`,
        });
    }
    const toolIds = new Set(Object.keys(flow.tools));
    for (const phase of flow.phases) {
        for (const step of phase.steps) {
            if (step.toolCalls === undefined) {
                continue;
            }
            for (const toolId of step.toolCalls) {
                if (!toolIds.has(toolId)) {
                    ctx.addIssue({
                        code: "custom",
                        path: ["phases", phase.id, "steps", step.id, "toolCalls"],
                        message: `step "${step.id}" calls tool "${toolId}" not declared in flow.tools`,
                    });
                }
            }
        }
    }
    const reservedNames = new Set(RESERVED_TOKEN_NAMES);
    for (const name of Object.keys(flow.variables)) {
        if (reservedNames.has(name)) {
            ctx.addIssue({
                code: "custom",
                path: ["variables", name],
                message: `variable "${name}" collides with the target token of the same name`,
            });
        }
    }
});
/**
 * Compute the set of tool logical-ids called by any step in the flow,
 * preserving registry order. Used by the renderer to auto-derive the
 * Claude `allowed-tools` frontmatter list.
 */
export function collectCalledToolIds(flow) {
    const called = new Set();
    for (const phase of flow.phases) {
        for (const step of phase.steps) {
            for (const id of step.toolCalls ?? []) {
                called.add(id);
            }
        }
    }
    return Object.keys(flow.tools).filter((id) => called.has(id));
}
/**
 * Resolve the next step(s) reachable from `step` when invoked in `mode`.
 * Returns an array because branch/loop kinds have multiple outgoing edges.
 *
 * For action/reference/parallel with `byMode` next, returns the per-mode
 * target. With a linear next, returns it for every mode. Branch and loop
 * kinds aren't mode-discriminating so they return all outgoing edges.
 */
export function nextRefsForMode(step, mode) {
    switch (step.kind) {
        case "action":
        case "reference":
        case "parallel":
            if (step.next === null || typeof step.next === "string") {
                return [step.next];
            }
            // byMode: missing entry = step is a dead end in that mode
            return [step.next.byMode[mode] ?? null];
        case "branch":
            return step.branches.map((b) => {
                if (b.next === null || typeof b.next === "string") {
                    return b.next;
                }
                return b.next.byMode[mode] ?? null;
            });
        case "loop":
            return [step.backTo, step.onExit];
    }
}
/**
 * Set of step ids reachable in any mode for the given target. The renderer
 * uses this as the visibility filter — a step renders iff it's reachable in
 * the current target. Replaces the hand-maintained `onlyTargets` flag.
 */
export function computeReachableSteps(flow, target) {
    return new Set(computeModesPerStep(flow, target).keys());
}
/**
 * For each step in the flow, the set of modes it is reachable in for a given
 * target. Used by the renderer to emit "skip this step if mode is X" prose
 * without the author having to hand-maintain a `modes` field.
 */
export function computeModesPerStep(flow, target) {
    const result = new Map();
    const stepById = new Map();
    for (const phase of flow.phases) {
        for (const step of phase.steps) {
            stepById.set(`${phase.id}/${step.id}`, step);
        }
    }
    for (const [mode, entry] of Object.entries(flow.entries)) {
        const start = resolvePerTarget(entry, target);
        if (start === undefined) {
            continue;
        }
        const queue = [start];
        const seen = new Set();
        while (queue.length > 0) {
            const id = queue.shift();
            if (id === undefined || seen.has(id)) {
                continue;
            }
            seen.add(id);
            const set = result.get(id) ?? new Set();
            set.add(mode);
            result.set(id, set);
            const step = stepById.get(id);
            if (step === undefined) {
                continue;
            }
            for (const n of nextRefsForMode(step, mode)) {
                if (n !== null) {
                    queue.push(n);
                }
            }
        }
    }
    return result;
}
/**
 * Diagram edge label for a branch entry. Derived from `option` when present,
 * otherwise reads `when` directly.
 */
export function branchLabel(branch) {
    if (branch.option !== undefined) {
        return `${branch.option.letter} — ${branch.option.label}`;
    }
    if (branch.when !== undefined) {
        return branch.when;
    }
    return "(unlabeled branch)";
}
