// bitfab-plugin-lib/dist/flows/setup.js
import { Flow } from "./schema.js";
import { SHARED_VARIABLES } from "./sharedVariables.js";
export const setupFlow = Flow.parse({
    schemaVersion: 1,
    id: "setup",
    title: "Bitfab Setup",
    frontmatter: {
        description: "Set up and update the loop to improve AI features autonomously",
        argumentHint: "[<mode>] [<what to do>]",
    },
    tools: {
        bash: { kind: "builtin", name: "Bash" },
        read: { kind: "builtin", name: "Read" },
        glob: { kind: "builtin", name: "Glob" },
        grep: { kind: "builtin", name: "Grep" },
        edit: { kind: "builtin", name: "Edit" },
        write: { kind: "builtin", name: "Write" },
        webFetch: { kind: "builtin", name: "WebFetch" },
        ask: { kind: "builtin", name: "AskUserQuestion" },
        apiKey: { kind: "mcp", name: "get_bitfab_api_key" },
        createTracePlan: { kind: "mcp", name: "create_trace_plan" },
        getTracePlan: { kind: "mcp", name: "get_trace_plan" },
        listFunctions: { kind: "mcp", name: "list_trace_functions" },
        searchTraces: { kind: "mcp", name: "search_traces" },
        getTemplateReference: { kind: "mcp", name: "get_template_reference" },
        getTemplate: { kind: "mcp", name: "get_template" },
        updateTemplate: { kind: "mcp", name: "update_template" },
    },
    intro: `{{^codex}}**Always use \`AskUserQuestion\` when asking questions or presenting choices.** Never print a question as text and wait.{{/codex}}{{#codex}}**When the flow branches, always present the options clearly and wait for the user's answer before proceeding.** Number or letter the options so the user can pick by reference.{{/codex}} Rules:
- Recommend an option first, explain why in one line
- Present 2-5 concrete options
- One decision per question — never batch
{{#codex}}
**🚨 Blocking-process rule (applies to any plugin command described as "blocks until the user does X"):** When you launch a plugin CLI that blocks on a browser handoff (\`login.js\`, \`startDataset.js\`, \`openTracePlan.js\`, etc.), you MUST keep the exec session alive and keep polling it until the process exits on its own.

- Printing the URL is NOT completion. The process is still running, racing a loopback callback and a server-polled ticket — it exits only after the user's browser action delivers a result (or after the timeout).
- After sending the URL to the user, keep polling the live shell/exec session every few seconds with your normal "read more output" tool. Do not idle waiting for a user message.
- The user's confirmation does NOT come back as a chat message; it comes back as the plugin process exiting with output on stdout.
- Stop polling only when: (a) the process exits 0 with its completion summary, (b) the process exits non-zero, or (c) the user explicitly cancels.
- When the process exits, immediately continue with the next step — don't wait for another user message.
{{/codex}}
This skill has six phases: **login**, **instrument**, **modify**, **view**, **replay**, and **templates**. Run individually or all at once (\`all\` runs login → instrument → replay; \`modify\` is only invoked explicitly or as a branch from the Instrument step {{step:instrument/search-existing}} menu; \`view\` is only invoked explicitly; \`templates\` is only invoked explicitly).

Within an Instrument cycle, **instrumentation and the replay pipeline for the cycle's trace function are written {{^codex}}in parallel{{/codex}}{{#codex}}in the same batch of tool calls{{/codex}}** once the trace plan is confirmed (see step {{step:instrument/write-instrumentation}}). The Replay phase in \`all\` mode is therefore a coverage-verification/backfill sweep — it typically finds every key already wired up.
{{#codex}}
**Before running plugin commands below**, resolve \`BITFAB_PLUGIN_DIR\` in the shell. Codex does not inject a plugin-root env var, so you must determine it. Copy and run this block verbatim — it auto-detects whichever install is active (dev, prod, or a custom \`CODEX_HOME\`):

\`\`\`bash
BITFAB_PLUGIN_DIR=$(
  hit=$(find "\${CODEX_HOME:-$HOME/.codex}/plugins/cache" -maxdepth 6 -type f -name status.js \\
    \\( -path '*/bitfab-internal/bitfab/local/dist/commands/*' \\
    -o -path '*/bitfab/bitfab/*/dist/commands/*' \\) 2>/dev/null | head -1)
  echo "\${hit%/dist/commands/status.js}"
)
export BITFAB_PLUGIN_DIR
test -n "$BITFAB_PLUGIN_DIR" && echo "Found: $BITFAB_PLUGIN_DIR" || echo "ERROR: Bitfab plugin not installed"
\`\`\`

If the block prints \`ERROR: Bitfab plugin not installed\`, the user hasn't installed the plugin yet — stop and tell them to add the marketplace. Otherwise proceed. The cache path includes a trailing \`local\` for dev installs and \`<version>\` for prod installs; the commands live at \`$BITFAB_PLUGIN_DIR/dist/commands/*.js\`.
{{/codex}}
**SDK reference:** https://docs.bitfab.ai is the source of truth for SDK install, initialization, API surface, and replay. Fetch in this order before writing any code — do not improvise from memory:
- **Canonical API surface (preferred for agents):** the dense reference pages at \`/reference/typescript\`, \`/reference/python\`, \`/reference/ruby\`, \`/reference/go\`. These list every public export, signature, type, default, and error semantic — no tutorials, no prose. Read these first.
- **Cross-SDK shared semantics:** \`/reference/overview\` (invariants), \`/reference/span-types\` (the \`SpanType\` enum), \`/reference/http\` (wire protocol).
- **Framework integrations (fetch when a framework is detected in step 1 of Instrument):** \`/frameworks/langgraph\`, \`/frameworks/openai-agents\`, \`/frameworks/claude-agent-sdk\`, \`/frameworks/baml\`. Each page documents the SDK's native handler/processor/wrapper for that framework, which is usually preferable to hand-wrapping every node/agent call with \`withSpan\`/\`@span\`.
- **Tutorials / walkthroughs / replay script template:** the language-specific guide pages (\`/typescript-sdk\`, \`/python-sdk\`, \`/ruby-sdk\`, \`/go-sdk\`). Use these for the copy-pasteable replay script and the replay output contract. During Instrument, fetch the \`#replay\` section before step {{step:instrument/write-instrumentation}} so the replay script can be written {{parallelism}} instrumentation.

**MCP tools:** This skill uses \`get_bitfab_api_key\`, \`create_trace_plan\`, and \`get_trace_plan\` (login / instrument / modify), and — for the \`templates\` mode only — \`list_trace_functions\`, \`search_traces\`, \`get_template_reference\`, \`get_template\`, and \`update_template\`. All come from the **local plugin MCP server** (bundled with this plugin){{#claude}}. Do NOT use the remote Bitfab MCP tools (\`mcp__Simforge__*\` or \`mcp__Bitfab__*\`) — use only the \`mcp__plugin_bitfab_Bitfab__*\` variants{{/claude}}{{^claude}}, exposed under the \`mcp__Bitfab__*\` prefix{{/claude}}.

| Invocation | Action |
|---|---|
| \`{{cmd}}setup\` or \`{{cmd}}setup all\` | Run login, then instrument + replay ({{togetherness}} per workflow) |
| \`{{cmd}}setup login\` | Authenticate via browser OAuth and retrieve API key |
| \`{{cmd}}setup login headless\` | Authenticate by pasting a token (no browser callback needed) |
| \`{{cmd}}setup instrument\` | Instrument AI workflows with Bitfab tracing |
| \`{{cmd}}setup modify\` | Modify an existing trace setup (add context, change depth, or move the root) |
| \`{{cmd}}setup view\` | Open the trace planner UI for an existing trace function (read-only) |
| \`{{cmd}}setup replay\` | Create or update replay scripts for instrumented workflows |
| \`{{cmd}}setup templates [<key>]\` | Iterate on the span-rendering templates for one trace function |`,
    entries: {
        all: "preamble/render-intro",
        login: "login/status-check",
        "login-headless": "login-headless/determine-service-url",
        instrument: "instrument/detect-language",
        modify: "modify/gather-existing",
        view: "view/gather-existing",
        replay: "replay/gather-keys",
        templates: "templates/pick-function",
    },
    defaultMode: "all",
    variables: {
        ...SHARED_VARIABLES,
        authTarget: {
            claude: "claude",
            cursor: "cursor",
            codex: "codex",
        },
        /**
         * Tool-call syntax for spawning the 11b replay-pipeline subagent. Claude
         * Code's Task tool is invoked as `Agent(subagent_type="...")`; Cursor has
         * subagent capability but uses different invocation, so we describe it
         * generically and let the agent pick its own dispatch syntax.
         */
        subagentInvocation: {
            claude: 'one `Agent(subagent_type="general-purpose")` call',
            default: "a parallel subagent dispatch",
        },
    },
    appendices: [
        {
            id: "refactor-confirmation",
            title: "Refactor confirmation (applies to Instrument step {{step:instrument/present-workflows}} and Replay step {{step:replay/safety-net}})",
            body: `Whenever the user picks "refactor to extract a pure core" (or any option that modifies existing functions/call sites, not just adds new wrappers), you must:

1. **Build a refactor plan** listing:
   - **Flavor** — **visibility** (extract + export, logic unchanged) or **structural** (new pure-core fn with serializable inputs, may require callers to construct them). Most cases are visibility.
   - **Source** — the function(s) that will be modified, with file path and current signature
   - **Extraction** — the new function name, its signature, and (for visibility refactors) an explicit note that the logic moves unchanged
   - **Trace wrap** — which function will carry the \`getFunction(...)\` / SDK trace wrap after the refactor
   - **Call sites** — every caller that will be rewritten, with file path and line range

2. **Present the plan verbatim** to the user, in the same format above.

3. **{{^codex}}AskUserQuestion{{/codex}}{{#codex}}Ask the user{{/codex}}** with exactly two options:
   - **"Apply refactor"** — proceed to write the changes
   - **"Cancel"** — return to the previous {{askNoun}} (step {{step:instrument/present-workflows}}'s (a)/(b), or Replay step {{step:replay/safety-net}}'s three-option prompt) so the user can pick a different resolution

Never modify existing code on a refactor path without completing this three-step confirmation. Adding new instrumentation wrappers to unchanged functions is not a refactor — this rule does not apply to step {{step:instrument/write-instrumentation}}'s purely-additive instrumentation.`,
        },
        {
            id: "reference",
            title: "Reference",
            body: `These sections are consulted during the Instrument phase — not executed sequentially.

### Trace Plan Format

The trace plan is a strict format. Do not improvise — follow the legend, grammar, and template selection rule below. When in doubt, copy the matching canonical example verbatim and substitute names.

#### Legend

| Symbol | Meaning | Where it appears |
|---|---|---|
| \`●\` | Instrumented span | Default + Expanded + Processor views |
| \`○\` | Skipped function (not instrumented) | Only when the expand modifier is applied (on top of any base template) |
| \`[root]\` | Literal label for the trace function entry point | Always, on its own line above the tree |
| \`[loop]\` | Control-flow group: children execute in a loop | Inside the tree, in place of a span |
| \`[branch]\` | Control-flow group: children are conditional branches | Inside the tree, in place of a span |
| \`[parallel]\` | Control-flow group: children execute concurrently | Inside the tree, in place of a span |
| \`[auto]\` | Auto-captured by a trace processor — no manual instrumentation | Trace-processor view only |
| \`(function)\` \`(llm)\` \`(tool)\` \`(agent)\` \`(handoff)\` | Span type annotation | Immediately after every \`●\` span name |

Brackets \`[…]\` are structural labels (not spans). Parens \`(…)\` are span type annotations (only on \`●\` lines).

#### Grammar rules

1. **Header line** — exactly: \`Trace function: "<trace-function-key>"\` followed by one blank line.
2. **Root** — the next line is the literal \`[root]\`, with no symbol prefix.
3. **Tree body** — uses box-drawing characters only:
   - \`├─\` for every child except the last
   - \`└─\` for the last child
   - Children of a \`├─\` node indent with \`│  \` (pipe + two spaces)
   - Children of a \`└─\` node indent with \`   \` (three spaces, no pipe)
4. **Span lines** — \`<prefix>● <name> (<type>)\`. Type annotation is **required** on every \`●\` line.
5. **Skipped lines** — \`<prefix>○ <name>\`. No type annotation, no description.
6. **Control-flow lines** — \`<prefix>[loop]\` / \`[branch]\` / \`[parallel]\`. They take children but have no symbol and no type.
7. **Footer** — one blank line, then one or both of:
   - \`Files changed:\` followed by a numbered list — every file the cycle will touch. This always includes the replay script path for non-Go projects (\`scripts/replay.*\` new or edited, per step 11b) alongside any instrumented source files. Go-only projects list only the instrumented source files.
   - \`Setup: <one-line setup description>\` (any plan that registers a trace processor)
   Hybrid plans (manual spans + processor) include both, with \`Setup:\` first then \`Files changed:\`. A pure-processor plan still lists \`Files changed:\` because the processor-registration file is edited and the replay script (non-Go) is written. Go-only pure-processor plans with a single registration file and no manual spans may include only \`Setup:\` plus that one file under \`Files changed:\`.
8. **No descriptions, no counts, no parameter details, no blank lines between siblings, no trailing whitespace.**
9. **One trace function per plan.** A trace plan describes exactly one trace function — exactly one \`Trace function: "..."\` header, exactly one \`[root]\`, exactly one tree, exactly one \`Files changed:\` section. If the cycle would require instrumenting two trace functions, that's two cycles, not one plan with two trees.

#### Which template to use (precedence — check top to bottom, stop at first match)

Pick the **base template** from SDK capability and surrounding work:

1. **Trace processor (hybrid) template** — if the SDK guide says to register a processor (e.g. OpenAI Agents SDK \`addTraceProcessor\`) AND there is meaningful work above, alongside, or below the SDK call. The trace function root wraps the broader workflow with manual \`●\` spans; the SDK call appears as one \`(agent)\` child whose grandchildren are the \`[auto]\` lines; other manual spans capture work outside the SDK. This is the default for any trace processor SDK whenever there's surrounding workflow logic — which is almost always.
2. **Trace processor (bare) template** — only when the workflow truly is *just* the SDK call with no surrounding work. Children of the root span are auto-captured and shown as \`[auto]\` lines. Confirm before using this — if the workflow has any input prep, orchestration, retries, post-processing, or non-SDK LLM/tool calls, use the hybrid template instead.
3. **Default view** — every other case (no processor in play). This is the recommended default for SDKs without a processor.

Then apply the **expand modifier**, orthogonally:

- If the user explicitly asks for more detail ("show details", "expand", "include skipped") or selects "Expand details" from the {{askNoun}} preview, add \`○\` skipped lines to whichever base template was picked. Never drop \`[auto]\` lines when expanding a processor template — skipped lines and auto-captured lines coexist in the tree. Without an explicit ask, do not add skipped lines.

Never mix base templates beyond the hybrid pattern. Never invent a fifth variant.

#### Canonical examples (copy-edit-substitute, do not restructure)

**Default view** — instrumented spans only:

\`\`\`
Trace function: "<trace-function-key>"

[root]
● outerFunction (function)
├─ ● llmCall (llm)
└─ [loop]
   ├─ ● anotherLlmCall (llm)
   └─ ● refinementCall (llm)

Files changed:
  1. client.ts
  2. pipeline.ts
\`\`\`

**Default + expand modifier** — adds skipped (○) functions in true execution order. The same modifier applies to processor templates (hybrid or bare) when the user asks for expansion — \`○\` lines coexist with \`[auto]\` lines in that case:

\`\`\`
Trace function: "<trace-function-key>"
● instrumented   ○ skipped

[root]
● outerFunction (function)
├─ ○ helperFormat
├─ ● llmCall (llm)
└─ [loop]
   ├─ ○ evaluateBatch
   ├─ ○ calculateScore
   ├─ ● anotherLlmCall (llm)
   ├─ ● refinementCall (llm)
   └─ ○ evaluateBatch

Files changed:
  1. client.ts
  2. pipeline.ts
\`\`\`

The legend line \`● instrumented   ○ skipped\` appears **only** in the expanded view, immediately under the header.

**Trace-processor (hybrid) view** — workflow with manual spans wrapping auto-captured agent internals (default for processor SDKs):

\`\`\`
Trace function: "handle-user-request"

[root]
● handleUserRequest (function)
├─ ● validateAndPrepareInput (function)
├─ ● runAgent (agent)
│  ├─ LLM calls    [auto]
│  ├─ tool calls   [auto]
│  └─ handoffs     [auto]
├─ ● scoreAgentOutput (llm)
└─ ● persistResult (function)

Setup: addTraceProcessor(processor) registered at startup
Files changed:
  1. handler.ts
  2. tracing/setup.ts
\`\`\`

The \`[auto]\` lines are auto-captured spans — the processor emits them inside the SDK call without manual instrumentation. They use \`├─\`/\`└─\` like normal children but carry no \`●\`/\`○\` symbol because you're not writing the span yourself. Manual \`●\` spans wrap the broader workflow above, alongside, and below the SDK call.

**Trace-processor (bare) view** — only when the workflow IS just the SDK call:

\`\`\`
Trace function: "my-agent"

[root]
● runAgent (function)
├─ LLM calls    [auto]
├─ tool calls   [auto]
└─ handoffs     [auto]

Setup: addTraceProcessor(processor) registered at startup
\`\`\`

Use this **only** when there is genuinely no work above, alongside, or below the SDK call. If there's any input prep, orchestration, retry, post-processing, or non-SDK LLM/tool call, use the hybrid view instead.

#### Anti-examples (do NOT do these)

- ❌ \`* outerFunction (function)\` — use \`●\`, never \`*\` or \`-\` or \`•\`
- ❌ \`● outerFunction\` — type annotation is mandatory on every instrumented span
- ❌ \`● outerFunction (function) — calls the LLM with retries\` — no descriptions, no em dashes
- ❌ \`● outerFunction (llm-call)\` — only the listed types are valid; do not invent new ones
- ❌ \`[Root]\` or \`[ROOT]\` — literal label is lowercase \`[root]\`
- ❌ Mixed indentation widths (2 spaces in one branch, 4 in another)
- ❌ Blank lines between siblings inside the tree
- ❌ Omitting \`Files changed:\` from any plan that has manual \`●\` spans (hybrid trace-processor plans MUST include both \`Setup:\` and \`Files changed:\`)
- ❌ Defaulting to the bare trace-processor view when the workflow has work above, alongside, or below the SDK call — use the hybrid view and add manual spans
- ❌ Putting the SDK's agent call (e.g. \`runAgent\`, \`Runner.run\`) at \`[root]\` when the actual workflow has a clear outer function — the workflow function is the root, the SDK call is a child
- ❌ Inventing extra sections like \`Notes:\` or \`Estimated coverage:\`
- ❌ Two \`Trace function: "..."\` headers in one plan — split into two cycles
- ❌ \`● someFn (llm)   ← description here\` — no inline descriptions, arrows, or trailing commentary on span lines
- ❌ \`● <kind>DocumentCreate (llm)\` — no placeholder/template span names; expand to concrete spans (e.g., three siblings, or under a \`[branch]\`)
- ❌ \`Files changed\` without the trailing colon
- ❌ \`1. lib/bitfab.ts (new) — Bitfab client + exported pipelines\` — file entries are paths only, no annotations or descriptions
- ❌ Recommending an approach that requires "a tiny behavior change" — disqualified at trace plan construction; restructure the tree instead

#### Presentation step

After building the plan according to the rules above, {{askVerb}} with these three options:
- **Proceed** (recommended) — accept the default view as shown
- **Expand details** — re-render using the expanded view template
- **Adjust** — user wants changes; ask what

### Trace Plan Accuracy

Read function signatures {{^codex}}with the \`Read\` tool {{/codex}}when the trace plan will reference their parameter names or return fields. Skipped leaf functions can be named from grep results if their shape isn't exposed in the plan. Never guess names that appear in the plan.`,
        },
    ],
    phases: [
        {
            id: "preamble",
            title: "Preamble",
            stepStyle: "list",
            steps: [
                {
                    id: "render-intro",
                    kind: "action",
                    title: "Render the introduction block",
                    body: `Render the block below **verbatim** as a single message, then continue straight to Login. Do **not** ask for confirmation, do **not** {{^codex}}use AskUserQuestion{{/codex}}{{#codex}}ask any extra questions{{/codex}}, do **not** summarize in your own words.

\`\`\`
Bitfab captures what your AI code does, turns runs into reusable datasets, and verifies fixes by replaying them against real data.

┌──────────┐    ┌──────────┐    ┌──────────┐    ┌──────────┐
│   CODE   │───▶│  TRACES  │───▶│ DATASETS │───▶│ IMPROVE  │
│          │    │ (what it │    │(reusable │    │ (edit +  │
│          │    │   did)   │    │test set) │    │ verify)  │
└──────────┘    └──────────┘    └──────────┘    └──────────┘

Primitives
  • Trace   — a recording of one workflow run (inputs, outputs, every step inside).
              Ground truth for what your code actually did.
  • Dataset — a curated collection of traces (failures, a specific workflow, custom).
              The reusable test set your changes get measured against.
  • Replay  — a tool that re-runs a dataset through your current code.
              Turns production data into a ready-made regression test.

Setup runs in two phases:
  1. LOGIN                 — authenticate (15s, browser)
  2. INSTRUMENT + REPLAY   — {{^codex}}run in parallel per workflow{{/codex}}{{#codex}}written together per workflow{{/codex}}:
     • INSTRUMENT          — wrap your workflows with tracing (purely additive)
     • REPLAY              — generate a replay script for your trace functions
\`\`\`

Then proceed to Login.`,
                    next: "login/status-check",
                },
            ],
        },
        {
            id: "login",
            title: "Login",
            stepStyle: "list",
            intro: "Authenticate with Bitfab and retrieve the API key.",
            outro: "**If running `login` only**, stop here and report the result.",
            steps: [
                {
                    id: "status-check",
                    kind: "branch",
                    toolCalls: ["bash"],
                    title: "Run the status check",
                    body: `Run the status check:

\`\`\`bash
node "{{pluginRoot}}/dist/commands/status.js"
\`\`\`{{#codex}}

If the output includes \`v<X> available — run ... to update\`, surface that line verbatim to the user once before proceeding; do not block on it.{{/codex}}

If **already authenticated**, skip to step {{step:login/get-api-key}}.`,
                    branches: [
                        {
                            when: "already authenticated",
                            next: "login/get-api-key",
                        },
                        {
                            when: "not authenticated",
                            next: "login/run-login",
                        },
                    ],
                },
                {
                    id: "run-login",
                    kind: "branch",
                    toolCalls: ["bash"],
                    title: "Run the login script",
                    body: `If **"not authenticated"**, run the login script yourself — do NOT ask the user to run it manually:

\`\`\`bash
node "{{pluginRoot}}/dist/commands/login.js"
\`\`\`
{{^codex}}
This opens the browser for OAuth and waits for the loopback callback. Run with 600000ms timeout (10 minutes). If the command **exits with an error**, **fails to reach the browser**, or **times out** — fall through to the **Login (headless)** flow below. This commonly happens on SSH sessions, sandboxed environments, cloud IDEs, and Codespaces where the browser can't reach the CLI's loopback port.{{/codex}}{{#codex}}
Run with 600000ms (10 minute) timeout. This opens the browser to a URL that **races two delivery paths**: a loopback HTTP callback on 127.0.0.1, and a server-polled handoff ticket (PKCE). Whichever resolves first wins — the ticket path works fine even when the CLI is sandboxed (Codex seatbelt, containers, SSH, cloud IDEs) because the browser delivers the token to the server and the CLI polls for it.

**Per the Blocking-process rule at the top:** after \`login.js\` prints its URL, keep polling the live exec session until it exits. Do not send a "waiting for you to sign in" message and then idle — the user's sign-in will NOT arrive as a chat message; it arrives as \`login.js\` exiting with \`Logged in as <email>\` on stdout.

**If the browser auto-launch appears to fail (Chrome crashes, you see a crash report, the browser never opens, no default browser is configured), do NOT abandon \`login.js\`.** The loopback server and the ticket channel both live inside the running \`login.js\` process — they stay valid for the full 10-minute timeout whether or not the auto-launch worked. Your job in that case is:

1. Leave \`login.js\` running. Do not kill it, do not start a new login, do not switch to the headless paste flow.
2. Copy the exact URL \`login.js\` printed (the one with \`?callbackPort=…&ticket=…\`) and surface it to the user verbatim. Ask them to open it in any browser on any device.
3. After ~4 seconds \`login.js\` will re-print the same URL with a "browser didn't open — paste this URL manually" banner; watch for that and relay it to the user if you see it.
4. Keep waiting on the same process. When the user opens the URL, the ticket channel (or the loopback callback if the browser is on the same machine and can reach loopback) will resolve and \`login.js\` will exit 0 with the credentials saved.

**Do NOT fall through to the headless flow pre-emptively** just because the environment looks sandboxed or the browser crashed. The ticket path is designed for exactly those environments, and it's still live. Only fall through to **Login (headless)** below when all of the following are true:
- \`login.js\` exited non-zero (not just "the loopback didn't bind"), OR
- the 10-minute timeout actually elapsed, OR
- the printed URL contained neither \`callbackPort=\` nor \`ticket=\` (indicating both channels failed to start — log the full login output so the user can debug).

If \`login.js\` is still running and the URL it printed includes \`ticket=<uuid>\`, the ticket channel is live — keep waiting; do not abandon it for a paste flow.{{/codex}}`,
                    branches: [
                        {
                            when: "login succeeds",
                            next: "login/get-api-key",
                        },
                        {
                            when: "login fails, errors, or times out",
                            next: "login-headless/determine-service-url",
                        },
                    ],
                },
                {
                    id: "get-api-key",
                    kind: "action",
                    toolCalls: ["apiKey"],
                    title: "Retrieve the API key",
                    body: "Call {{tool:apiKey}} to retrieve the API key — **NEVER print or log the full key**. Stored at `~/.config/bitfab/credentials.json`, used for the `BITFAB_API_KEY` environment variable.",
                    next: "login/ask-session-log-consent",
                },
                {
                    id: "ask-session-log-consent",
                    kind: "action",
                    toolCalls: ["bash", "ask"],
                    title: "Ask for session log consent (first login only)",
                    body: `Check whether session log consent has already been recorded:

\`\`\`bash
node -e "const fs=require('fs'),os=require('os'),p=require('path').join(os.homedir(),'.config/bitfab/config.json');const c=JSON.parse(fs.existsSync(p)?fs.readFileSync(p,'utf8'):'{}');console.log(c.sessionLogConsent??'null')"
\`\`\`

If the output is already \`true\` or \`false\`, skip the prompt and continue. If the output is \`null\`, {{askVerb}}:
- **Question:** "Allow Bitfab to collect session logs?"
- **Description:** Used to diagnose issues and improve the product.
- **Options:** "Allow" / "Don't allow"

Save the answer (replace \`CONSENT\` with \`true\` or \`false\`):

\`\`\`bash
node -e "const fs=require('fs'),os=require('os'),p=require('path').join(os.homedir(),'.config/bitfab/config.json');fs.mkdirSync(require('path').dirname(p),{recursive:true});const c=JSON.parse(fs.existsSync(p)?fs.readFileSync(p,'utf8'):'{}');c.sessionLogConsent=CONSENT;fs.writeFileSync(p,JSON.stringify(c,null,2)+'\\n')"
\`\`\``,
                    next: {
                        byMode: {
                            all: "instrument/detect-language",
                            login: null,
                            "login-headless": null,
                            instrument: null,
                            modify: null,
                            view: null,
                            replay: null,
                            templates: null,
                        },
                    },
                },
            ],
        },
        {
            id: "login-headless",
            title: "Login (headless)",
            stepStyle: "list",
            intro: "{{^codex}}Use this flow when the browser callback can't reach the terminal — SSH sessions, sandboxed environments, cloud IDEs, Codespaces, CI runners. Triggered explicitly by `{{cmd}}setup login headless`, or as an automatic fallback when the normal Login flow above fails.{{/codex}}{{#codex}}Use this flow only when `login.js` itself can't deliver — i.e. both the loopback callback AND the server-polled ticket channel failed to start, or `login.js` exited with an error / hit the 10-minute timeout. Sandboxed environments (Codex seatbelt, containers, SSH, cloud IDEs) on their own are NOT a reason to use this flow — the ticket channel inside `login.js` handles those. Triggered explicitly by `{{cmd}}setup login headless`, or as an automatic fallback per the strict conditions in the Login step above.{{/codex}}",
            steps: [
                {
                    id: "determine-service-url",
                    kind: "action",
                    toolCalls: ["bash", "read"],
                    title: "Determine the service URL",
                    body: "Determine the service URL. Default is `https://bitfab.ai`. If the user has a custom deployment, read it from `~/.config/bitfab/config.json` (field `serviceUrl`) or the `BITFAB_SERVICE_URL` environment variable.",
                    next: "login-headless/show-url",
                },
                {
                    id: "show-url",
                    kind: "action",
                    title: "Show the auth URL to the user",
                    body: `Tell the user:
> Open this URL in a browser on any device: **{serviceUrl}/plugin/auth/{{authTarget}}**
>
> Sign in with your Bitfab account. The page will show an API key with a copy button. Paste the token here when you have it.`,
                    next: "login-headless/wait-for-token",
                },
                {
                    id: "wait-for-token",
                    kind: "action",
                    title: "Wait for the user's token",
                    body: "Wait for the user's next message — it will contain the token. Do NOT {{askVerb}} here (it adds an unnecessary extra step before the user can paste).",
                    next: "login-headless/validate-token",
                },
                {
                    id: "validate-token",
                    kind: "action",
                    toolCalls: ["bash"],
                    title: "Validate the token",
                    body: `When the user pastes the token, validate it with curl — do NOT echo the token back to the user or print it in any output:

\`\`\`bash
curl -fsS -H "Authorization: Bearer <TOKEN>" "{serviceUrl}/api/plugin/whoami"
\`\`\`

If this returns 200 with a JSON body containing \`user.email\`, the token is valid. If it fails, tell the user the token was invalid and ask them to re-paste (do not re-print the bad token).`,
                    next: "login-headless/save-token",
                },
                {
                    id: "save-token",
                    kind: "action",
                    toolCalls: ["bash", "write"],
                    title: "Save the token to credentials",
                    body: `Save the token to \`~/.config/bitfab/credentials.json\` {{#codex}}writing a file{{/codex}}{{^codex}}using the \`Write\` tool{{/codex}} with this exact content (replace \`<TOKEN>\` with the pasted value, nothing else):

\`\`\`json
{
  "apiKey": "<TOKEN>"
}
\`\`\`

Create the \`~/.config/bitfab/\` directory first if it doesn't exist:

\`\`\`bash
mkdir -p ~/.config/bitfab
\`\`\``,
                    next: "login-headless/confirm-success",
                },
                {
                    id: "confirm-success",
                    kind: "action",
                    title: "Confirm sign-in to the user",
                    body: 'Confirm success to the user by referencing the email returned from `/api/plugin/whoami` — e.g. "Signed in as alice@example.com." **Never echo the token back.**',
                    next: "login-headless/ask-session-log-consent",
                },
                {
                    id: "ask-session-log-consent",
                    kind: "action",
                    toolCalls: ["bash", "ask"],
                    title: "Ask for session log consent (first login only)",
                    body: `Check whether session log consent has already been recorded:

\`\`\`bash
node -e "const fs=require('fs'),os=require('os'),p=require('path').join(os.homedir(),'.config/bitfab/config.json');const c=JSON.parse(fs.existsSync(p)?fs.readFileSync(p,'utf8'):'{}');console.log(c.sessionLogConsent??'null')"
\`\`\`

If the output is already \`true\` or \`false\`, skip the prompt and continue. If the output is \`null\`, {{askVerb}}:
- **Question:** "Allow Bitfab to collect session logs?"
- **Description:** Used to diagnose issues and improve the product.
- **Options:** "Allow" / "Don't allow"

Save the answer (replace \`CONSENT\` with \`true\` or \`false\`):

\`\`\`bash
node -e "const fs=require('fs'),os=require('os'),p=require('path').join(os.homedir(),'.config/bitfab/config.json');fs.mkdirSync(require('path').dirname(p),{recursive:true});const c=JSON.parse(fs.existsSync(p)?fs.readFileSync(p,'utf8'):'{}');c.sessionLogConsent=CONSENT;fs.writeFileSync(p,JSON.stringify(c,null,2)+'\\n')"
\`\`\``,
                    next: "login-headless/continue-or-stop",
                },
                {
                    id: "continue-or-stop",
                    kind: "action",
                    title: "Continue with setup or stop",
                    body: "Continue with the rest of setup, or stop if running `login headless` only.",
                    next: {
                        byMode: {
                            all: "instrument/detect-language",
                            login: null,
                            "login-headless": null,
                            instrument: null,
                            modify: null,
                            view: null,
                            replay: null,
                            templates: null,
                        },
                    },
                },
            ],
        },
        {
            id: "instrument",
            title: "Instrument",
            stepStyle: "list",
            intro: "Instrument the codebase with Bitfab tracing. Requires authentication (run Login first if needed).\n\nBitfab captures every AI function call — inputs, outputs, and errors — so you can see exactly what your AI is doing and discover what's going wrong. The goal is to have enough context in each trace to tell whether a call succeeded or failed, and why.",
            steps: [
                {
                    id: "detect-language",
                    kind: "action",
                    toolCalls: ["bash", "grep", "glob", "read"],
                    title: "Detect the project language and frameworks",
                    body: `**Detect the project language** (TypeScript, Python, Ruby, or Go). In a monorepo, identify which directories are **applications** (services, APIs, agents) vs **libraries** (SDKs, shared packages). Focus on application directories. Also scan imports and package manifests for supported framework signals, and note which framework each application directory uses — step {{step:instrument/read-sdk-reference}} fetches the matching framework page alongside the language reference:
- **LangGraph / LangChain** — TS: \`@langchain/langgraph\`, \`@langchain/core\`; Python: \`langgraph\`, \`langchain\`, \`langchain_core\`
- **OpenAI Agents SDK** — TS: \`@openai/agents\`, \`setTraceProcessors\`; Python: \`agents\` (\`from agents import ...\`)
- **Claude Agent SDK** — TS: \`@anthropic-ai/claude-agent-sdk\`, \`ClaudeSDKClient\`; Python: \`claude_agent_sdk\`, \`ClaudeSDKClient\`
- **BAML** — TS: \`@boundaryml/baml\`, \`baml_client\` import; Python: \`baml-py\`, \`from baml_client import b\``,
                    next: "instrument/search-existing",
                },
                {
                    id: "search-existing",
                    kind: "branch",
                    toolCalls: ["grep", "glob", "ask"],
                    title: "Search for existing SDK usage",
                    body: `**Search for existing SDK usage** (\`withSpan\`, \`@span\`, \`bitfab_span\`, \`client.Span\`, \`getFunction\`, \`get_function\`, etc.). In a monorepo, search **each application directory separately** — a root-level search can miss subdirectories.
- If found: list the trace function keys, then {{askVerb}}:

{{branches}}

  If "Modify", jump to the Modify phase. If "Continue", skip to Replay.
- **If usage routes through a project-local shim** (a wrapper file that re-exports \`withSpan\` / \`@span\` / \`bitfab_span\` / \`getCurrentTrace\` / \`getCurrentSpan\` with custom init, often named \`lib/bitfab.*\` or after a predecessor SDK such as \`lib/simforge.*\`), audit the shim before instrumenting anything new. The shim must (a) construct the SDK client (\`new Bitfab(...)\`, \`bitfab_init()\`, \`Bitfab::Client.new\`, etc.) at module load, **synchronously**, never lazily inside the wrapped function; and (b) hand off to the SDK call synchronously, with no \`await\` between the user's entry to the shim and \`client.withSpan(...)\` / \`@bitfab.span(...)\`. Lazy or async client init (e.g. \`await getOrCreateTraceFunction(key)\` inside the wrapped body) breaks the SDK's nesting context (TypeScript \`AsyncLocalStorage\`, Python \`contextvars\`) under any parallel fan-out (\`Promise.all\`, \`Promise.allSettled\`, \`asyncio.gather\`, parallel workers): every span becomes its own top-level trace instead of nesting inside its caller. Fix the shim before instrumenting anything new. (Direct callers of the SDK with no shim already satisfy this rule, skip the audit.)
- If not found: **proceed to step {{step:instrument/get-api-key}}** — no SDK usage does NOT mean nothing to instrument, it means the SDK hasn't been installed yet. NEVER conclude "nothing to instrument" before completing step {{step:instrument/explain-trace-boundary}}.`,
                    branches: [
                        {
                            when: "no existing SDK usage found",
                            next: "instrument/get-api-key",
                        },
                        {
                            option: {
                                letter: "A",
                                label: "Search for more workflows",
                                description: "find uninstrumented gaps",
                            },
                            recommended: true,
                            next: "instrument/get-api-key",
                        },
                        {
                            option: {
                                letter: "B",
                                label: "Modify an existing trace setup",
                                description: "jump to the Modify phase",
                            },
                            next: "modify/gather-existing",
                        },
                        {
                            option: {
                                letter: "C",
                                label: "Continue",
                                description: "skip to replay",
                            },
                            next: {
                                byMode: {
                                    all: "replay/gather-keys",
                                    login: null,
                                    "login-headless": null,
                                    instrument: null,
                                    modify: null,
                                    view: null,
                                    replay: null,
                                    templates: null,
                                },
                            },
                        },
                    ],
                },
                {
                    id: "get-api-key",
                    kind: "action",
                    toolCalls: ["apiKey"],
                    title: "Use or retrieve the API key",
                    body: "Use the API key from the Login phase (or retrieve it now if already authenticated)",
                    next: "instrument/install-sdk",
                },
                {
                    id: "install-sdk",
                    kind: "action",
                    toolCalls: ["bash", "read", "edit"],
                    title: "Install the SDK",
                    body: `**Install the SDK now.** Detect the project's package manager from its manifest (\`pyproject.toml\` → \`uv\`/\`poetry\`; \`package.json\` → \`pnpm\`/\`npm\`/\`yarn\`/\`bun\`; \`Gemfile\` → \`bundle\`; \`go.mod\` → \`go get\`; \`requirements.txt\` → edit file + \`pip install -r\`) and run its canonical add command — do NOT stop to ask about version pinning or dep groups. Prefer \`uv add\`/\`poetry add\` over bare \`pip install\` (bare \`pip install\` doesn't persist to pyproject.toml). In monorepos, scope to the correct workspace (e.g. \`pnpm add --filter <pkg>\`, or cd into the app directory first) — running from the repo root will install into the wrong package. Default to a runtime dep for applications; a dev dep for libraries/SDKs where a runtime dep would propagate to downstream users. Then set the \`BITFAB_API_KEY\` environment variable.

**Tell the user what you did.** Pick the env-handling approach that fits the project's existing convention. Whatever you do, surface it explicitly: name the file (with absolute path) or mechanism you used, so the user knows where the key now lives. Do not print the key value itself. If the key landed in a \`.env\`-style file, additionally tell the user that any already-running dev server, REPL, or test runner may need a restart to pick it up, since most file watchers reload code on save but not env files.`,
                    next: "instrument/read-sdk-reference",
                },
                {
                    id: "read-sdk-reference",
                    kind: "action",
                    toolCalls: ["webFetch"],
                    title: "Read the SDK reference",
                    body: "**Read the SDK reference.** Fetch the dense canonical reference page first — `/reference/typescript`, `/reference/python`, `/reference/ruby`, or `/reference/go` — for every signature, type, default, and error semantic you need (initialization, `withSpan` / `@span` / `bitfab_span` / `client.Span`, `getFunction` / `get_function` / `GetFunction` / `bitfab_function`, `SpanType`, `getCurrentSpan`/`getCurrentTrace`, `wrapBAML`/`wrap_baml`). If step {{step:instrument/detect-language}} detected a framework in this application directory, also fetch the matching framework page — it documents the handler/processor/wrapper the SDK exposes for that framework, which is usually preferable to hand-wrapping every node/agent call with `withSpan`/`@span`: LangGraph/LangChain → `/frameworks/langgraph` (`getLangGraphCallbackHandler` / `get_langgraph_callback_handler`); OpenAI Agents SDK → `/frameworks/openai-agents` (`getOpenAiTracingProcessor` / `get_openai_tracing_processor`); Claude Agent SDK → `/frameworks/claude-agent-sdk` (`getClaudeAgentHandler` / `get_claude_agent_handler`); BAML → `/frameworks/baml` (`wrapBAML` / `wrap_baml`). Then fetch the language guide (`/typescript-sdk`, `/python-sdk`, `/ruby-sdk`, `/go-sdk`) — including the `#replay` section for non-Go projects — for the install command, the multi-file project layout example, the BAML auto-instrumentation walkthrough, and the replay script template. Read the replay section upfront (not later) because step {{step:instrument/write-instrumentation}} writes the replay pipeline {{parallelism}} instrumentation. {{^codex}}Use WebFetch or ask the user to share the pages{{/codex}}{{#codex}}Fetch those pages (via web-fetch capability, or ask the user to share them){{/codex}}. **Do not improvise instrumentation from memory** — the API has moved and guessing will produce broken code.",
                    next: "instrument/explain-trace-boundary",
                },
                {
                    id: "explain-trace-boundary",
                    kind: "action",
                    title: "Establish trace-boundary rules and serializability constraint",
                    body: `**The root exists so the replay harness can re-invoke it as a plain lambda with serialized inputs** — that's what makes traces searchable (a coherent unit of behavior) and replayable (runnable against current code). The root must own its state setup, not consume a pre-built stateful object the replay script can't reconstruct. Frameworks are the sharpest case (LangGraph compiled graphs, Claude Agent SDK clients, LangChain chains all require constructors + special setup), but the rule generalizes to anything stateful — configured SDK clients, prepared models, cached routers, DB sessions. The root is therefore the outer workflow function that **builds** the framework / stateful object + invokes it + processes the output (API handler, message processor, job runner, pipeline coordinator) — almost never the SDK's \`run()\` / \`invoke()\` itself.

**Hard constraint: every wrapped function's inputs and outputs must be serializable by the SDK's tracing layer so traces can be replayed.** Every span input and output gets serialized into the trace using the SDK's language-native serialization (TypeScript/JSON, Python/JSON via Pydantic, Ruby/\`to_json\`, Go/\`json.Marshal\`). If a wrapped function takes live runtime objects that don't round-trip through that serialization, the trace can't be replayed, and badly-failing inputs can drop the entire span on the floor (not just garble the input field). Examples of unserializable inputs:
- browser objects (\`MediaStream\`, \`RTCPeerConnection\`, \`WebSocket\`, DOM refs)
- HTTP \`Request\` / \`Response\`, stream writers, open sockets
- framework request contexts whose content is genuinely opaque (not reconstructible from headers + user id)
- **live SDK client instances passed as arguments** (LLM clients like \`OpenAI\` / \`Anthropic\` / Bedrock, configured agents, DB connection objects, HTTP agents): class instances whose internals carry circular references, function members, or platform handles all sink superjson and \`JSON.stringify\`. Watch especially for an options/config bag (e.g. \`options.llmProvider\`, \`ctx.db\`) that smuggles a live client into an otherwise-serializable signature.

Module-level dependencies (DB clients, env vars, config loaders, LLM clients) do **not** count *when accessed via module scope or closure*: replay inherits them from the app's loaded environment. The same client passed *as a function argument* IS captured as input and WILL fail. The fix when an SDK client is the only unserializable piece is usually trivial: hoist it to module scope (or capture via closure) and drop it from the argument list, leaving the wrapped function's serializable args (issue, request, options-without-the-client) intact. When the natural outer boundary still has unserializable inputs after that, do **one** of the following **before writing code**:
- **Move the trace boundary inward** to the first function whose inputs are serializable (e.g. trace \`processTurn(transcript, context)\` instead of \`handleSession(stream, peerConnection)\`). This is not a refactor.
- **Refactor** so a function with serializable inputs exists. Two flavors, chosen per case in the refactor plan:
  - **Visibility refactor (common)** — the logic that takes serializable inputs already exists inline but isn't importable (embedded in a route handler, not exported). Extract it into a named, exported function at module scope. No semantic change.
  - **Structural refactor (rare overall, common for realtime/streaming/browser apps)** — no function with serializable inputs exists yet. Introduce one: a pure core whose parameters are serializable, with callers constructing them. A real rewrite.

Raise this with the user in step {{step:instrument/present-workflows}} (not later) — never instrument a root with unserializable inputs and try to fix it in the Replay phase.`,
                    next: "instrument/identify-workflows",
                },
                {
                    id: "identify-workflows",
                    kind: "action",
                    toolCalls: ["read", "grep"],
                    title: "Identify all AI workflows in the codebase",
                    body: "Read the codebase to identify ALL AI workflows — every place the app makes LLM calls, runs agents, or makes AI-driven decisions. For each, find the **outer workflow boundary** (per the rule in step {{step:instrument/explain-trace-boundary}}), and also note any meaningful work **above** the agent/LLM call (auth, validation, input prep, retry/orchestration loops, multi-agent coordination), **alongside** it (custom LLM calls outside the SDK, tools that aren't registered with the SDK, downstream services), and **below** it (post-processing, parsing, persistence). These are the manual spans that will sit around any auto-captured SDK content.",
                    next: "instrument/present-workflows",
                },
                {
                    id: "present-workflows",
                    kind: "action",
                    toolCalls: ["ask"],
                    refs: ["refactor-confirmation"],
                    title: "Present workflows and pick one",
                    body: `Present a numbered list of workflows found, ordered by value (most complex or LLM-heavy first). For each, give:
- **Trace boundary** — the outer workflow function that will be the trace function root (per step {{step:instrument/explain-trace-boundary}} — NOT the SDK/agent call itself)
- **Inputs** — the shape of the function's inputs, and an explicit note that they're serializable by the SDK's tracing layer. If the natural outer boundary's inputs are unserializable (live browser/runtime objects, HTTP req/res, stream writers, sockets, opaque request contexts), state that here and present the two resolutions from step {{step:instrument/explain-trace-boundary}} as part of this workflow's entry: **(a) move the boundary inward to \`<specific inner function with serializable inputs>\`** (recommended when an obvious candidate exists — not a refactor), or **(b) refactor**. Do not proceed to step {{step:instrument/read-signatures}} until the user picks one — never instrument an unserializable root. **If the user picks (b), present a refactor plan — labeled as *visibility* (extract + export, logic unchanged) or *structural* (new pure-core fn) — and get an explicit second confirmation before modifying code. See the "Refactor confirmation" rule below.**
- **What's covered end-to-end** — the work above, alongside, and below any agent/LLM/SDK call that this trace will capture (be specific: list the orchestration, custom LLM calls, tools, downstream services that will become spans)
- **Why tracing it is valuable**

The description must commit to the actual scope. If the plan will only auto-capture an SDK's internals, say so explicitly — do NOT use language like "complete tracing of X workflow" when the trace will only cover an SDK call's internals.

Recommend one to start with. **Ask the user to pick exactly ONE workflow to instrument first.** Never accept "multiple" or "all" — each Instrument cycle produces exactly one trace function with one trace plan and one set of code changes. If the user wants to instrument several, they will be done sequentially via the loop in step {{step:instrument/mandatory-stop}}, one at a time.`,
                    next: "instrument/read-signatures",
                },
                {
                    id: "read-signatures",
                    kind: "action",
                    toolCalls: ["read"],
                    refs: ["reference"],
                    title: "Read function signatures referenced by the trace plan",
                    body: '**Read function signatures you\'ll reference in the trace plan** — root function first, then any whose parameter names or return fields aren\'t already obvious from the step {{step:instrument/identify-workflows}} scan. Skipped leaf functions only need their names; don\'t Read them unless their shape appears in the plan. Never guess names. See "Trace Plan Format" and "Trace Plan Accuracy" in the Reference section below.',
                    next: "instrument/build-trace-plan",
                },
                {
                    id: "build-trace-plan",
                    kind: "reference",
                    refs: ["reference"],
                    toolCalls: ["ask", "bash", "createTracePlan", "getTracePlan"],
                    title: "Build and present the trace plan",
                    body: `**Build the trace plan under a hard constraint: the resulting instrumentation must be purely additive.** If a candidate tree requires *any* behavior change to make spans nest correctly (awaiting a stream that wasn't awaited, delaying a call, reordering operations, blocking a callback, restructuring control flow), the tree is invalid — restructure the *tree* instead (make spans siblings, split into separate trace functions across separate cycles, or accept a flatter shape). Never present a behavior-changing approach as an option, not even as a non-recommended alternative.

**For trace processor SDKs (OpenAI Agents SDK, etc.) — extend beyond the processor.** The processor only auto-captures what runs *inside* the SDK's instrumented call (LLM calls, tool calls, handoffs). Everything above it (orchestration, retries, input prep), alongside it (non-SDK LLM calls, unregistered tools, downstream services), and below it (post-processing, persistence) is invisible unless you add manual spans. Default to a **hybrid plan**: trace function root wraps the workflow with manual \`●\` spans, the SDK call appears as one \`(agent)\` child whose grandchildren are \`[auto]\` lines, and other manual spans capture the work around it. A bare auto-only plan (root = the SDK call, no surrounding manual spans) is only valid when the workflow truly is just the SDK call with no surrounding work — confirm there's nothing meaningful above/alongside/below before defaulting to it.

**One flow = one trace function key.** When an outer \`@bitfab.span\` / \`withSpan\` / \`bitfab_span\` and a framework handler wrap the same work (LangGraph \`get_langgraph_callback_handler\`, Claude Agent SDK \`get_claude_agent_handler\`), pass the **same key** to both — a second key splits one flow into two overlapping trace functions. Separate trace functions describe separate flows with their own standalone roots, never a sub-range of an outer flow.

Then post the plan to the browser confirmation UI via {{tool:createTracePlan}} and open it with the \`openTracePlan.js\` CLI — that command races a loopback callback and a server-polled handoff ticket so the browser → CLI hand-off works on SSH, Docker, WSL, cloud IDEs, etc. Same delivery pattern as \`login.js\` and \`startDataset.js\`.

- Build a \`TracePlanTree\` (\`{ rootId, nodes: { [id]: TraceNode } }\`) from the same span tree you'd otherwise render. Each \`TraceNode\` carries \`id\` (stable, e.g. hash of \`file:line:name\`), \`name\`, \`kind\` ("manual" | "auto" | "pure"), \`file\`, \`line\`, \`signature\`, \`parentId\`, \`childIds\`, plus \`framework\` (for \`[auto]\` lines).
- **Every captured node MUST include \`sampleInput\` and \`sampleOutput\`.** Without samples the confirmation page can't show the user what gets captured, which is the whole point. Construct realistic example values from the function's parameter and return types (Read the file and its return-type imports if needed); for SDK calls (\`openai.chat.completions.create\`, \`generateText\`, \`cohere.rerank\`, etc.) use the documented response shape. Do NOT call \`create_trace_plan\` with a captured node missing either field.
- **Include surrounding code as \`pure\` context nodes** so the captured set is legible inside its codebase context and the user can toggle additional nodes into the capture directly in the UI without leaving the page. The test for inclusion is **"would the user plausibly want this as its own span?"** — anything they might promote to a wider root, wrap as a deeper child, or add as a peer at the same depth. Walk in three directions:
  - **~10 callers above the root** — candidates for **promoting the root upward** to a wider scope. Walk via Grep (callers of the root, then callers of those, etc.) and attach each as a \`pure\` ancestor. Stop at process entry points (HTTP handlers, queue workers, CLI \`main\`, cron jobs, page handlers, framework boot — there is no useful root above those) or when you've gathered ~10 nodes.
  - **~10 callees below each leaf** — candidates for **wrapping deeper spans**. For every captured leaf, walk downward (callees of that leaf, callees of those, etc.) and attach each as a \`pure\` descendant. Include any callee the user might plausibly want as its own span — LLM / tool / agent calls, prompt construction, response parsing, retry loops, fan-outs, post-processing that drives another model. Stop at pure plumbing (pass-through returns, trivial formatting or arithmetic, no further interesting activity) or ~10 nodes per leaf. **Don't stop just because you crossed an SDK / framework / stdlib boundary** — the test is "is this plausibly its own span?", not "is this in our code?".
  - **~5 siblings per captured non-root node** — candidates for **peer spans at the same depth**. For each captured non-root node, include the parent's other callees (other functions invoked from the same wrapper) as \`pure\` siblings. These are the nodes the user might wrap alongside the existing capture to widen the trace sideways.
  All surrounding nodes get \`kind: "pure"\` and are **not** included in \`capturedNodeIds\`. They serve two ends: **legibility** (the captured set sits inside its surrounding code so the user sees what is and isn't traced) and **modification** (they are the levers in the UI for expanding capture deeper, broader, or sideways).
- Call {{tool:createTracePlan}} with \`{ language, tree, capturedNodeIds, traceFunctionKey }\` (and \`stats\` if you have a sample run) — \`capturedNodeIds\` is your initial recommendation, must form a connected sub-tree (selecting any descendant implies its ancestors). \`traceFunctionKey\` is the key you'll pass to \`getFunction\` / \`get_function\` / \`bitfab_function\` / \`WithFunctionName\` in step {{step:instrument/write-instrumentation}}; persisting it lets future Modify cycles bootstrap their \`before\` tree from this plan via \`get_trace_plan({ traceFunctionKey })\` instead of re-deriving from code. The tool returns a plan id (and a \`https://bitfab.ai/trace-plan/<id>\` URL).
- Open the trace plan in the browser by running:

\`\`\`bash
node "{{pluginRoot}}/dist/commands/openTracePlan.js" <planId>
\`\`\`

(\`{{pluginRoot}}\` resolves to the plugin directory; \`<planId>\` is the id returned by {{tool:createTracePlan}}.) The script opens the trace plan page and **blocks** until the user clicks **Confirm** or **Chat about this** in the browser{{#codex}} (up to 30 minutes){{/codex}}.{{#codex}} The script prints a handoff URL (with \`callbackPort=\` and \`ticket=\` query params) — surface it to the user verbatim so they can paste it into a browser if the auto-launch didn't open one.

**Polling (mandatory — see the Blocking-process rule at the top of this skill):** after surfacing the URL, keep polling the live exec session until the process exits. Do NOT stop after printing the URL. Do NOT wait for a chat message from the user — their confirmation arrives as stdout on the already-running process, not as a new prompt. Poll every few seconds until one of these terminal conditions, then route accordingly:{{/codex}}

- On exit, parse the final stdout line:
  - \`Trace plan confirmed [via …]\` — the user confirmed in the browser. Call {{tool:getTracePlan}} with the plan id to read the authoritative \`capturedNodeIds\` for step {{step:instrument/write-instrumentation}}. If it differs from your initial recommendation, prune \`[auto]\` lines whose ancestor manual span was uncaptured, and drop manual \`●\` wraps that aren't in the set.
  - \`Trace plan cancelled [via …]\` — the user aborted from the browser. Tell them the trace setup was dropped and ask what they'd like to do instead. Do not write instrumentation.
  - non-zero exit / timeout — surface the error to the user. Do not write instrumentation.

**Inline fallback** (use only if {{tool:createTracePlan}} errors, e.g. offline or MCP unreachable): present the trace plan **using the format defined in the "Trace Plan Format" reference section below** (legend → grammar → template precedence → canonical example). **STOP** — {{askVerb}} to confirm before writing code.`,
                    next: "instrument/write-instrumentation",
                },
                {
                    id: "write-instrumentation",
                    kind: "action",
                    toolCalls: ["edit", "webFetch"],
                    title: "Write instrumentation and replay pipeline together",
                    body: `{{^codex}}**Write instrumentation (main agent) AND replay pipeline (subagent) concurrently — to overlap code *generation*, not just file I/O.** Dispatch in a single message: your Edit calls for 11a, plus {{subagentInvocation}} for 11b. The subagent generates its replay code in parallel with your instrumentation generation — parallel Edit calls alone only overlap millisecond file writes, a subagent overlaps the seconds-to-minutes of token generation. Skip the subagent entirely for Go-only projects (Go does not support replay).

- **11a. Instrumentation edits (main agent)** — follow the SDK reference exactly, purely additive. Never change behavior, arguments, return values, error handling, variable names, types, control flow, or code structure. Batch repetitive edits in parallel; for large mechanical fan-outs (>10 files of the same wrapper pattern), validate the pattern on one file, then delegate the rest to a separate subagent (distinct from the 11b subagent).

- **11b. Replay pipeline subagent** — the subagent won't see your conversation. Brief it fully and self-containedly:
  - **Language + SDK replay reference URL** — \`https://docs.bitfab.ai/<language>-sdk#replay\` (TypeScript / Python / Ruby). Tell it to WebFetch this first to ground its code in the current API — do not paste the reference content into the brief.
  - **Trace function key** — confirmed in the trace plan.
  - **Trace function root** — name, full signature (param names + types), return type, absolute file path, and import path the replay script will use.
  - **Replay script target** — path to an existing script if one exists (\`scripts/replay.*\` or the project's equivalent — add a new pipeline entry), otherwise the path to create new.
  - **Non-negotiables**: CLI arg for pipeline name; optional \`--limit N\` (default 10) and \`--trace-ids id1,id2\` flags; replay fn imports and invokes the real function (never a stub); runs in the app's loaded \`.env\` environment (no mocked DB clients / env vars / config / models); mocks only what has no live counterpart at replay time (stream writers, session/request stubs); follows the Replay Output Contract (emit the full \`ReplayResult\` as one JSON block via \`JSON.stringify(result, null, 2)\` / \`json.dumps(result, indent=2, default=str)\` / \`JSON.pretty_generate(result)\`, including every item's \`durationMs\`/\`duration_ms\`, \`tokens\`, and \`model\`; never swap the JSON block for per-field log lines, counts, lengths, hashes, or previews); prints a short human-readable summary + test run URL before the JSON dump; lives under \`scripts/\` (or the project's existing scripts location).
  - **Match the \`#replay\` template's fn signature verbatim — no speculative defense.** The SDK invokes the replay wrapper with captured args in their original shape; don't branch on arg arity/shape, don't add type-checker escape hatches (\`any\` casts, \`cast(Any, ...)\`, ignore comments, untyped passthroughs), and don't guard against cases the contract precludes. If the root signature in the brief contradicts what the reference template expects, return that fact so the main agent can re-check; don't paper over it in code. A hard error at the call site beats silent passthrough of malformed input.
  - **Per-item error tolerance** — \`bitfab.replay\` records thrown wrapped-fn errors in \`item.error\` and keeps going; rely on that. Don't wrap the fn in try/catch returning a placeholder — that turns infra failures (stale rows, FK violations, rejected writes) into fake successes. Only allowed top-level catch: a fatal handler around \`main()\` that exits non-zero, so callers can tell a whole-replay crash from a clean run with some unreplayable items.
  - **Side-effect check** — if importing the instrumented function triggers module-level side effects (booting listeners/ports/prod connections), the subagent must not work around it silently — return that fact in its report so the main agent can flag it to the user.
  - **Expected return** — one-line confirmation of the script path written/edited, plus any flags worth surfacing to the user.{{/codex}}{{#codex}}**Write instrumentation AND the replay pipeline for this trace function in the same cycle — batched into a single message of tool calls.** Instrumentation edits go in one apply_patch / set of Edit calls; the replay script (new or updated \`scripts/replay.*\`) goes in the same turn. Skip the replay script entirely for Go-only projects (Go does not support replay).

- **11a. Instrumentation edits** — follow the SDK reference exactly, purely additive. Never change behavior, arguments, return values, error handling, variable names, types, control flow, or code structure. Batch repetitive edits in parallel (one message, many Edit calls); for large mechanical fan-outs (>10 files of the same wrapper pattern), validate the pattern on one file, then do the remaining files.

- **11b. Replay pipeline edits** — write or update the replay script (\`scripts/replay.*\` or the project's equivalent) alongside the instrumentation, grounded in the docs you already fetched in step {{step:instrument/read-sdk-reference}}:
  - **Language + SDK replay reference URL** — \`https://docs.bitfab.ai/<language>-sdk#replay\` (TypeScript / Python / Ruby). Re-check the \`#replay\` section now to confirm the current signature — do not write from memory.
  - **Trace function key** — as confirmed in the trace plan.
  - **Trace function root** — the function's real name, full signature (param names + types), return type, absolute file path, and import path the replay script will use.
  - **Replay script target** — if \`scripts/replay.*\` (or the project's equivalent) already exists, add a new pipeline entry for this key. Otherwise create the file.
  - **Non-negotiables**: CLI arg for pipeline name; optional \`--limit N\` (default 10) and \`--trace-ids id1,id2\` flags; replay fn imports and invokes the real function (never a stub); runs in the app's loaded \`.env\` environment (no mocked DB clients / env vars / config / models); mocks only what has no live counterpart at replay time (stream writers, session/request stubs); follows the Replay Output Contract (emit the full \`ReplayResult\` as one JSON block via \`JSON.stringify(result, null, 2)\` / \`json.dumps(result, indent=2, default=str)\` / \`JSON.pretty_generate(result)\`, including every item's \`durationMs\`/\`duration_ms\`, \`tokens\`, and \`model\`; never swap the JSON block for per-field log lines, counts, lengths, hashes, or previews); prints a short human-readable summary + test run URL before the JSON dump; lives under \`scripts/\` (or the project's existing scripts location).
  - **Match the \`#replay\` template's fn signature verbatim — no speculative defense.** The SDK invokes the replay wrapper with captured args in their original shape; don't branch on arg arity/shape, don't add type-checker escape hatches (\`any\` casts, \`cast(Any, ...)\`, ignore comments, untyped passthroughs), and don't guard against cases the contract precludes. If the root signature contradicts what the reference template expects, surface that to the user before writing code; don't paper over it. A hard error at the call site beats silent passthrough of malformed input.
  - **Per-item error tolerance** — \`bitfab.replay\` records thrown wrapped-fn errors in \`item.error\` and keeps going; rely on that. Don't wrap the fn in try/catch returning a placeholder — that turns infra failures (stale rows, FK violations, rejected writes) into fake successes. Only allowed top-level catch: a fatal handler around \`main()\` that exits non-zero, so callers can tell a whole-replay crash from a clean run with some unreplayable items.
  - **Side-effect check** — if importing the instrumented function triggers module-level side effects (booting listeners/ports/prod connections), do not work around it silently — flag it to the user so they can scope the replay env correctly.{{/codex}}

The trace plan's \`Files changed:\` list must include the replay script path for this cycle (new or edited) alongside the instrumented files.`,
                    next: "instrument/tell-user-how-to-run",
                },
                {
                    id: "tell-user-how-to-run",
                    kind: "action",
                    title: "Tell the user how to generate traces and run replay",
                    body: "Tell the user how to run the app to generate the first trace AND, once traces exist, how to run the replay script for this pipeline — give exact command(s) for both. Do NOT run them yourself. (Omit the replay command for Go-only projects.)",
                    next: "instrument/mandatory-stop",
                },
                {
                    id: "mandatory-stop",
                    kind: "branch",
                    toolCalls: ["bash", "ask"],
                    title: "Mandatory stop — pick the next move",
                    body: `**MANDATORY STOP — never silently end the cycle without the A/B/C/D prompt.** {{askVerbCap}} (we recommend **A**: generate traces before instrumenting the next workflows):

{{branches}}

**For option A**, present the script to run to the user (allow them to let you run it for them). Before starting the wait, tell the user verbatim: \`Polling for first trace (up to ~10 min) — press Esc to cancel.\` Then run {{^codex}}with \`Bash\` (timeout: 660000ms){{/codex}}{{#codex}}in a shell (allow up to ~11 min){{/codex}}: \`node "{{pluginRoot}}/dist/commands/waitForTrace.js" <trace-function-key>\`. The command blocks inside Node — polling Bitfab every 10s until a trace lands or the ~10 min timeout fires — so no agent tokens are burned while waiting. When it exits, parse the final stdout line as JSON: \`{"status":"found","traceId":"…","url":"…"}\` → report the trace URL; \`{"status":"timeout",…}\` → note that no trace arrived yet; \`{"status":"interrupted",…}\` → the user cancelled.

A, B, and C all return to step {{step:instrument/present-workflows}} for the selected workflow. Only D exits the Instrument loop.

**After D in \`all\` mode, Replay ALWAYS runs** as a coverage-verification/backfill sweep. Step {{step:instrument/write-instrumentation}} already wrote a replay pipeline for every trace function instrumented in this session, so Replay is usually a no-op that confirms coverage; it still runs to catch any pre-existing trace function keys that don't yet have a pipeline and to verify Replay Output Contract compliance across all pipelines. Replay does not depend on traces existing — replay scripts are built from trace function keys in the instrumented code, not captured trace data. In \`instrument\` mode, D stops after the Instrument loop.`,
                    branches: [
                        {
                            option: {
                                letter: "A",
                                label: "Generate traces [current workflow]",
                            },
                            recommended: true,
                            next: "instrument/present-workflows",
                        },
                        {
                            option: {
                                letter: "B",
                                label: "Instrument [next workflow]",
                                description: "[why it's the next highest value]",
                            },
                            next: "instrument/present-workflows",
                        },
                        {
                            option: {
                                letter: "C",
                                label: "Instrument [other workflow]",
                                description: "[alternative]",
                            },
                            next: "instrument/present-workflows",
                        },
                        {
                            option: {
                                letter: "D",
                                label: "Done instrumenting",
                                description: "proceed to Replay (in `all` mode) / Done (in `instrument` mode)",
                            },
                            next: {
                                byMode: {
                                    all: "replay/gather-keys",
                                    login: null,
                                    "login-headless": null,
                                    instrument: null,
                                    modify: null,
                                    view: null,
                                    replay: null,
                                    templates: null,
                                },
                            },
                        },
                    ],
                },
            ],
        },
        {
            id: "modify",
            title: "Modify",
            stepStyle: "list",
            intro: `Adjust an **existing** trace setup. Requires existing SDK usage in the codebase — if none exists, run Instrument first. Triggered explicitly by \`{{cmd}}setup modify\`, or selected from the {{askNoun}} at Instrument step {{step:instrument/search-existing}} when existing SDK usage is found.

Every Modify cycle targets **exactly one** trace function. Never batch multiple trace functions in one cycle — if the user wants more, loop via the step {{step:modify/mandatory-stop}} menu.`,
            steps: [
                {
                    id: "gather-existing",
                    kind: "action",
                    toolCalls: ["grep", "glob"],
                    title: "Gather existing trace functions",
                    body: "**Gather existing trace functions** by searching for SDK patterns ({{sdkPatterns}}). List each key alongside its root function. If none are found, tell the user Modify needs existing instrumentation and suggest `{{cmd}}setup instrument`.",
                    next: "modify/pick-function",
                },
                {
                    id: "pick-function",
                    kind: "action",
                    toolCalls: ["ask"],
                    title: "Pick exactly ONE trace function",
                    body: "**Pick exactly ONE trace function to modify.** {{askVerbCap}} with the list of existing keys. Recommend the one the user most recently instrumented (or the one most recently referenced in the current session) and explain why in one line.",
                    next: "modify/reconstruct-current-plan",
                },
                {
                    id: "reconstruct-current-plan",
                    kind: "action",
                    toolCalls: ["getTracePlan", "read"],
                    refs: ["reference"],
                    title: "Reconstruct the current trace plan (bootstrap from prior plan when possible)",
                    body: `**Bootstrap the \`before\` \`TracePlanTree\` from the most recent confirmed trace plan for this trace function key**, falling back to reading the code only when no prior plan exists. The plan from the previous Instrument or Modify cycle is the source of truth for what's currently captured — re-deriving from code drops sample inputs/outputs and surrounding-context nodes the user previously confirmed.

1. Call {{tool:getTracePlan}} with \`{ traceFunctionKey: "<chosen key>" }\` (no \`planId\`). Two outcomes:
   - **Prior plan found** — parse the JSON block in the response. Use its \`tree\` as the \`before\` \`TracePlanTree\` and its \`capturedNodeIds\` as the current capture set. You do not need to re-read the instrumented files. Skip step 2.
   - **"No prior confirmed trace plan found"** — there is no plan for this key yet (key created outside the skill, or first Modify cycle that predates this column). Fall through to step 2.
2. **Code-reading fallback.** Read the instrumented files to map the existing span tree into a \`TracePlanTree\` (\`{ rootId, nodes: { [id]: TraceNode } }\`, same shape used in Instrument step {{step:instrument/build-trace-plan}}). Each \`TraceNode\` carries \`id\`, \`name\`, \`kind\` ("manual" | "auto" | "pure"), \`file\`, \`line\`, \`signature\`, \`parentId\`, \`childIds\`, plus \`framework\` for \`[auto]\` lines.

Either way, hold the \`before\` tree in memory — it seeds the \`after\` tree you build in step {{step:modify/build-modified-plan}} and becomes the left-hand side of the inline-fallback diff in step {{step:modify/present-diff}}. Do not present it yet.`,
                    next: "modify/build-modified-plan",
                },
                {
                    id: "build-modified-plan",
                    kind: "reference",
                    refs: ["reference"],
                    toolCalls: ["read", "grep"],
                    title: "Build the modified trace plan (with surrounding context)",
                    body: `**Build the modified trace plan as a \`TracePlanTree\` under the same PURELY ADDITIVE constraint as Instrument step {{step:instrument/build-trace-plan}}.** Start from the \`before\` tree built in step {{step:modify/reconstruct-current-plan}} and produce an \`after\` tree of the same shape (\`{ rootId, nodes: { [id]: TraceNode } }\`) that applies the user's requested modifications. Reuse node ids unchanged for nodes that survive — that lets the trace plan UI show only what actually changes — and mint new ids for added nodes.

**If the user didn't request anything specific** (no modifications were named in the skill invocation or earlier in the conversation), produce an \`after\` tree identical to the \`before\` tree. Don't invent changes. The user will edit the capture set directly in the UI in step {{step:modify/present-diff}}.

The modified tree must be implementable without behavior changes. If a requested modification requires awaiting a stream that wasn't awaited, delaying a call, reordering operations, blocking a callback, or restructuring control flow, tell the user which part doesn't fit and why, and ask them to refine the request (or suggest splitting into multiple cycles). Never present a behavior-changing approach as an option.

**Every captured node MUST include \`sampleInput\` and \`sampleOutput\`** — same hard rule as Instrument step {{step:instrument/build-trace-plan}}. Carry samples forward unchanged for surviving nodes; for newly added nodes (intermediate spans, deeper leaves, a new upstream/downstream root), construct realistic example values from the function's parameter and return types (Read the file and its return-type imports if needed). Do not advance to step {{step:modify/present-diff}} with a captured node missing either field.

**Include surrounding code as \`pure\` context nodes** so the modified capture is legible inside its codebase context and the user can toggle additional nodes into the capture directly in the UI without leaving the page. The test for inclusion is **"would the user plausibly want this as its own span?"** — anything they might promote to a wider root, wrap as a deeper child, or add as a peer at the same depth. Walk in three directions:
- **~10 callers above the root** — candidates for **promoting the root upward** to a wider scope. Walk via Grep (callers of the root, then callers of those, etc.) and attach each as a \`pure\` ancestor. Stop at process entry points (HTTP handlers, queue workers, CLI \`main\`, cron jobs, page handlers, framework boot — there is no useful root above those) or when you've gathered ~10 nodes.
- **~10 callees below each leaf** — candidates for **wrapping deeper spans**. For every existing leaf in the captured sub-tree, walk downward (callees of that leaf, callees of those, etc.) and attach each as a \`pure\` descendant. Include any callee the user might plausibly want as its own span — LLM / tool / agent calls, prompt construction, response parsing, retry loops, fan-outs, post-processing that drives another model. Stop at pure plumbing (pass-through returns, trivial formatting or arithmetic, no further interesting activity) or ~10 nodes per leaf. **Don't stop just because you crossed an SDK / framework / stdlib boundary** — the test is "is this plausibly its own span?", not "is this in our code?".
- **~5 siblings per captured non-root node** — candidates for **peer spans at the same depth**. For each captured non-root node, include the parent's other callees (other functions invoked from the same wrapper) as \`pure\` siblings. These are the nodes the user might wrap alongside the existing capture to widen the trace sideways.

Mark every surrounding node with \`kind: "pure"\` (uncaptured) and **do not** add their ids to \`capturedNodeIds\`. They serve two ends: **legibility** (the captured set sits inside its surrounding code so the user sees what is and isn't traced) and **modification** (they are the levers in the UI for expanding capture deeper, broader, or sideways).

When applying a requested modification, read the relevant signatures so the plan stays accurate: for added context, name the exact keys/values and the span they attach to; for new instrumented spans, read each callee's signature and pick a type annotation (\`function\`, \`llm\`, \`tool\`, \`agent\`, \`handoff\`); for span removals, list each by name and confirm the underlying call is left untouched; for a new upstream/downstream root, read the new function's signature and confirm it still covers the interesting LLM/tool activity (upstream) or remains a common ancestor of every LLM/tool span (downstream).`,
                    next: "modify/present-diff",
                },
                {
                    id: "present-diff",
                    kind: "branch",
                    toolCalls: ["ask", "bash", "createTracePlan", "getTracePlan"],
                    refs: ["reference"],
                    title: "Open the modified plan in the trace plan UI; confirm = apply, cancel = ask what to change",
                    body: `**Send the modified plan straight to the trace plan UI — it is the user's primary surface for confirming or editing the change**, not the inline before/after diff. The user can adjust the captured set directly in the UI (selecting/deselecting any of the surrounding \`pure\` context nodes added in step {{step:modify/build-modified-plan}}). Confirm in the UI = apply the diff. Cancel = ask the user what they want to change. Same delivery pattern as Instrument step {{step:instrument/build-trace-plan}}.

1. **Post the modified plan and open the UI.** Call {{tool:createTracePlan}} with \`{ language, tree, capturedNodeIds, traceFunctionKey }\` (and \`stats\` if you have a sample run from the existing trace function):
   - \`tree\` — the modified \`after\` \`TracePlanTree\` from step {{step:modify/build-modified-plan}}, with the ~10 surrounding callers / ~10 surrounding callees included as \`pure\` context nodes.
   - \`capturedNodeIds\` — your initial recommendation. Must form a connected sub-tree (selecting any descendant implies its ancestors). Surrounding \`pure\` context nodes are not included.
   - \`traceFunctionKey\` — the existing key from step {{step:modify/pick-function}}. Persisting it lets the next Modify cycle bootstrap from this plan.

   The tool returns a plan id (and a \`https://bitfab.ai/trace-plan/<id>\` URL).

2. **Open the trace plan in the browser** by running:

\`\`\`bash
node "{{pluginRoot}}/dist/commands/openTracePlan.js" <planId>
\`\`\`

(\`{{pluginRoot}}\` resolves to the plugin directory; \`<planId>\` is the id returned by {{tool:createTracePlan}}.) The script opens the trace plan page and **blocks** until the user clicks **Confirm** or **Cancel** in the browser{{#codex}} (up to 30 minutes){{/codex}}.{{#codex}} The script prints a handoff URL (with \`callbackPort=\` and \`ticket=\` query params) — surface it to the user verbatim so they can paste it into a browser if the auto-launch didn't open one.

**Polling (mandatory — see the Blocking-process rule at the top of this skill):** after surfacing the URL, keep polling the live exec session until the process exits. Do NOT stop after printing the URL. Do NOT wait for a chat message from the user — their confirmation arrives as stdout on the already-running process, not as a new prompt. Poll every few seconds until one of these terminal conditions, then route accordingly:{{/codex}}

3. **On exit, route by the final stdout line:**
   - \`Trace plan confirmed [via …]\` — call {{tool:getTracePlan}} with \`{ planId }\` to read the authoritative \`capturedNodeIds\` (the user may have toggled \`pure\` context nodes into the captured set or removed previously-captured nodes in the UI). Reconcile your edit plan with what's now in \`capturedNodeIds\` — drop manual \`●\` wraps no longer captured, add wraps for any newly captured nodes — then take branch **A** (Proceed).
   - \`Trace plan cancelled [via …]\` — the user cancelled from the browser. Take branch **C** (Modifications) — {{askVerb}}: what do they want to change? Their answer feeds back into step {{step:modify/build-modified-plan}}.
   - non-zero exit / timeout — surface the error to the user, then fall back to the inline AskUserQuestion below.

**Inline fallback** (use only if {{tool:createTracePlan}} errors, e.g. offline or MCP unreachable, or \`openTracePlan.js\` exits non-zero): present an inline before/after diff using the Default view template from the **Trace Plan Format** reference section, list \`Files changed:\` (paths only, no annotations), and **STOP** — {{askVerb}}:

{{branches}}`,
                    branches: [
                        {
                            option: {
                                letter: "A",
                                label: "Proceed",
                                description: "apply the diff using the confirmed capture set",
                            },
                            recommended: true,
                            next: "modify/apply-changes",
                        },
                        {
                            option: {
                                letter: "B",
                                label: "Expand details",
                                description: "re-render the inline diff in the expanded view (fallback only)",
                            },
                            next: "modify/present-diff",
                        },
                        {
                            option: {
                                letter: "C",
                                label: "Modifications",
                                description: "ask what the user wants to change, then return to building the modified plan",
                            },
                            next: "modify/build-modified-plan",
                        },
                        {
                            option: {
                                letter: "D",
                                label: "Abort entirely",
                                description: "drop this cycle without writing edits",
                            },
                            next: null,
                        },
                    ],
                },
                {
                    id: "apply-changes",
                    kind: "action",
                    toolCalls: ["edit"],
                    title: "Apply the changes (purely additive)",
                    body: "**Apply the changes — purely additive to behavior.** Same rules as Instrument step {{step:instrument/write-instrumentation}}: never change arguments, return values, error handling, variable names, types, control flow, or code structure. Removing a `withSpan`/`@span` wrapper is the only structural edit allowed, and only when it leaves the wrapped call, its arguments, and its return value untouched. The trace function key from step {{step:modify/pick-function}} stays the same — do not rename keys. Batch repetitive edits in parallel (one message, many Edit calls).",
                    next: "modify/mandatory-stop",
                },
                {
                    id: "mandatory-stop",
                    kind: "branch",
                    toolCalls: ["bash", "ask"],
                    title: "Mandatory stop — pick the next move",
                    body: `Tell the user how to run the app to generate a trace with the modified setup — exact command(s). Do NOT run it yourself. Then **MANDATORY STOP** — {{askVerb}}:
> We recommend **A**: generate a trace with the modified setup so the diff is observable end-to-end.

{{branches}}

B returns to step {{step:modify/pick-function}}. A and C exit the Modify loop. After exit, stop (Modify does not auto-continue to Replay — the user can invoke \`{{cmd}}setup replay\` separately).`,
                    branches: [
                        {
                            option: {
                                letter: "A",
                                label: "Generate a trace for the modified setup",
                                description: "present the script to run; allow the user to let you run it",
                            },
                            recommended: true,
                            next: null,
                        },
                        {
                            option: {
                                letter: "B",
                                label: "Modify another trace function",
                                description: "returns to step 2",
                            },
                            next: "modify/pick-function",
                        },
                        {
                            option: {
                                letter: "C",
                                label: "Done",
                                description: "stop here",
                            },
                            next: null,
                        },
                    ],
                },
            ],
        },
        {
            id: "view",
            title: "View",
            stepStyle: "list",
            intro: `Open the trace planner UI for an **existing** trace function — read-only. Triggered explicitly by \`{{cmd}}setup view\`. Useful for inspecting what's currently captured (tree shape, captured node ids, sample inputs/outputs) without making any code edits.

Every View invocation targets **exactly one** trace function. The browser UI's Confirm/Cancel controls have no effect here — the user is just looking at the plan.`,
            steps: [
                {
                    id: "gather-existing",
                    kind: "action",
                    toolCalls: ["grep", "glob"],
                    title: "Gather existing trace functions",
                    body: "**Gather existing trace functions** by searching for SDK patterns ({{sdkPatterns}}). List each key alongside its root function. If none are found, tell the user View needs existing instrumentation and suggest `{{cmd}}setup instrument`.",
                    next: "view/pick-function",
                },
                {
                    id: "pick-function",
                    kind: "action",
                    toolCalls: ["ask"],
                    title: "Pick exactly ONE trace function",
                    body: "**Pick exactly ONE trace function to view.** {{askVerbCap}} with the list of existing keys. Recommend the one the user most recently instrumented (or the one most recently referenced in the current session) and explain why in one line.",
                    next: "view/fetch-plan",
                },
                {
                    id: "fetch-plan",
                    kind: "branch",
                    toolCalls: ["getTracePlan"],
                    title: "Fetch the latest confirmed trace plan",
                    body: `Call {{tool:getTracePlan}} with \`{ traceFunctionKey: "<chosen key>" }\` (no \`planId\`). Two outcomes:

- **Prior plan found** — parse the response for the \`Plan id:\` line and hold that id for the next step. Take branch **A** (Open).
- **"No prior confirmed trace plan found"** — there is no plan to view (key created outside the skill, never confirmed, or never instrumented via this skill). Tell the user there's nothing to view yet and suggest \`{{cmd}}setup modify\` to build and confirm a plan for this key. Take branch **B** (Stop).`,
                    branches: [
                        {
                            when: "prior plan found",
                            option: {
                                letter: "A",
                                label: "Open",
                                description: "open the trace plan in the browser using the returned plan id",
                            },
                            recommended: true,
                            next: "view/open-in-ui",
                        },
                        {
                            when: "no prior plan",
                            option: {
                                letter: "B",
                                label: "Stop",
                                description: "no plan exists for this key; suggest `setup modify` and exit",
                            },
                            next: null,
                        },
                    ],
                },
                {
                    id: "open-in-ui",
                    kind: "action",
                    toolCalls: ["bash"],
                    title: "Open the trace plan in the browser",
                    body: `Open the trace plan in the browser by running:

\`\`\`bash
node "{{pluginRoot}}/dist/commands/openTracePlan.js" <planId>
\`\`\`

(\`{{pluginRoot}}\` resolves to the plugin directory; \`<planId>\` is the id parsed from step {{step:view/fetch-plan}}.) The script opens the trace plan page and **blocks** until the user closes the browser tab or clicks Confirm/Cancel{{#codex}} (up to 30 minutes){{/codex}}. View is read-only — whichever button the user clicks, do **not** apply edits or call {{tool:getTracePlan}} again. When the process exits, report that the plan was viewed and stop.{{#codex}}

The script prints a handoff URL (with \`callbackPort=\` and \`ticket=\` query params) — surface it to the user verbatim so they can paste it into a browser if the auto-launch didn't open one.

**Polling (mandatory — see the Blocking-process rule at the top of this skill):** after surfacing the URL, keep polling the live exec session until the process exits. Do NOT stop after printing the URL. Do NOT wait for a chat message from the user — their dismissal arrives as stdout on the already-running process, not as a new prompt.{{/codex}}`,
                    next: null,
                },
            ],
        },
        {
            id: "replay",
            title: "Replay",
            stepStyle: "list",
            intro: `Create or update replay scripts for instrumented trace functions. Requires instrumentation in the codebase; does **not** require existing traces — replay scripts are created from trace function keys in the code, not captured trace data.

Replay scripts let the team regression-test any trace function against production data with one command — they fetch historical traces, re-run them through the current code, and report old vs. new outputs side-by-side. Note: **Go does not support replay** — skip this phase if the project is Go-only.

**Relationship to Instrument.** When Replay runs via \`all\` mode or directly after Instrument, most (often all) trace function keys already have pipelines because Instrument step {{step:instrument/write-instrumentation}} writes them {{parallelism}} the instrumentation edits. This phase is then a coverage + contract-compliance sweep. Run it standalone (\`{{cmd}}setup replay\`) to catch pre-existing trace function keys that predate the parallel-write step or were added outside the skill.

**Source of truth:** two pages — read both before creating or modifying a replay script. Do not improvise from memory.
- **Canonical \`replay\` API signature, options, and return shape:** \`/reference/typescript#replay\`, \`/reference/python#replay\`, \`/reference/ruby#replay\` (Go has no replay). Use this for the exact field names (\`result\` / \`originalOutput\` vs \`original_output\`), default \`limit\`, \`maxConcurrency\`/\`max_concurrency\`, error behavior.
- **Copy-pasteable script template + replay output contract + input serialization caveat:** \`/typescript-sdk#replay\`, \`/python-sdk#replay\`, \`/ruby-sdk#replay\`. Use this for the \`scripts/replay.<ext>\` shape and the rules for what to print to stdout.`,
            steps: [
                {
                    id: "gather-keys",
                    kind: "action",
                    toolCalls: ["grep", "glob"],
                    title: "Gather all trace function keys",
                    body: "**Gather all trace function keys** by searching for SDK patterns ({{sdkPatterns}}). This is the source of truth for what replay must cover.",
                    next: "replay/find-existing-scripts",
                },
                {
                    id: "find-existing-scripts",
                    kind: "action",
                    toolCalls: ["glob", "grep"],
                    title: "Search for existing replay scripts",
                    body: "**Search for existing replay scripts** — files matching `scripts/replay.*`, `scripts/*replay*`, or any file importing/calling the SDK's replay API.",
                    next: "replay/compare-coverage",
                },
                {
                    id: "compare-coverage",
                    kind: "action",
                    toolCalls: ["read"],
                    title: "Compare coverage and act",
                    body: `**Compare coverage.** Replay is non-interactive once entered — do not ask the user whether to create or add scripts:
- If replay scripts exist and cover all keys: verify each one already conforms to the Replay Output Contract in the docs (emits the full \`ReplayResult\` as one JSON block, including every item's \`durationMs\`/\`duration_ms\`, \`tokens\`, and \`model\`, never just counts or per-field log lines). If any don't, fix them; otherwise report up to date and stop.
- If replay scripts exist but are missing trace function keys: add the missing scripts in step {{step:replay/create-scripts}}.
- If no replay scripts exist: create them in step {{step:replay/create-scripts}}.`,
                    next: "replay/create-scripts",
                },
                {
                    id: "create-scripts",
                    kind: "action",
                    toolCalls: ["webFetch", "write", "edit"],
                    title: "Create or update replay scripts",
                    body: `**Create the replay script** following the example in the SDK reference's Replay section (\`https://docs.bitfab.ai/<language>-sdk#replay\`), adapted to this codebase. The non-negotiables (enforced by the docs page, repeated here so the script review catches them):
- **Ground the script in the docs, not memory.** Before writing the replay call, fetch \`https://docs.bitfab.ai/reference/<language>#replay\` for the canonical signature and return shape, then \`https://docs.bitfab.ai/<language>-sdk#replay\` for the script template and output contract. Quote the exact function signature + return-shape fields verbatim in your plan. Field names differ per language (Python: \`result\`, \`original_output\`; TypeScript: \`result\`, \`originalOutput\`; Ruby: \`:result\`, \`:original_output\`) — do not paraphrase or invent names like \`new_output\`/\`trace_id\`.
- **Pass the decorated function itself, not an undecorated wrapper.** The trace function key is read from the decorator/attribute on the function you pass in. For Python class methods, pass \`Class.method\` (or a bound \`instance.method\`). For TypeScript, the key is passed as a string arg alongside the function — use the exact key from the instrumented code. For Ruby, pass \`receiver\` + \`method_name:\` + \`trace_function_key:\` matching the \`traceable\` decoration.
- **Use the same \`Bitfab\` client across instrumentation and replay.** Import it from the instrumented module (or a shared singleton) — never construct a second client inside the replay script, or registered trace functions won't resolve.
- Accept a pipeline name as a CLI argument
- Accept optional \`--limit N\` (default 10) and \`--trace-ids id1,id2\` flags
- Map pipeline names to trace function keys and their replay functions
- **Each pipeline's replay function MUST import and call the actual instrumented function** — never a stub or identity function. If the function signature doesn't match the raw input shape, reshape arguments in the wrapper.
- **Replay runs in the app's environment.** The script imports the app as a library — DB clients, env vars, config loaders, and model IDs resolve from the loaded environment. Do **not** mock them. Run the script with \`.env\` loaded (e.g. \`pnpm with-env tsx scripts/replay.ts\`, \`dotenv run -- python scripts/replay.py\`, or the project's equivalent) so the app's normal bootstrap applies.
- **Only mock what has no live counterpart at replay time.** For factory-created instrumented functions (taking session, stream writers via closure), the wrapper passes:
  - Stream/socket writers: no-op (\`{ write: () => {}, merge: () => {} }\`) — no client on the other end
  - Session/request identifiers: minimal stub with the fields the function reads
- **Caveat: watch for module-level import side effects.** Importing the instrumented function transitively runs the app's module initialization — if that opens listeners, binds ports, or connects to prod, the replay script inherits it. When in doubt, confirm the replay env points at a staging/local DB before running.
- **Follow the docs' Replay Output Contract**: capture the full \`ReplayResult\` (items + \`testRunId\` + \`testRunUrl\`, including \`durationMs\`/\`duration_ms\`, \`tokens\`, and \`model\` per item) into one variable and emit it as a single JSON object to stdout via \`JSON.stringify(result, null, 2)\` (TS), \`json.dumps(result, indent=2, default=str)\` (Python), or \`JSON.pretty_generate(result)\` (Ruby). A subagent reading the output must be able to \`JSON.parse\` / \`json.loads\` one contiguous block — do not replace the JSON dump with per-field log lines, counts, lengths, hashes, or previews. Writing the same JSON to \`scripts/replay-result.json\` in parallel is optional but encouraged.
- Print a short human-readable summary (total replayed, same, changed, errors) and the test run URL ahead of the JSON dump
- Live in a \`scripts/\` directory (or the project's existing scripts location)`,
                    next: "replay/safety-net",
                },
                {
                    id: "safety-net",
                    kind: "branch",
                    toolCalls: ["ask", "edit"],
                    refs: ["refactor-confirmation"],
                    title: "Safety net for legacy instrumentation",
                    body: `**Safety net for legacy instrumentation.** If an already-instrumented function (introduced before step {{step:instrument/explain-trace-boundary}}'s serializability gate, or via another path) can't be invoked from the replay script — most commonly because it isn't exported, is defined inline in a route handler, or takes unserializable inputs — {{askVerb}} offering step {{step:instrument/explain-trace-boundary}}'s two resolutions:

{{branches}}

Reason from the function's signature and visibility; do not execute the script to detect this. **If the user picks "Refactor" (or a boundary move that requires rewriting callers), apply the "Refactor confirmation" rule below — present a refactor plan labeled as *visibility* or *structural* and get a second confirmation before modifying code.**`,
                    branches: [
                        {
                            option: {
                                letter: "A",
                                label: "Move trace boundary inward",
                            },
                            next: null,
                        },
                        {
                            option: { letter: "B", label: "Refactor" },
                            recommended: true,
                            next: null,
                        },
                        {
                            option: {
                                letter: "C",
                                label: "Leave as-is",
                                description: "add a header comment noting why the function isn't callable and flag that the script will rot",
                            },
                            next: null,
                        },
                    ],
                },
            ],
        },
        {
            id: "templates",
            title: "Templates",
            stepStyle: "list",
            intro: `Iterate on the **span-rendering templates** for one trace function. Each round: the user describes what should look different, you call {{tool:getTemplate}} → edit → {{tool:updateTemplate}} **with \`traceFunctionKey\` set to the picked key**, and the user refreshes the chromeless template-preview page to see the change rendered against a real trace. Loop until the user is satisfied. Triggered explicitly by \`{{cmd}}setup templates [<key>]\` — never reached from \`all\`.

Templates control how a span's input / output renders in the Bitfab UI. They are scoped per **span type** (\`llm\`, \`agent\`, \`function\`, \`guardrail\`, \`handoff\`, \`custom\`). This phase **always passes \`traceFunctionKey\`** so edits become **per-function overrides**: they apply only to spans on traces of the picked function, not to other functions in the org. Resolution at render time is per-key row → org-global → file default, so the seed you see in {{tool:getTemplate}} reflects whatever is currently rendering for this function. Surface this scope when the user asks for a change so they know nothing else in the org is affected.`,
            steps: [
                {
                    id: "pick-function",
                    kind: "branch",
                    toolCalls: ["ask", "listFunctions", "grep", "glob"],
                    title: "Resolve the trace function key",
                    body: `If the user passed a key as the argument, use it directly and continue.

Otherwise, follow the same picker pattern as \`{{cmd}}assistant\`:

1. Call {{tool:listFunctions}} to enumerate the org's traced functions. The tool returns flat \`FUNCTION: <key>\` lines; work from those keys directly. Use **only** the keys returned: do NOT invent or infer descriptions of what each function does from its name. Key names are often ambiguous, and guessing produces hallucinated summaries that confuse the user.
2. Grep this repo for each key in parallel (across \`*.ts\`, \`*.tsx\`, \`*.py\`, \`*.rb\`, \`*.go\`, \`*.baml\`) so you know which keys are instrumented here. Mark each as ✅ instrumented here (with file path) or ⚠️ not found in this repo.
3. Present a compact list in the question text showing only: \`<key>\` · \`<repo marker + path>\`. No invented summaries.
4. {{askVerbCap}} with 2 options: the recommended function (prefer ✅ instrumented here, and matching session context when one is clearly relevant) and a free-text "Type a function key" option. If nothing is instrumented in this repo, say so explicitly in the question, don't hide it.

{{whens}}`,
                    branches: [
                        {
                            when: "argument supplied",
                            description: "use it as the trace function key and continue",
                            next: "templates/load-reference",
                        },
                        {
                            when: "no argument",
                            description: "list trace functions, ask the user, then continue with the chosen key",
                            next: "templates/load-reference",
                        },
                    ],
                },
                {
                    id: "load-reference",
                    kind: "action",
                    toolCalls: ["getTemplateReference"],
                    title: "Load the template reference",
                    body: `Call {{tool:getTemplateReference}} **once** before any edit. It returns a stable agent-facing schema for Bitfab span templates: the rendering engine (Nunjucks, Jinja2-compatible), the render-context shape (top-level keys, \`SpanData\` / \`ParsedSpanData\`), the registered custom filters and tests, common patterns from the live default templates, and error-fallback behavior. Without this you cannot write a correct edit; references to undeclared variables silently render empty in production.

Hold the reference in your working context for the rest of the loop. Do NOT call it again on subsequent edits.`,
                    next: "templates/ground-in-code",
                },
                {
                    id: "ground-in-code",
                    kind: "action",
                    toolCalls: ["grep", "glob", "read"],
                    title: "Ground edits in the user's code",
                    body: `Before opening the preview, grep the codebase for the trace function key (\`<key>\`) so you can see what the function actually does. The user's "change" requests are usually about surfacing something domain-specific (an input field, a tool name, a context label), and knowing the function helps you map the request to the right span type and the right field path. If grep returns nothing (the function has been renamed or the user is operating on traces from a different repo), continue without it.`,
                    next: "templates/check-traces",
                },
                {
                    id: "check-traces",
                    kind: "branch",
                    toolCalls: ["searchTraces"],
                    title: "Verify a trace exists for the function",
                    body: `The preview page renders the most recent trace for the function. Without at least one trace it has nothing to render, so check before opening it.

Call {{tool:searchTraces}} with \`{ traceFunctionKey: "<key>", limit: 1 }\`. If the response contains a trace ID, continue. If the response indicates no traces exist (e.g. \`No traces found matching the filter criteria.\`), exit and tell the user in one short line: \`No traces yet for <key>. Run your app (or the replay script) to generate one, then re-run \\\`{{cmd}}setup templates <key>\\\` to preview.\` Do NOT block waiting; the user re-invokes when they have a trace.

{{whens}}`,
                    branches: [
                        {
                            when: "trace exists",
                            description: "continue and open the preview",
                            next: "templates/open-preview",
                        },
                        {
                            when: "no traces yet for this function",
                            description: "exit and tell the user to generate a trace and re-run",
                            next: null,
                        },
                    ],
                },
                {
                    id: "open-preview",
                    kind: "action",
                    toolCalls: ["bash"],
                    title: "Open the chromeless template-preview page (background)",
                    body: `Launch the preview command **in the background** so the agent can keep iterating while the page stays open:

\`\`\`bash
node "{{pluginRoot}}/dist/commands/startTemplatePreview.js" <functionKey>
\`\`\`

{{#claude}}Run this with \`run_in_background: true\` on the Bash tool. The harness returns a task id and an output file path, and will deliver a \`<task-notification>\` with \`status: completed\` automatically when the process exits. Capture both: you'll need the output file path to poll between edit rounds.{{/claude}}{{#cursor}}Run this as a background process and capture the handle plus its stdout so you can poll its status between edit rounds.{{/cursor}}{{#codex}}Run this in a long-lived exec session (\`exec_command\` with \`is_background: true\`); capture the session id so you can \`read\` it between edit rounds to check whether it has exited.{{/codex}}

The command **blocks until the user clicks the Close button on the page**, then exits 0 with a single line like \`Template preview closed [via loopback]\`. If the user instead just closes the browser tab without clicking Close, the process keeps running until the 30-minute timeout. The page auto-redirects to the most recent trace for the function and renders it with the org's current templates; it subscribes to SSE \`template:updated\` events and re-renders the affected span automatically, so the user does NOT need to refresh after each edit.

🚨 **Stdout is a mixed JSONL + free-form stream.** Two event shapes flow over the same channel as the user interacts with the live preview:

\`\`\`json
{"event":"click","ts":"...","traceId":"...","spanId":"...","spanType":"...","sectionPath":"metadata","fieldPath":"metadata.tokens","rawText":"1234","selector":"..."}
{"event":"focus","ts":"...","traceId":"...","spanId":"...","viewMode":"span","expandedSections":["metadata"]}
\`\`\`

\`click\` events fire when the user clicks a decorated element. \`focus\` events fire on initial load, on every span/trace selection change, and on shadow-root \`<details>\` open / close — so you always know the starting viewport even before any click.

Free-form text (browser-handoff status lines, errors) goes through the same stdout. **You MUST filter to lines that parse as JSON before routing.** Skip anything that doesn't parse — never error out on non-JSON lines. The click event payload follows the template-anchor catalog returned by {{tool:getTemplateReference}}; \`fieldPath\` matches a row there, \`sectionPath\` matches a section id. Unknown anchor values are omitted (the click handler drops them); \`rawText\` and \`selector\` are always present so you can disambiguate. Focus event fields are always present; \`spanId\` is null when the user is on the trace overview, \`viewMode\` is \`"trace"\` or \`"span"\`, and \`expandedSections\` lists the \`data-section\` ids whose \`<details>\` is currently open.`,
                    next: "templates/edit-loop",
                },
                {
                    id: "edit-loop",
                    kind: "branch",
                    toolCalls: ["ask", "bash", "getTemplate", "updateTemplate"],
                    title: "Edit loop: change → render → confirm (poll for Close)",
                    body: `Each round of the loop. **Every {{tool:getTemplate}} and {{tool:updateTemplate}} call must include \`traceFunctionKey: <key>\`** (the key picked in step {{step:templates/pick-function}}); without it you'd edit the org-global instead of this function's override.

1. **Tail the background process's stdout** for any \`{"event":"click",...}\` or \`{"event":"focus",...}\` JSON lines that arrived since the previous round. Parse each line; skip non-JSON status lines.
   - **Most recent click** (if any) is ground truth for "what the user is referring to": its \`spanType\` is the template to edit, \`sectionPath\` + \`fieldPath\` (against the anchor catalog from {{tool:getTemplateReference}}) tell you which region to change. If \`fieldPath\` is absent, fall back to \`sectionPath\` + \`rawText\`.
   - **Most recent focus** tells you what the user is currently looking at, even without a click. Use it to anchor a question when the user's instruction is ambiguous (e.g. "make this less verbose" while their focus is on a specific span) and to pick the span type when no click is available. Focus is also helpful to confirm in your acknowledgement that you're editing the same span the user is viewing.
   - If neither signal is present since the last round, fall through to step 2 and ask normally.
2. Ask {{askUser}}: **"Tell me how you want your trace data to look and I'll make the changes in Bitfab. You'll see the changes update live in the Bitfab Studio (the browser tab opened from here)."** **If there was a click in the previous round, anchor the question to it** by prepending a one-line acknowledgement (e.g. "You clicked the tokens value in metadata."). Keep the framing open-ended — do NOT list the six span types up front; let the user describe what they want and pick the span type from their answer. If the user names one of the six span types (\`llm\`, \`agent\`, \`function\`, \`guardrail\`, \`handoff\`, \`custom\`), use that. If their answer is unambiguous about the rendered region but doesn't name a span type AND there was no click, fall back to {{askUser}}which of the six span templates they want to edit. Don't guess the span type from a description like "make this less verbose," since the same description fits multiple templates.
3. Call {{tool:getTemplate}} with \`spanType\` and \`traceFunctionKey: <key>\` to read the **live** content. The response labels its source: \`scoped to traceFunctionKey "<key>"\` (a per-key row already exists), \`org-global override\` (no per-key row yet — this is your seed for the first save), or \`source: file <name>\` (no DB rows at all). **Always** read before write: the prior round may have edited the same template, and overwriting blindly drops that work.
4. Edit the returned source in-context — **one focused change per round**. Resist the urge to bundle multiple unrelated tweaks into a single save: small steps let the user see each effect land on the preview and redirect mid-loop if the change isn't quite right. Stay inside the documented Nunjucks variables and filters (per the reference). Don't introduce \`{% extends %}\`; the assembler injects into \`base.njk\`'s content block, so extends will break composition. When adding new visible regions, **decorate them with the catalog anchors** (\`data-section\`, \`data-field-path\`, \`data-iter-index\`) so future clicks resolve cleanly.
5. Call {{tool:updateTemplate}} with \`spanType\`, \`traceFunctionKey: <key>\`, and the full edited body. The tool upserts the per-function row in place (no version bump, no row juggling). On the first save for a span type the row is created; subsequent edits update it. The browser shows a brief "Editing..." status banner while the call is in flight, then a "Saved" flash when it returns — no extra signaling needed from your side.
6. Acknowledge the save in one short line (e.g. "Saved."). The preview page subscribes to SSE \`template:updated\` events and re-renders automatically — do NOT tell the user to refresh. Do not paste the template body back into chat. After a non-trivial change you may briefly ask {{askUser}} whether the result looks right before starting the next round; for obvious tweaks (a label rename, a colour swap), skip the check and proceed.

Before asking the user about another change, **check whether the background process from step {{step:templates/open-preview}} has exited**. The terminal signal is a line containing \`Template preview closed\` on stdout (the process exits 0 right after).

{{#claude}}Two equivalent ways to detect it: (a) if you've already received a \`<task-notification>\` for the captured task id with \`status: completed\`, the user has clicked Close; (b) otherwise, use the \`Read\` tool on the captured output file path and look for the \`Template preview closed\` line. Either signal means the loop should exit. **Use the same \`Read\` call to also harvest any new \`{"event":"click",...}\` and \`{"event":"focus",...}\` JSON lines for step 1 of the next round.**{{/claude}}{{#cursor}}Read the background process's stdout; the \`Template preview closed\` line means the user clicked Close and the process has exited. **Use the same read to harvest any new \`{"event":"click",...}\` and \`{"event":"focus",...}\` JSON lines for step 1 of the next round.**{{/cursor}}{{#codex}}Use \`read\` against the captured exec session; the \`Template preview closed\` line on stdout means the user clicked Close and the session has exited. **Use the same read to harvest any new \`{"event":"click",...}\` and \`{"event":"focus",...}\` JSON lines for step 1 of the next round.**{{/codex}}

Two ways the loop ends:

{{whens}}`,
                    branches: [
                        {
                            when: "background process exited (user clicked Close)",
                            description: "exit the loop and acknowledge that template editing is done",
                            next: null,
                        },
                        {
                            when: "user explicitly says they're done",
                            description: "exit the loop and acknowledge",
                            next: null,
                        },
                        {
                            when: "user wants another change",
                            description: "loop back and apply the next edit",
                            next: "templates/edit-loop",
                        },
                    ],
                },
            ],
        },
    ],
});
