// bitfab-plugin-lib/dist/flows/targets.js
const renderClaudeFrontmatter = ({ description, argumentHint, allowedTools, }) => {
    const lines = ["---", `description: ${description}`];
    if (argumentHint !== undefined) {
        lines.push(`argument-hint: ${argumentHint}`);
    }
    if (allowedTools !== undefined) {
        const formatted = `[${allowedTools.map((t) => JSON.stringify(t)).join(", ")}]`;
        lines.push(`allowed-tools: ${formatted}`);
    }
    lines.push("---");
    return lines.join("\n");
};
const renderCursorFrontmatter = ({ flowId, description, argumentHint, }) => {
    const usage = argumentHint !== undefined
        ? `. Usage: /bitfab-${flowId} ${argumentHint}`
        : "";
    return [
        "---",
        `name: bitfab-${flowId}`,
        `description: "${description}${usage}"`,
        "---",
    ].join("\n");
};
const renderCodexFrontmatter = ({ flowId, description, argumentHint, }) => {
    const usage = argumentHint !== undefined
        ? `. Invoke with $bitfab:${flowId} ${argumentHint}.`
        : "";
    return [
        "---",
        `name: ${flowId}`,
        `description: "${description}${usage}"`,
        "---",
    ].join("\n");
};
export const TARGETS = {
    claude: {
        name: "claude",
        tokens: {
            cmd: "/bitfab:",
            editor: "Claude Code",
            // biome-ignore lint/suspicious/noTemplateCurlyInString: shell variable reference rendered into skill markdown, not a JS template
            pluginRoot: "${CLAUDE_PLUGIN_ROOT}",
            askUser: "with `AskUserQuestion` ",
            askVerb: "use `AskUserQuestion`",
            askVerbCap: "Use `AskUserQuestion`",
            askNoun: "AskUserQuestion",
        },
        mcpPrefix: "mcp__plugin_bitfab_Bitfab__",
        frontmatter: renderClaudeFrontmatter,
    },
    cursor: {
        name: "cursor",
        tokens: {
            cmd: "/bitfab-",
            editor: "Cursor",
            // biome-ignore lint/suspicious/noTemplateCurlyInString: shell variable reference rendered into skill markdown, not a JS template
            pluginRoot: "${CURSOR_PLUGIN_ROOT:-${CLAUDE_PLUGIN_ROOT}}",
            askUser: "with `AskUserQuestion` ",
            askVerb: "use `AskUserQuestion`",
            askVerbCap: "Use `AskUserQuestion`",
            askNoun: "AskUserQuestion",
        },
        mcpPrefix: "mcp__Bitfab__",
        frontmatter: renderCursorFrontmatter,
    },
    codex: {
        name: "codex",
        tokens: {
            cmd: "$bitfab:",
            editor: "Codex",
            // biome-ignore lint/suspicious/noTemplateCurlyInString: shell variable reference rendered into skill markdown, not a JS template
            pluginRoot: "${BITFAB_PLUGIN_DIR}",
            askUser: "",
            askVerb: "ask the user",
            askVerbCap: "Ask the user",
            askNoun: "choice prompt",
        },
        mcpPrefix: "mcp__Bitfab__",
        frontmatter: renderCodexFrontmatter,
    },
};
