// bitfab-plugin-lib/dist/flows/templates.js
import { Flow } from "./schema.js";
import { CODEX_RESOLVE_PLUGIN_DIR_BASH } from "./shared.js";
export const templatesFlow = Flow.parse({
    schemaVersion: 1,
    id: "templates",
    title: "Bitfab Templates",
    frontmatter: {
        description: {
            claude: "Iterate on Bitfab span-rendering templates against a live trace, with the agent editing via MCP and the user previewing in the browser",
            cursor: "Iterate on Bitfab span-rendering templates against a live trace",
            codex: "Iterate on Bitfab span-rendering templates. Use when the user wants to change how a traced function's spans are rendered in the Bitfab UI. The agent edits the org's global templates via MCP while the user previews against a real trace in the browser",
        },
        argumentHint: "[<trace-function-key>]",
    },
    tools: {
        bash: { kind: "builtin", name: "Bash" },
        grep: { kind: "builtin", name: "Grep" },
        glob: { kind: "builtin", name: "Glob" },
        read: { kind: "builtin", name: "Read" },
        ask: { kind: "builtin", name: "AskUserQuestion" },
        listFunctions: { kind: "mcp", name: "list_trace_functions" },
        searchTraces: { kind: "mcp", name: "search_traces" },
        getTemplateReference: { kind: "mcp", name: "get_template_reference" },
        getTemplate: { kind: "mcp", name: "get_template" },
        updateTemplate: { kind: "mcp", name: "update_template" },
    },
    intro: `Open the chromeless **template-preview** page for one trace function and iterate on the org's global span-rendering templates with the user. Each round: the user describes what should look different, you call {{tool:getTemplate}} → edit → {{tool:updateTemplate}}, and the user refreshes the preview to see the change rendered against a real trace. Loop until the user is satisfied.

**MCP tools:** This skill uses \`list_trace_functions\`, \`search_traces\`, \`get_template_reference\`, \`get_template\`, and \`update_template\` from the **local plugin MCP server** (bundled with this plugin){{#claude}}. Do NOT use the remote Bitfab MCP tools (\`mcp__Simforge__*\` or \`mcp__Bitfab__*\`); use only the \`mcp__plugin_bitfab_Bitfab__*\` variants.{{/claude}}{{^claude}}, exposed under the \`mcp__Bitfab__*\` prefix.{{/claude}}

Templates control how a span's input / output renders in the Bitfab UI. They are scoped per **span type** (\`llm\`, \`agent\`, \`function\`, \`guardrail\`, \`handoff\`, \`custom\`) and apply across the org. Editing one template affects every trace that contains a span of that type, so surface this when the user asks for a change that's narrower than "change all llm spans look like X."

| Invocation | Action |
|---|---|
| \`{{cmd}}templates <key>\` | Open the preview for the given trace function key, then iterate |
| \`{{cmd}}templates\` | List trace functions, ask which one to preview, then iterate |`,
    entries: {
        default: {
            default: "main/pick-function",
            codex: "main/resolve-plugin-dir",
        },
    },
    defaultMode: "default",
    phases: [
        {
            id: "main",
            title: "Templates",
            stepStyle: "headers",
            steps: [
                {
                    id: "resolve-plugin-dir",
                    kind: "action",
                    preamble: true,
                    toolCalls: ["bash"],
                    title: "Resolve `BITFAB_PLUGIN_DIR`",
                    body: `Codex does not inject a plugin-root env var. Resolve it first (same block used by \`setup\` and \`assistant\`):

${CODEX_RESOLVE_PLUGIN_DIR_BASH}`,
                    next: "main/pick-function",
                },
                {
                    id: "pick-function",
                    kind: "branch",
                    toolCalls: ["ask", "listFunctions", "grep", "glob"],
                    title: "Resolve the trace function key",
                    body: `If the user passed a key as the argument, use it directly and continue.

Otherwise, follow the same picker pattern as \`{{cmd}}assistant\`:

1. Call {{tool:listFunctions}} to enumerate the org's traced functions. The tool returns flat \`FUNCTION: <key>\` lines; work from those keys directly. Use **only** the keys returned: do NOT invent or infer descriptions of what each function does from its name. Key names are often ambiguous, and guessing produces hallucinated summaries that confuse the user.
2. Grep this repo for each key in parallel (across \`*.ts\`, \`*.tsx\`, \`*.py\`, \`*.rb\`, \`*.go\`, \`*.baml\`) so you know which keys are instrumented here. Mark each as ✅ instrumented here (with file path) or ⚠️ not found in this repo.
3. Present a compact list in the question text showing only: \`<key>\` · \`<repo marker + path>\`. No invented summaries.
4. {{askVerbCap}} with 2 options: the recommended function (prefer ✅ instrumented here, and matching session context when one is clearly relevant) and a free-text "Type a function key" option. If nothing is instrumented in this repo, say so explicitly in the question, don't hide it.

{{whens}}`,
                    branches: [
                        {
                            when: "argument supplied",
                            description: "use it as the trace function key and continue",
                            next: "main/load-reference",
                        },
                        {
                            when: "no argument",
                            description: "list trace functions, ask the user, then continue with the chosen key",
                            next: "main/load-reference",
                        },
                    ],
                },
                {
                    id: "load-reference",
                    kind: "action",
                    toolCalls: ["getTemplateReference"],
                    title: "Load the template reference",
                    body: `Call {{tool:getTemplateReference}} **once** before any edit. It returns a stable agent-facing schema for Bitfab span templates: the rendering engine (Nunjucks, Jinja2-compatible), the render-context shape (top-level keys, \`SpanData\` / \`ParsedSpanData\`), the registered custom filters and tests, common patterns from the live default templates, and error-fallback behavior. Without this you cannot write a correct edit; references to undeclared variables silently render empty in production.

Hold the reference in your working context for the rest of the loop. Do NOT call it again on subsequent edits.`,
                    next: "main/ground-in-code",
                },
                {
                    id: "ground-in-code",
                    kind: "action",
                    toolCalls: ["grep", "glob", "read"],
                    title: "Ground edits in the user's code",
                    body: `Before opening the preview, grep the codebase for the trace function key (\`<key>\`) so you can see what the function actually does. The user's "change" requests are usually about surfacing something domain-specific (an input field, a tool name, a context label), and knowing the function helps you map the request to the right span type and the right field path. If grep returns nothing (the function has been renamed or the user is operating on traces from a different repo), continue without it.`,
                    next: "main/check-traces",
                },
                {
                    id: "check-traces",
                    kind: "branch",
                    toolCalls: ["searchTraces"],
                    title: "Verify a trace exists for the function",
                    body: `The preview page renders the most recent trace for the function. Without at least one trace it has nothing to render, so check before opening it.

Call {{tool:searchTraces}} with \`{ traceFunctionKey: "<key>", limit: 1 }\`. If the response contains a trace ID, continue. If the response indicates no traces exist (e.g. \`No traces found matching the filter criteria.\`), exit and tell the user in one short line: \`No traces yet for <key>. Run your app (or the replay script) to generate one, then re-run \\\`{{cmd}}templates <key>\\\` to preview.\` Do NOT block waiting; the user re-invokes when they have a trace.

{{whens}}`,
                    branches: [
                        {
                            when: "trace exists",
                            description: "continue and open the preview",
                            next: "main/open-preview",
                        },
                        {
                            when: "no traces yet for this function",
                            description: "exit and tell the user to generate a trace and re-run",
                            next: null,
                        },
                    ],
                },
                {
                    id: "open-preview",
                    kind: "action",
                    toolCalls: ["bash"],
                    title: "Open the chromeless template-preview page (background)",
                    body: `Launch the preview command **in the background** so the agent can keep iterating while the page stays open:

\`\`\`bash
node "{{pluginRoot}}/dist/commands/startTemplatePreview.js" <functionKey>
\`\`\`

{{#claude}}Run this with \`run_in_background: true\` on the Bash tool. The harness returns a task id and an output file path, and will deliver a \`<task-notification>\` with \`status: completed\` automatically when the process exits. Capture both: you'll need the output file path to poll between edit rounds.{{/claude}}{{#cursor}}Run this as a background process and capture the handle plus its stdout so you can poll its status between edit rounds.{{/cursor}}{{#codex}}Run this in a long-lived exec session (\`exec_command\` with \`is_background: true\`); capture the session id so you can \`read\` it between edit rounds to check whether it has exited.{{/codex}}

The command **blocks until the user clicks the Close button on the page**, then exits 0 with a single line like \`Template preview closed [via loopback]\`. If the user instead just closes the browser tab without clicking Close, the process keeps running until the 30-minute timeout. The page auto-redirects to the most recent trace for the function and renders it with the org's current templates; the user refreshes after each edit to see the new render.`,
                    next: "main/edit-loop",
                },
                {
                    id: "edit-loop",
                    kind: "branch",
                    toolCalls: ["ask", "bash", "getTemplate", "updateTemplate"],
                    title: "Edit loop: change → render → confirm (poll for Close)",
                    body: `Each round of the loop:

1. Ask {{askUser}}what they want changed about the rendering. If the user names one of the six span types in their answer (\`llm\`, \`agent\`, \`function\`, \`guardrail\`, \`handoff\`, \`custom\`), use that. If they don't, ask {{askUser}}which of the six span templates they want to edit before making any changes. Don't guess the span type from a description like "make this less verbose," since the same description fits multiple templates.
2. Call {{tool:getTemplate}} for that span type to read the **live** content (database override if present, otherwise the system default file). **Always** read before write: the prior round may have edited the same template, and overwriting blindly drops that work.
3. Edit the returned source in-context. Stay inside the documented Nunjucks variables and filters (per the reference). Don't introduce \`{% extends %}\`; the assembler injects into \`base.njk\`'s content block, so extends will break composition.
4. Call {{tool:updateTemplate}} with the full edited body. The tool upserts the org's global override in place (no version bump, no row juggling).
5. Tell the user "Refresh the preview to see the change," in one short line. Do not paste the template body back into chat.

Before asking the user about another change, **check whether the background process from step {{step:main/open-preview}} has exited**. The terminal signal is a line containing \`Template preview closed\` on stdout (the process exits 0 right after).

{{#claude}}Two equivalent ways to detect it: (a) if you've already received a \`<task-notification>\` for the captured task id with \`status: completed\`, the user has clicked Close; (b) otherwise, use the \`Read\` tool on the captured output file path and look for the \`Template preview closed\` line. Either signal means the loop should exit.{{/claude}}{{#cursor}}Read the background process's stdout; the \`Template preview closed\` line means the user clicked Close and the process has exited.{{/cursor}}{{#codex}}Use \`read\` against the captured exec session; the \`Template preview closed\` line on stdout means the user clicked Close and the session has exited.{{/codex}}

Two ways the loop ends:

{{whens}}`,
                    branches: [
                        {
                            when: "background process exited (user clicked Close)",
                            description: "exit the loop and acknowledge that template editing is done",
                            next: null,
                        },
                        {
                            when: "user explicitly says they're done",
                            description: "exit the loop and acknowledge",
                            next: null,
                        },
                        {
                            when: "user wants another change",
                            description: "loop back and apply the next edit",
                            next: "main/edit-loop",
                        },
                    ],
                },
            ],
        },
    ],
});
