// bitfab-plugin-lib/dist/flows/update.js
import { Flow } from "./schema.js";
import { CODEX_RESOLVE_PLUGIN_DIR_BASH } from "./shared.js";
export const updateFlow = Flow.parse({
    schemaVersion: 1,
    id: "update",
    title: "Bitfab Update",
    frontmatter: {
        description: "Update Bitfab plugin and SDK to the latest versions",
        argumentHint: "[plugin|sdk|all]",
    },
    tools: {
        bash: { kind: "builtin", name: "Bash" },
        read: { kind: "builtin", name: "Read" },
        grep: { kind: "builtin", name: "Grep" },
        glob: { kind: "builtin", name: "Glob" },
        edit: { kind: "builtin", name: "Edit" },
        ask: { kind: "builtin", name: "AskUserQuestion" },
    },
    intro: `Update the Bitfab {{editor}} plugin and/or every workspace's SDK in the current project.

| Invocation | What runs |
|---|---|
| \`{{cmd}}update\` or \`{{cmd}}update all\` | Plugin update **and** SDK update per workspace |
| \`{{cmd}}update plugin\` | Plugin update only — skips all SDK steps |
| \`{{cmd}}update sdk\` | SDK update only — skips the plugin check |`,
    entries: {
        all: {
            default: "main/run-update-script",
            codex: "main/resolve-plugin-dir",
        },
        plugin: {
            default: "main/run-update-script",
            codex: "main/resolve-plugin-dir",
        },
        sdk: {
            default: "main/run-update-script",
            codex: "main/resolve-plugin-dir",
        },
    },
    defaultMode: "all",
    languageColumnHeaders: {
        detection: "Detection (from workspace/repo)",
        outdated: "Command (run from workspace dir)",
        update: "Command",
    },
    languages: [
        {
            name: "typescript",
            displayName: "TypeScript",
            imports: "`import .* from [\"']bitfab[\"']` or `require\\([\"']bitfab[\"']\\)`",
            detection: "`pnpm-lock.yaml` at repo root → pnpm; `yarn.lock` → yarn; `bun.lock` → bun; otherwise npm",
            outdated: "`pnpm outdated bitfab --json` / `npm outdated bitfab --json` / `yarn outdated bitfab` / `bun outdated bitfab`",
            update: "`pnpm update bitfab@latest` / `yarn upgrade bitfab@latest` / `bun update bitfab` / `npm install bitfab@latest`",
        },
        {
            name: "python",
            displayName: "Python",
            imports: "`^\\s*(from|import) bitfab\\b`",
            detection: "`uv.lock` → uv; `poetry.lock` → poetry; otherwise pip",
            outdated: "`uv pip list --outdated --format=json` / `poetry show -o bitfab-py` / `pip list --outdated --format=json`",
            update: "`uv add bitfab-py@latest` / `poetry add bitfab-py@latest` / `pip install -U bitfab-py` (and bump the pin in `requirements.txt` via Edit)",
        },
        {
            name: "ruby",
            displayName: "Ruby",
            imports: "`require ['\"]bitfab['\"]`",
            detection: "`Gemfile.lock`",
            outdated: "`bundle outdated bitfab --parseable`",
            update: "`bundle update bitfab`",
        },
        {
            name: "go",
            displayName: "Go",
            imports: '`"github.com/Project-White-Rabbit/bitfab-go"`',
            detection: "`go.mod`",
            outdated: "`go list -m -u -json github.com/Project-White-Rabbit/bitfab-go`",
            update: "`go get github.com/Project-White-Rabbit/bitfab-go@latest && go mod tidy`",
        },
    ],
    phases: [
        {
            id: "main",
            title: "Update",
            stepStyle: "headers",
            steps: [
                {
                    id: "resolve-plugin-dir",
                    kind: "action",
                    preamble: true,
                    toolCalls: ["bash"],
                    title: "Resolve `BITFAB_PLUGIN_DIR`",
                    body: `Codex does not inject a plugin-root env var. Resolve it first (same block used by \`setup\` and \`assistant\`):

${CODEX_RESOLVE_PLUGIN_DIR_BASH}`,
                    next: "main/run-update-script",
                },
                {
                    id: "run-update-script",
                    kind: "action",
                    toolCalls: ["bash"],
                    title: "Run the update script",
                    body: `Pass the mode argument the user invoked through to the script (omit for the default \`all\`):

\`\`\`bash
node "{{pluginRoot}}/dist/commands/update.js" <mode>
\`\`\`

- For \`{{cmd}}update plugin\`, run with \`plugin\`.
- For \`{{cmd}}update sdk\`, run with \`sdk\`.
- For \`{{cmd}}update\` or \`{{cmd}}update all\`, run with no argument (or \`all\`).

The script does up to two things depending on mode:
- **Plugin phase** (\`all\` or \`plugin\`) — updates the plugin if a newer version is available.
- **SDK phase** (\`all\` or \`sdk\`) — queries the registry for the latest SDK version and prints a \`<bitfab-sdk-status>\` block with one JSON entry per \`(workspace, language)\` pair. Falls back to the baked snapshot ({{#codex}}sets{{/codex}}{{^codex}}set{{/codex}} \`remoteCheckFailed: true\`) if the registry lookup fails.`,
                    next: {
                        byMode: {
                            all: "main/report-plugin-result",
                            plugin: "main/report-plugin-result",
                            sdk: "main/parse-sdk-status",
                        },
                    },
                },
                {
                    id: "report-plugin-result",
                    kind: "action",
                    title: "Report the plugin result",
                    body: `If the plugin was updated, remind the user to restart {{editor}} to apply the update. If the mode was \`plugin\`, stop here — do not run steps {{step:main/parse-sdk-status}}-{{step:main/run-batch-update}}.`,
                    next: {
                        byMode: {
                            all: "main/parse-sdk-status",
                            plugin: null,
                            sdk: null,
                        },
                    },
                },
                {
                    id: "parse-sdk-status",
                    kind: "action",
                    title: "Parse the SDK status",
                    body: `Each line inside \`<bitfab-sdk-status>\` is a JSON object with fields:

- \`workspacePath\` — relative path of the workspace from the repo root (\`"."\` for non-monorepos or the root package itself)
- \`language\` — \`"typescript" | "python" | "ruby" | "go"\`
- \`packageName\` — the SDK package name (\`bitfab\`, \`bitfab-py\`, \`bitfab\`, \`github.com/Project-White-Rabbit/bitfab-go\`)
- \`declaredVersion\` — the range from the workspace's manifest. May be \`null\` or a loose range.
- \`resolvedVersion\` — the exact version from the lockfile (workspace's own lockfile, or the monorepo root lockfile used as fallback). This is the truth — what the user is actually running.
- \`current\` — \`resolvedVersion ?? declaredVersion\`. Use this for user-facing messages.
- \`latest\` — the latest published version (from the source indicated by \`latestSource\`)
- \`latestSource\` — \`"remote"\` (fetched live from the registry) or \`"baked"\` (snapshot from the plugin build)
- \`remoteCheckFailed\` — \`true\` when the live registry lookup failed; trigger the agent fallback (step {{step:main/agent-fallback}})
- \`updateAvailable\` — \`true\` only when \`latest > current\`
- \`manifestPath\` / \`lockfilePath\` — absolute paths of the files the info came from

If there are **no lines** inside \`<bitfab-sdk-status>\`, the programmatic check found no SDK — but don't stop yet, run step {{step:main/verify-with-agent}} first. After that step, distinguish two cases:

- **No SDK anywhere** (no lines AND step {{step:main/verify-with-agent}} found no imports): the project isn't instrumented yet. Tell the user the Bitfab SDK isn't installed and suggest running \`{{cmd}}setup\` to instrument the project. Stop.
- **SDKs present and current** (every entry has \`updateAvailable: false\` AND step {{step:main/verify-with-agent}} finds no extras): tell the user their SDKs are up to date and stop.{{#codex}}

**Note on Codex's seatbelt sandbox:** by default Codex blocks all outbound network (including localhost). If \`latestSource\` is \`"baked"\` and \`remoteCheckFailed\` is \`true\` for every entry, the registry lookup was likely blocked by the sandbox. Point the user at the [macOS network-access issue](https://github.com/openai/codex/issues/10390) and suggest launching Codex with \`--sandbox danger-full-access\` to get live registry data. In the meantime step {{step:main/agent-fallback}}'s fallback commands give accurate results as long as the user's shell itself has network access.{{/codex}}`,
                    next: "main/verify-with-agent",
                },
                {
                    id: "verify-with-agent",
                    kind: "action",
                    toolCalls: ["grep", "glob"],
                    title: "Verify with an agent pass (always)",
                    body: `The programmatic detection is regex-based and only knows the workspace formats we hand-coded (pnpm/npm/yarn workspaces, uv workspaces, go.work). It can miss unusual monorepo layouts, vendored SDKs, or projects using package managers we don't parse. **Always run this verification** before offering updates.

- Grep the project for SDK imports (run these in parallel):
{{languages:imports}}
- For each import, find its workspace directory by walking up to the nearest \`package.json\` / \`pyproject.toml\` / \`Gemfile\` / \`go.mod\`.
- Compare that set against the \`workspacePath\` values in \`<bitfab-sdk-status>\`.
- For each workspace that has imports but **no** corresponding status entry, treat it as a missed detection: ask the user which package manager that workspace uses, then go to step {{step:main/agent-fallback}} for it (same flow as \`remoteCheckFailed: true\`).
- If the sets match, proceed.`,
                    next: "main/agent-fallback",
                },
                {
                    id: "agent-fallback",
                    kind: "action",
                    toolCalls: ["bash"],
                    title: "Agent fallback for `remoteCheckFailed` or detection gaps",
                    body: `For each entry where \`remoteCheckFailed: true\`, or any workspace discovered only in step {{step:main/verify-with-agent}}, run the package manager's native outdated command from the workspace directory. The command is authoritative — it respects private registries, mirrors, and offline caches.

{{languages:detection,outdated}}

Use the real latest from the command's output in place of \`latest\` when deciding whether to offer an upgrade.`,
                    next: "main/ask-batch",
                },
                {
                    id: "ask-batch",
                    kind: "branch",
                    toolCalls: ["ask"],
                    title: "Ask whether to batch-update or step through workspaces",
                    body: `If there are **3 or more** workspaces with \`updateAvailable: true\`, ask {{askUser}}— **one decision per question**:

{{branches}}

**Always recommend "Update all" (option A).** Do not downgrade the recommendation based on the range specifier or lockfile shape — not for \`workspace:*\` / \`workspace:^\`, not for git refs, not for pinned \`"=X.Y.Z"\`, not for path deps. An outdated SDK is an outdated SDK. If the user is working inside a monorepo where the dep is workspace-linked to a sibling SDK package, they are free to pick **Skip** themselves, but the recommended action is still **Update**.

If there are **fewer than 3** outdated workspaces, skip this prompt and go straight to per-workspace.`,
                    branches: [
                        {
                            when: "fewer than 3 outdated workspaces",
                            next: "main/ask-per-workspace",
                        },
                        {
                            option: {
                                letter: "A",
                                label: "Update all N outdated workspaces",
                            },
                            recommended: true,
                            next: "main/run-batch-update",
                        },
                        {
                            option: { letter: "B", label: "Ask me per workspace" },
                            next: "main/ask-per-workspace",
                        },
                        {
                            option: { letter: "C", label: "Skip everything" },
                            next: null,
                        },
                    ],
                },
                {
                    id: "ask-per-workspace",
                    kind: "branch",
                    toolCalls: ["ask"],
                    title: "Ask per workspace whether to update",
                    body: `For the next workspace with \`updateAvailable: true\`, ask {{askUser}}— **one decision per question**:

> We recommend **Update**: \`<workspacePath>\` — \`<language>\` SDK \`<current>\` → \`<latest>\`.
>
{{branches}}

When no outdated workspaces remain, exit and acknowledge.`,
                    branches: [
                        {
                            option: {
                                letter: "A",
                                label: "Update",
                                description: "run the package manager update command now",
                            },
                            recommended: true,
                            next: "main/run-single-update",
                        },
                        {
                            option: {
                                letter: "B",
                                label: "Skip",
                                description: "leave this workspace on `<current>`",
                            },
                            next: "main/ask-per-workspace",
                        },
                        {
                            when: "no outdated workspaces remain",
                            next: null,
                        },
                    ],
                },
                {
                    id: "run-single-update",
                    kind: "action",
                    toolCalls: ["bash", "read", "edit"],
                    title: "Run the update for the chosen workspace",
                    body: `Detect the package manager from the lockfiles and run the update **from the workspace directory** (not repo root — matters in monorepos):

{{languages:update}}

After the update, Read the manifest to verify the new version and confirm to the user, then return to the per-workspace prompt for the next workspace.`,
                    next: "main/ask-per-workspace",
                },
                {
                    id: "run-batch-update",
                    kind: "action",
                    toolCalls: ["bash", "read", "edit"],
                    title: "Run updates for every outdated workspace",
                    body: `For every workspace with \`updateAvailable: true\`, detect the package manager from the lockfiles and run the update **from each workspace directory** (not repo root):

{{languages:update}}

After each update, Read the manifest to verify the new version and confirm to the user.`,
                    next: null,
                },
            ],
        },
    ],
});
