// bitfab-plugin-lib/dist/commands/openStudio.js
import crypto from "node:crypto";
import { createAgentSession, pollAgentSessionEvents, pushAgentSessionEvent, } from "../agentSessionChannel.js";
import { openChromelessWindow } from "../browser.js";
import { getConfig, hasCredentials } from "../config.js";
import { recordFocus } from "../frontmostApp.js";
export async function openStudio({ serviceUrl, apiKey, sessionId, }) {
    const restoreFocus = recordFocus();
    const client = { serviceUrl, apiKey, sessionId };
    await createAgentSession({ serviceUrl, apiKey, sessionId });
    await pushAgentSessionEvent(client, {
        type: "agent:navigate",
        data: { path: "/studio" },
    });
    const url = `${serviceUrl}/studio?session=${encodeURIComponent(sessionId)}`;
    openChromelessWindow(url);
    const abortController = new AbortController();
    pollAgentSessionEvents({
        serviceUrl,
        apiKey,
        sessionId,
        abortSignal: abortController.signal,
        onEvent: (event) => {
            if (event.type === "studio:return-to-agent") {
                restoreFocus();
            }
            if (event.type === "studio:session-ended") {
                console.log("Studio session ended by user.");
                abortController.abort();
            }
        },
        onError: (err) => {
            console.error(`studio agent-event poll: ${err.message}`);
        },
    }).catch((err) => {
        if (!abortController.signal.aborted) {
            console.error(`studio agent-event poll terminated: ${err.message}`);
        }
    });
    return {
        sessionId,
        abort: () => abortController.abort(),
    };
}
export async function runOpenStudio(platform) {
    if (!hasCredentials()) {
        console.error(`Not authenticated. Run ${platform.loginHint} first.`);
        process.exit(1);
    }
    const config = getConfig();
    if (!config.apiKey) {
        console.error("No API key configured.");
        process.exit(1);
    }
    const sessionId = crypto.randomUUID();
    console.log(`Opening Studio: ${config.serviceUrl}/studio?session=${sessionId}`);
    const session = await openStudio({
        serviceUrl: config.serviceUrl,
        apiKey: config.apiKey,
        sessionId,
    });
    process.on("SIGINT", () => {
        session.abort();
        process.exit(0);
    });
}
