// bitfab-plugin-lib/dist/commands/openStudio.js
import crypto from "node:crypto";
import { clearActiveStudioSession, writeActiveStudioSession, } from "../activeStudioSession.js";
import { createAgentSession, pollAgentSessionEvents, } from "../agentSessionChannel.js";
import { openChromelessWindow } from "../browser.js";
import { getConfig, hasCredentials } from "../config.js";
import { recordFocus } from "../frontmostApp.js";
function emitJson(event) {
    process.stdout.write(`${JSON.stringify(event)}\n`);
}
export async function openStudio({ serviceUrl, apiKey, sessionId, initialPath = "/studio", clientHeaders, }) {
    const restoreFocus = recordFocus();
    await createAgentSession({
        serviceUrl,
        apiKey,
        sessionId,
        extraHeaders: clientHeaders,
    });
    // Announce the session to other plugin processes in this worktree so that
    // Studio-aware commands (e.g. openTracePlan) can route into the existing
    // tab instead of spawning a duplicate window. Stale records from crashed
    // openStudio runs are detected on read via pid liveness, so a hard exit
    // that misses the cleanup hook below is self-correcting.
    writeActiveStudioSession({ sessionId, serviceUrl });
    const separator = initialPath.includes("?") ? "&" : "?";
    const url = `${serviceUrl}${initialPath}${separator}session=${encodeURIComponent(sessionId)}`;
    openChromelessWindow(url);
    const abortController = new AbortController();
    pollAgentSessionEvents({
        serviceUrl,
        apiKey,
        sessionId,
        abortSignal: abortController.signal,
        onEvent: (event) => {
            if (event.type === "studio:return-to-agent") {
                emitJson({ event: "return-to-agent", sessionId });
                restoreFocus();
            }
            if (event.type === "studio:edit-with-agent") {
                emitJson({
                    event: "edit-with-agent",
                    sessionId,
                    datasetId: event.data.datasetId,
                });
                restoreFocus();
            }
            if (event.type === "studio:session-ended") {
                emitJson({ event: "session-ended", sessionId });
                clearActiveStudioSession();
                restoreFocus();
                abortController.abort();
            }
            if (event.type === "studio:auth-required") {
                emitJson({ event: "auth-required", sessionId });
            }
            if (event.type === "studio:authenticated") {
                emitJson({ event: "authenticated", sessionId });
            }
        },
        onError: (err) => {
            console.error(`studio agent-event poll: ${err.message}`);
        },
    }).catch((err) => {
        if (!abortController.signal.aborted) {
            console.error(`studio agent-event poll terminated: ${err.message}`);
        }
    });
    return {
        sessionId,
        abort: () => abortController.abort(),
    };
}
export async function runOpenStudio(platform) {
    if (!hasCredentials()) {
        console.error(`Not authenticated. Run ${platform.loginHint} first.`);
        process.exit(1);
    }
    const config = getConfig();
    if (!config.apiKey) {
        console.error("No API key configured.");
        process.exit(1);
    }
    const initialPath = process.argv[2] ?? "/studio";
    const sessionId = crypto.randomUUID();
    const separator = initialPath.includes("?") ? "&" : "?";
    console.error(`Opening Studio: ${config.serviceUrl}${initialPath}${separator}session=${sessionId}`);
    const session = await openStudio({
        serviceUrl: config.serviceUrl,
        apiKey: config.apiKey,
        sessionId,
        initialPath,
        clientHeaders: {
            "x-bitfab-client-name": platform.cliBinary,
            "x-bitfab-client-display-name": platform.displayName,
        },
    });
    emitJson({ event: "started", sessionId });
    const shutdown = (exitCode) => {
        session.abort();
        clearActiveStudioSession();
        process.exit(exitCode);
    };
    process.on("SIGINT", () => shutdown(0));
    process.on("SIGTERM", () => shutdown(0));
    process.on("exit", () => {
        // `exit` is sync-only; rely on the synchronous fs.rmSync inside
        // clearActiveStudioSession. Safe to call even if abort() was already
        // invoked above — clear is idempotent.
        clearActiveStudioSession();
    });
}
