// bitfab-plugin-lib/dist/commands/update.js
import { execSync } from "node:child_process";
import { checkForSdkUpdates, checkForSdkUpdatesOffline, formatSdkUpdateLine, } from "../sdkUpdates.js";
import { checkForUpdate } from "../updates.js";
function updatePlugin(current, latest, platform) {
    console.log(`Plugin update available: v${current} → v${latest}`);
    console.log("Updating marketplace...");
    try {
        execSync(`${platform.cliBinary} plugin marketplace update ${platform.marketplaceName}`, { stdio: "inherit" });
    }
    catch {
        console.error(`Failed to update marketplace '${platform.marketplaceName}'.`);
        if (!platform.marketplacePreRegistered) {
            console.error(`You can add it with: ${platform.cliBinary} plugin marketplace add ${platform.repo}`);
        }
        process.exit(1);
    }
    console.log("Updating plugin...");
    try {
        execSync(`${platform.cliBinary} plugin update ${platform.pluginName}@${platform.marketplaceName}`, { stdio: "inherit" });
    }
    catch {
        console.error("Failed to update plugin.");
        process.exit(1);
    }
    console.log(`\nBitfab plugin updated to v${latest}. Restart ${platform.displayName} to apply the update.`);
}
function printSdkStatus(statuses) {
    if (statuses.length === 0) {
        console.log("\nSDK: no Bitfab SDK detected in the current directory.");
        return;
    }
    const outdated = statuses.filter((s) => s.updateAvailable && s.latest);
    const upToDate = statuses.filter((s) => !s.updateAvailable);
    const needsAgent = statuses.filter((s) => s.remoteCheckFailed);
    console.log("\nSDK status:");
    console.log("<bitfab-sdk-status>");
    for (const s of statuses) {
        console.log(JSON.stringify({
            workspacePath: s.workspacePath,
            language: s.language,
            packageName: s.packageName,
            declaredVersion: s.declaredVersion,
            resolvedVersion: s.resolvedVersion,
            current: s.current,
            latest: s.latest,
            latestSource: s.latestSource,
            remoteCheckFailed: s.remoteCheckFailed,
            updateAvailable: s.updateAvailable,
            manifestPath: s.manifestPath,
            lockfilePath: s.lockfilePath,
        }));
    }
    console.log("</bitfab-sdk-status>");
    for (const s of upToDate) {
        const v = s.current ?? s.latest ?? "unknown";
        const suffix = s.remoteCheckFailed ? " (baked check only)" : "";
        const prefix = s.workspacePath && s.workspacePath !== "." ? `${s.workspacePath} ` : "";
        console.log(`  ${prefix}${s.language} (${s.packageName}) v${v} — up to date${suffix}`);
    }
    if (outdated.length > 0) {
        console.log("\nSDK updates available:");
        for (const s of outdated) {
            const suffix = s.latestSource === "baked" ? " (baked check)" : "";
            console.log(`  ${formatSdkUpdateLine(s)}${suffix}`);
        }
        console.log("\nRun the Bitfab update skill to apply SDK updates with the correct package manager.");
    }
    if (needsAgent.length > 0) {
        console.log("\nRemote registry lookup failed for some SDKs — the update skill should run the package manager's native outdated command (e.g. `pnpm outdated bitfab`, `pip list --outdated`, `bundle outdated bitfab`) to determine the true latest.");
    }
}
export function parseUpdateMode(raw) {
    if (!raw || raw === "all") {
        return "all";
    }
    if (raw === "plugin" || raw === "sdk") {
        return raw;
    }
    throw new Error(`Unknown update mode "${raw}". Expected one of: plugin, sdk, all.`);
}
async function runPluginUpdate(currentVersion, platform) {
    const { current, latest, updateAvailable } = await checkForUpdate(currentVersion, platform);
    if (!updateAvailable || !latest) {
        console.log(`Bitfab plugin v${current} is already up to date.`);
        return;
    }
    updatePlugin(current, latest, platform);
}
async function runSdkUpdate() {
    // Skip the offline-first preview: when the remote check succeeds, the
    // baked value is just noise (and misleading when the baked SDK is ahead
    // of what's actually published, as happens inside the monorepo that
    // builds the plugin). checkForSdkUpdates falls back to baked internally
    // via remoteCheckFailed if the network lookup fails, so printSdkStatus
    // still surfaces baked when it's the only thing we have.
    const offline = checkForSdkUpdatesOffline(process.cwd());
    if (offline.length === 0) {
        console.log("\nSDK: no Bitfab SDK detected in the current directory.");
        return;
    }
    try {
        const sdkStatuses = await checkForSdkUpdates(process.cwd());
        printSdkStatus(sdkStatuses);
    }
    catch (err) {
        const message = err instanceof Error ? err.message : String(err);
        console.error(`SDK update check failed: ${message}`);
    }
}
export async function runUpdate(currentVersion, platform, mode = "all") {
    if (mode === "plugin" || mode === "all") {
        await runPluginUpdate(currentVersion, platform);
    }
    if (mode === "sdk" || mode === "all") {
        await runSdkUpdate();
    }
}
