// bitfab-plugin-lib/dist/chatSessions/client.js
import { getConfig } from "../config.js";
export async function appendChatSession(request) {
    const { serviceUrl, apiKey } = getConfig();
    if (!apiKey) {
        return { ok: false, reason: "auth" };
    }
    let response;
    try {
        response = await fetch(`${serviceUrl}/api/plugin/chatSessions`, {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
                Authorization: `Bearer ${apiKey}`,
            },
            body: JSON.stringify(request),
        });
    }
    catch (error) {
        return {
            ok: false,
            reason: "network",
            detail: error instanceof Error ? error.message : String(error),
        };
    }
    if (response.status === 409) {
        const body = (await response.json().catch(() => ({})));
        return {
            ok: false,
            reason: "offset_conflict",
            lastTranscriptOffset: body.lastTranscriptOffset ?? 0,
            turnCount: body.turnCount ?? 0,
        };
    }
    if (response.status === 400) {
        const body = (await response.json().catch(() => ({})));
        return { ok: false, reason: "server", detail: body.error };
    }
    if (response.status === 401) {
        return { ok: false, reason: "auth" };
    }
    if (!response.ok) {
        return {
            ok: false,
            reason: "server",
            detail: `HTTP ${response.status}`,
        };
    }
    const body = (await response.json());
    return {
        ok: true,
        id: body.id,
        lastTranscriptOffset: body.lastTranscriptOffset,
        turnCount: body.turnCount,
    };
}
export async function getChatSessionState(externalId) {
    const { serviceUrl, apiKey } = getConfig();
    if (!apiKey) {
        return null;
    }
    try {
        const response = await fetch(`${serviceUrl}/api/plugin/chatSessions/${encodeURIComponent(externalId)}/state`, {
            method: "GET",
            headers: {
                Authorization: `Bearer ${apiKey}`,
            },
        });
        if (!response.ok) {
            return null;
        }
        return (await response.json());
    }
    catch {
        return null;
    }
}
