// bitfab-plugin-lib/dist/chatSessions/index.js
import { execFileSync } from "node:child_process";
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { extractSkillFromPrompt } from "../hooks/skillDetection.js";
import { arm, disarm, ensureSessionsDir, isArmed, isBitfabMcpToolName, listArmedSessionIds, localOffsetPath, readArmedMetadata, } from "./arming.js";
import { appendChatSession, getChatSessionState, } from "./client.js";
import { readTranscriptDelta } from "./transcriptReader.js";
function configFilePath() {
    return path.join(os.homedir(), ".config", "bitfab", "config.json");
}
export function isLocalCaptureDisabled() {
    const envValue = process.env.BITFAB_CAPTURE_SESSIONS;
    if (envValue === "1" || envValue === "true") {
        return false;
    }
    if (envValue === "0" || envValue === "false") {
        return true;
    }
    try {
        const raw = fs.readFileSync(configFilePath(), "utf-8");
        const parsed = JSON.parse(raw);
        return parsed.sessionLogConsent !== true;
    }
    catch {
        return true;
    }
}
/**
 * Single entry point for every Claude Code hook script. Decides whether to
 * arm the session, whether there's anything new to send, and posts the delta.
 *
 * The function is deliberately fire-and-forget from the hook's perspective —
 * any error is swallowed into a CaptureResult so the hook never blocks the
 * agent.
 */
export async function captureFromHook(hookType, input, agent = "claude-code") {
    if (isLocalCaptureDisabled()) {
        return { status: "skipped_opt_out" };
    }
    const sessionId = input.session_id;
    const transcriptPath = input.transcript_path;
    if (!sessionId || !transcriptPath) {
        return { status: "skipped_no_transcript" };
    }
    const armedMeta = {
        transcriptPath,
        cwd: input.cwd ?? process.cwd(),
        agent,
    };
    if (hookType === "PreToolUse" && input.tool_name) {
        if (isBitfabMcpToolName(input.tool_name) && !isArmed(sessionId)) {
            arm(sessionId, armedMeta);
        }
    }
    if (hookType === "UserPromptSubmit" && input.prompt) {
        if (extractSkillFromPrompt(input.prompt) !== null && !isArmed(sessionId)) {
            arm(sessionId, armedMeta);
        }
    }
    if (hookType === "SessionStart") {
        sweepStaleSessions(sessionId);
    }
    if (!isArmed(sessionId)) {
        return { status: "skipped_not_armed" };
    }
    if (!fs.existsSync(transcriptPath)) {
        return { status: "skipped_no_transcript" };
    }
    const isTerminal = hookType === "Stop" || hookType === "SubagentStop";
    const result = await flushDelta({
        sessionId,
        transcriptPath,
        cwd: input.cwd ?? process.cwd(),
        agent,
        isResume: hookType === "SessionStart" && input.source === "resume",
        ended: isTerminal,
    });
    if (isTerminal) {
        const flushedOrNoop = result.status === "appended" || result.status === "no_op";
        if (flushedOrNoop) {
            cleanupSession(sessionId);
        }
    }
    return result;
}
async function flushDelta(args) {
    ensureSessionsDir();
    const offsetFile = localOffsetPath(args.sessionId);
    let state = readOffsetState(offsetFile);
    if (args.isResume) {
        const remote = await getChatSessionState(args.sessionId);
        if (remote?.exists) {
            state = { offset: remote.lastTranscriptOffset };
        }
    }
    const stat = fs.statSync(args.transcriptPath);
    if (state.offset > stat.size) {
        const remote = await getChatSessionState(args.sessionId);
        const sequenceOffset = remote?.exists ? remote.turnCount : 0;
        const delta = readTranscriptDelta(args.transcriptPath, 0);
        if (delta.turns.length === 0 && !args.ended) {
            writeOffsetState(offsetFile, {
                offset: delta.newOffset,
                sequenceBase: sequenceOffset,
            });
            return { status: "no_op", detail: "compaction detected, no new turns" };
        }
        for (const turn of delta.turns) {
            turn.sequence += sequenceOffset;
        }
        const newSequenceBase = sequenceOffset + delta.turns.length;
        const startedAt = inferSessionStart(args.transcriptPath);
        const result = await appendChatSession({
            externalId: args.sessionId,
            agent: args.agent,
            startedAt,
            metadata: collectMetadata(args.cwd),
            fromOffset: 0,
            toOffset: delta.newOffset,
            turns: delta.turns,
            compacted: true,
            ended: args.ended || undefined,
        });
        return handleAppendResponse(result, offsetFile, newSequenceBase);
    }
    const delta = readTranscriptDelta(args.transcriptPath, state.offset, state.sequenceBase);
    if (delta.turns.length === 0 &&
        state.offset === delta.newOffset &&
        !args.ended) {
        return { status: "no_op" };
    }
    const startedAt = inferSessionStart(args.transcriptPath);
    const result = await appendChatSession({
        externalId: args.sessionId,
        agent: args.agent,
        startedAt,
        metadata: collectMetadata(args.cwd),
        fromOffset: state.offset,
        toOffset: delta.newOffset,
        turns: delta.turns,
        ended: args.ended || undefined,
    });
    return handleAppendResponse(result, offsetFile, state.sequenceBase);
}
function handleAppendResponse(response, offsetFile, sequenceBase) {
    if (response.ok) {
        writeOffsetState(offsetFile, {
            offset: response.lastTranscriptOffset,
            sequenceBase,
        });
        return {
            status: "appended",
            appended: {
                lastTranscriptOffset: response.lastTranscriptOffset,
                turnCount: response.turnCount,
            },
        };
    }
    if (response.reason === "offset_conflict") {
        writeOffsetState(offsetFile, {
            offset: response.lastTranscriptOffset,
            sequenceBase,
        });
        return {
            status: "no_op",
            detail: `offset_conflict; resynced to ${response.lastTranscriptOffset}`,
        };
    }
    if (response.reason === "auth") {
        return { status: "auth_missing" };
    }
    return { status: "error", detail: response.detail };
}
function readOffsetState(offsetFile) {
    try {
        const raw = fs.readFileSync(offsetFile, "utf-8").trim();
        if (raw.startsWith("{")) {
            const parsed = JSON.parse(raw);
            const offset = Number(parsed.offset);
            const seqBase = Number(parsed.sequenceBase);
            return {
                offset: Number.isFinite(offset) && offset >= 0 ? offset : 0,
                sequenceBase: Number.isFinite(seqBase) && seqBase >= 0 ? seqBase : undefined,
            };
        }
        const parsed = Number.parseInt(raw, 10);
        return { offset: Number.isFinite(parsed) && parsed >= 0 ? parsed : 0 };
    }
    catch {
        return { offset: 0 };
    }
}
function writeOffsetState(offsetFile, state) {
    fs.writeFileSync(offsetFile, JSON.stringify(state));
}
function inferSessionStart(transcriptPath) {
    try {
        const stat = fs.statSync(transcriptPath);
        return stat.birthtime.toISOString();
    }
    catch {
        return new Date().toISOString();
    }
}
function cleanupSession(sessionId) {
    disarm(sessionId);
    try {
        fs.unlinkSync(localOffsetPath(sessionId));
    }
    catch {
        // Never crash the host app
    }
}
function sweepStaleSessions(currentSessionId) {
    for (const sessionId of listArmedSessionIds()) {
        if (sessionId === currentSessionId) {
            continue;
        }
        const meta = readArmedMetadata(sessionId);
        if (!meta) {
            disarm(sessionId);
            continue;
        }
        if (!fs.existsSync(meta.transcriptPath)) {
            cleanupSession(sessionId);
            continue;
        }
        flushDelta({
            sessionId,
            transcriptPath: meta.transcriptPath,
            cwd: meta.cwd,
            agent: meta.agent,
            isResume: false,
            ended: true,
        })
            .then((result) => {
            if (result.status === "appended" || result.status === "no_op") {
                cleanupSession(sessionId);
            }
        })
            .catch(() => { });
    }
}
function collectMetadata(cwd) {
    const metadata = { cwd };
    const branch = readGitBranch(cwd);
    if (branch) {
        metadata.gitBranch = branch;
    }
    return metadata;
}
function readGitBranch(cwd) {
    try {
        const out = execFileSync("git", ["branch", "--show-current"], {
            cwd,
            timeout: 1500,
            encoding: "utf-8",
        }).trim();
        return out || null;
    }
    catch {
        return null;
    }
}
