// bitfab-plugin-lib/dist/agentSessionChannel.js
const DEFAULT_POLL_INTERVAL_MS = 1500;
export async function createAgentSession(opts) {
    const url = `${opts.serviceUrl}/api/studio/sessions`;
    const res = await fetch(url, {
        method: "POST",
        headers: {
            Authorization: `Bearer ${opts.apiKey}`,
            "Content-Type": "application/json",
            ...opts.extraHeaders,
        },
        body: JSON.stringify({ sessionId: opts.sessionId }),
    });
    if (!res.ok) {
        throw new Error(`createAgentSession failed (${res.status}): ${await res.text()}`);
    }
    return (await res.json());
}
/**
 * Push an event into a session's event stream. Use this to send commands
 * (e.g. "agent:navigate") to the browser tab paired with `sessionId`.
 * Event types and data shapes come from the typed `AgentEventDataMap`
 * union.
 */
export async function pushAgentSessionEvent(client, event) {
    const url = `${client.serviceUrl}/api/studio/events`;
    const res = await fetch(url, {
        method: "POST",
        headers: {
            Authorization: `Bearer ${client.apiKey}`,
            "Content-Type": "application/json",
        },
        body: JSON.stringify({
            session: client.sessionId,
            type: event.type,
            data: event.data,
        }),
    });
    if (!res.ok) {
        throw new Error(`pushAgentSessionEvent failed (${res.status}): ${await res.text()}`);
    }
    return (await res.json());
}
/**
 * Continuously poll a session's event stream. Each poll request is short-
 * lived (Vercel-friendly) and asks for events strictly after `since`. The
 * cursor advances after each successful poll.
 *
 * The loop runs until `abortSignal` fires. Network errors are surfaced via
 * `onError` and the loop continues; transient failures should not terminate
 * the channel.
 */
export async function pollAgentSessionEvents(opts) {
    const interval = opts.intervalMs ?? DEFAULT_POLL_INTERVAL_MS;
    let since = opts.startCursor ?? null;
    while (!opts.abortSignal.aborted) {
        try {
            const url = new URL(`${opts.serviceUrl}/api/studio/events`);
            url.searchParams.set("session", opts.sessionId);
            if (since) {
                url.searchParams.set("since", since);
            }
            const res = await fetch(url.toString(), {
                headers: { Authorization: `Bearer ${opts.apiKey}` },
                signal: opts.abortSignal,
            });
            if (res.ok) {
                const body = (await res.json());
                for (const event of body.events) {
                    opts.onEvent(event);
                    since = event.id;
                }
                if (opts.onBrowserStatus &&
                    typeof body.browserConnected === "boolean") {
                    opts.onBrowserStatus(body.browserConnected);
                }
            }
            else if (opts.onError) {
                opts.onError(new Error(`agent-session poll failed (${res.status})`));
            }
        }
        catch (err) {
            if (opts.abortSignal.aborted) {
                return;
            }
            if (opts.onError) {
                opts.onError(err instanceof Error ? err : new Error(String(err)));
            }
        }
        await new Promise((resolve) => {
            const timer = setTimeout(resolve, interval);
            opts.abortSignal.addEventListener("abort", () => {
                clearTimeout(timer);
                resolve();
            }, { once: true });
        });
    }
}
export async function closeStudio(client, message) {
    return pushAgentSessionEvent(client, {
        type: "agent:close-studio",
        data: message ? { message } : {},
    });
}
export async function navigateStudio(client, path) {
    return pushAgentSessionEvent(client, {
        type: "agent:navigate",
        data: { path },
    });
}
/**
 * How long to wait for the open Studio tab to echo back `studio:navigated`
 * after we push an `agent:navigate`. The browser shell emits it from a
 * `usePathname` effect, so an alive tab responds in milliseconds — a
 * silence past this window means the tab is closed, crashed, or stuck on
 * sign-in and the caller should fall back to a chromeless window.
 *
 * Module-private so every caller goes through `navigateStudioAndAwaitAck`
 * (CLI + in-process) and the 12-second contract documented in the
 * assistant skill stays consistent across them.
 */
const STUDIO_NAVIGATE_ACK_TIMEOUT_MS = 12_000;
/**
 * Push an `agent:navigate` event and resolve once the browser shell
 * acknowledges the navigation (via `studio:navigated` on the matching
 * path), the session ends, or the timeout fires. Deterministic prefix
 * step for any caller that wants to route the open Studio tab somewhere
 * and then decide what to do based on whether the tab actually moved.
 *
 * The CLI command (`runNavigateStudio`) and in-process callers (e.g.
 * `openTracePlan.js`'s Studio handoff path) both go through this helper
 * so the ack contract — and the 12s timeout that the assistant skill
 * documents — stay in one place.
 */
export async function navigateStudioAndAwaitAck(client, path, opts = {}) {
    const ackTimeoutMs = opts.ackTimeoutMs ?? STUDIO_NAVIGATE_ACK_TIMEOUT_MS;
    const abortController = new AbortController();
    // Lifted out of the Promise constructor so the navigate-POST catch
    // below can clear it on push-failed; otherwise the timer leaks and
    // keeps the Node event loop alive for `ackTimeoutMs` past the early
    // return (Graphite review on #1662).
    let timer = null;
    const result = new Promise((resolve) => {
        timer = setTimeout(() => {
            resolve({ acked: false, reason: "timeout" });
            abortController.abort();
        }, ackTimeoutMs);
        pollAgentSessionEvents({
            serviceUrl: client.serviceUrl,
            apiKey: client.apiKey,
            sessionId: client.sessionId,
            abortSignal: abortController.signal,
            onEvent: (event) => {
                if (event.type === "studio:navigated" && event.data.path === path) {
                    if (timer) {
                        clearTimeout(timer);
                    }
                    resolve({ acked: true });
                    abortController.abort();
                    return;
                }
                if (event.type === "studio:session-ended") {
                    if (timer) {
                        clearTimeout(timer);
                    }
                    resolve({ acked: false, reason: "session-ended" });
                    abortController.abort();
                }
            },
            onError: () => {
                // Swallow transient poll errors; pollAgentSessionEvents keeps
                // going. A terminal failure is handled by the catch below.
            },
        }).catch(() => {
            if (abortController.signal.aborted) {
                return;
            }
            if (timer) {
                clearTimeout(timer);
            }
            resolve({ acked: false, reason: "timeout" });
        });
    });
    try {
        await navigateStudio(client, path);
    }
    catch {
        if (timer) {
            clearTimeout(timer);
        }
        abortController.abort();
        return { acked: false, reason: "push-failed" };
    }
    return result;
}
/**
 * Read the agent-event stream once and report whether the browser tab
 * paired with `sessionId` has an active SSE connection right now.
 * Deterministic — backed by the same Redis-heartbeat `browserConnected`
 * field that the polling loop surfaces incrementally via
 * `onBrowserStatus`. Use this when you need a synchronous yes/no before
 * deciding whether to push events into the session.
 *
 * Returns `false` on any failure (network, auth, malformed body): callers
 * almost always want to treat "couldn't tell" identically to "no Studio
 * open" and fall back to whatever their non-Studio path is.
 */
export async function isStudioBrowserConnected(client, opts = {}) {
    const timeoutMs = opts.timeoutMs ?? 2_000;
    try {
        const url = new URL(`${client.serviceUrl}/api/studio/events`);
        url.searchParams.set("session", client.sessionId);
        url.searchParams.set("limit", "1");
        const res = await fetch(url.toString(), {
            headers: { Authorization: `Bearer ${client.apiKey}` },
            signal: AbortSignal.timeout(timeoutMs),
        });
        if (!res.ok) {
            return false;
        }
        const body = (await res.json());
        return body.browserConnected === true;
    }
    catch {
        return false;
    }
}
