// bitfab-plugin-lib/dist/browserHandoff.js
import http from "node:http";
import { openChromelessWindow } from "./browser.js";
import { startHandoffTicketChannel } from "./handoffTicketChannel.js";
import { startLoopbackServer, } from "./loopback.js";
const DEFAULT_TIMEOUT_MS = 10 * 60 * 1000;
const DEFAULT_REDISPLAY_DELAY_MS = 4 * 1000;
const DEFAULT_HEARTBEAT_INTERVAL_MS = 30 * 1000;
export function findOpenPort() {
    return new Promise((resolve, reject) => {
        const server = http.createServer();
        // Bind to 127.0.0.1 explicitly — sandboxed runtimes (Codex seatbelt,
        // container runtimes, restricted CI) reject listen() on 0.0.0.0 with
        // EPERM/EACCES, and the callback only needs loopback anyway. Also
        // subscribe to 'error' so those rejections reject the Promise instead
        // of escaping as an unhandled EventEmitter 'error' and crashing Node.
        server.on("error", (err) => {
            server.close(() => reject(err));
        });
        server.listen(0, "127.0.0.1", () => {
            const addr = server.address();
            if (addr && typeof addr === "object") {
                const port = addr.port;
                server.close(() => resolve(port));
            }
            else {
                reject(new Error("Could not find open port"));
            }
        });
    });
}
/**
 * Low-level primitive: start both channels + the browser, return handles so
 * the caller can compose additional racers (e.g. login's interactive paste)
 * and print their own hint messages. Most callers should use the higher-level
 * `runBrowserHandoff` instead.
 */
export async function startBrowserHandoff(opts) {
    const timeoutMs = opts.timeoutMs ?? DEFAULT_TIMEOUT_MS;
    // Debug flag: force ticket-only delivery even on a machine where
    // loopback would otherwise bind. Used to exercise the server-polled
    // ticket path locally without needing an SSH/Docker/container env.
    const forceTicket = process.env.BITFAB_FORCE_TICKET === "1";
    // Try to pick a free port. Sandboxed runtimes (Codex, some containers,
    // restricted CI) reject `listen(0)` with EPERM / EACCES; treat that as
    // "loopback unavailable" and fall through to the ticket-only path
    // instead of crashing the whole login.
    let port = null;
    let portBindError = null;
    if (!forceTicket) {
        try {
            port = await findOpenPort();
        }
        catch (err) {
            portBindError = err;
        }
    }
    const loopbackUnavailable = {
        promise: new Promise(() => { }),
        get done() {
            return false;
        },
        available: false,
        bindError: portBindError ?? new Error("BITFAB_FORCE_TICKET=1 disabled loopback"),
        resolve: () => { },
        reject: () => { },
        close: () => { },
    };
    const [loopback, ticket] = await Promise.all([
        forceTicket || port === null
            ? Promise.resolve(loopbackUnavailable)
            : startLoopbackServer(port, opts.loopbackRoutes),
        startHandoffTicketChannel({
            serviceUrl: opts.serviceUrl,
            kind: opts.ticketKind,
            timeoutMs,
            ttlSeconds: opts.ticketTtlSeconds,
            bearer: opts.ticketBearer,
        }),
    ]);
    const browserUrl = opts.buildUrl({
        port: loopback.available ? port : null,
        ticketId: ticket.ticketId,
    });
    return { loopback, ticket, port, browserUrl, timeoutMs };
}
/**
 * Attach session-keepalive UX to a running browser handoff. Two concerns,
 * each on its own timer:
 *
 *   1. **URL redisplay** at `redisplayDelayMs` (default 4s). Purely
 *      informational — re-prints the URL under a "still waiting for
 *      browser handoff" banner so the user/agent has a fresh signal that
 *      the session is live. Does NOT open any new browser tab. (No
 *      "backup launch" anymore — the primary browser-launch path is
 *      already sandbox-aware via the `CODEX_SANDBOX=seatbelt` branch in
 *      `openChromelessWindow`, so every environment opens the right
 *      kind of window on the first try and a deferred duplicate tab
 *      would be noise either way.)
 *
 *   2. **Periodic heartbeat** every `heartbeatIntervalMs` (default 30s)
 *      when stdout is a pipe (`!process.stdout.isTTY` — agent/CI).
 *      Prints a single "still waiting" line so agent shell-tool
 *      wrappers (Codex exec sessions are the motivating case) see
 *      fresh output and don't classify the blocked Node process as
 *      idle. Suppressed on interactive TTYs to keep human terminals
 *      clean.
 *
 * Returns a `cleanup()` callback. Callers must invoke it in their
 * resolve/reject/timeout path (typically `finally`).
 */
export function attachBrowserHandoffKeepalive(opts) {
    const redisplayDelay = opts.redisplayDelayMs ?? DEFAULT_REDISPLAY_DELAY_MS;
    const heartbeatInterval = opts.heartbeatIntervalMs ?? DEFAULT_HEARTBEAT_INTERVAL_MS;
    const startedAt = Date.now();
    const done = () => opts.loopback.done || opts.ticket.done;
    const redisplayTimer = setTimeout(() => {
        if (done()) {
            return;
        }
        console.log("");
        console.log("Still waiting for browser handoff. If the browser didn't open,");
        console.log("copy this URL into any browser — this session is still valid:");
        console.log("");
        console.log(`  ${opts.browserUrl}`);
        console.log("");
    }, redisplayDelay);
    // Only pulse on pipes: agent shell tools read stdout incrementally
    // and can classify a stale stream as "idle / ready to move on." A
    // TTY-attached human doesn't need the noise.
    const heartbeat = process.stdout.isTTY
        ? null
        : setInterval(() => {
            if (done()) {
                return;
            }
            const elapsed = Math.round((Date.now() - startedAt) / 1000);
            console.log(`[bitfab] waiting for browser handoff... ${elapsed}s elapsed, session still live`);
        }, heartbeatInterval);
    return () => {
        clearTimeout(redisplayTimer);
        if (heartbeat) {
            clearInterval(heartbeat);
        }
    };
}
/**
 * High-level flow: open a browser window to `buildUrl(...)` and race the
 * loopback and ticket channels until one delivers a value of type `T`. Tears
 * down both channels in a `finally` block.
 *
 * Use this for any browser → CLI hand-off that doesn't need extras like
 * interactive paste, SSH-specific hints, or short-circuit timeouts — e.g.
 * dataset selection, grader editing, experiment configuration. The login
 * flow uses `startBrowserHandoff` directly to layer on its extras.
 */
export async function runBrowserHandoff(opts) {
    const { loopback, ticket, browserUrl, timeoutMs } = await startBrowserHandoff(opts);
    if (!loopback.available && !ticket.available) {
        loopback.close();
        ticket.close();
        throw new Error(`Neither loopback nor ticket channel is available — can't complete the browser hand-off. Loopback: ${loopback.bindError?.message ?? "ok"}; ticket: ${ticket.startError?.message ?? "ok"}.`);
    }
    console.log("\nOpening your browser...");
    console.log(`  ${browserUrl}`);
    console.log("");
    console.log("(If the browser didn't open, visit the URL above manually.)");
    openChromelessWindow(browserUrl);
    const stopKeepalive = attachBrowserHandoffKeepalive({
        loopback,
        ticket,
        browserUrl,
    });
    const timeoutPromise = new Promise((_resolve, reject) => {
        setTimeout(() => reject(new Error(`Browser hand-off timed out after ${Math.round(timeoutMs / 1000)}s.`)), timeoutMs);
    });
    const channelAvailability = {
        loopbackAvailable: loopback.available,
        ticketAvailable: ticket.available,
    };
    const racers = [];
    if (loopback.available) {
        racers.push(loopback.promise.then((value) => ({
            value,
            wonBy: "loopback",
            ...channelAvailability,
        })));
    }
    if (ticket.available) {
        racers.push(ticket.promise.then((value) => ({
            value,
            wonBy: "ticket",
            ...channelAvailability,
        })));
    }
    try {
        return await Promise.race([...racers, timeoutPromise]);
    }
    finally {
        stopKeepalive();
        loopback.close();
        ticket.close();
    }
}
