// bitfab-plugin-lib/dist/browser.js
import { execSync, spawn } from "node:child_process";
import fs from "node:fs";
import os from "node:os";
function getChromiumBrowsers() {
    const platform = os.platform();
    if (platform === "darwin") {
        return [
            {
                name: "Google Chrome",
                defaultIds: ["com.google.chrome"],
                binaryPaths: [
                    "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome",
                ],
            },
            {
                name: "Chromium",
                defaultIds: ["org.chromium.chromium"],
                binaryPaths: ["/Applications/Chromium.app/Contents/MacOS/Chromium"],
            },
            {
                name: "Brave",
                defaultIds: ["com.brave.browser"],
                binaryPaths: [
                    "/Applications/Brave Browser.app/Contents/MacOS/Brave Browser",
                ],
            },
            {
                name: "Microsoft Edge",
                defaultIds: ["com.microsoft.edgemac"],
                binaryPaths: [
                    "/Applications/Microsoft Edge.app/Contents/MacOS/Microsoft Edge",
                ],
            },
        ];
    }
    if (platform === "linux") {
        return [
            {
                name: "Google Chrome",
                defaultIds: ["google-chrome.desktop", "google-chrome-stable.desktop"],
                binaryPaths: ["google-chrome", "google-chrome-stable"],
            },
            {
                name: "Chromium",
                defaultIds: ["chromium-browser.desktop", "chromium.desktop"],
                binaryPaths: ["chromium-browser", "chromium"],
            },
            {
                name: "Brave",
                defaultIds: ["brave-browser.desktop"],
                binaryPaths: ["brave-browser", "brave"],
            },
            {
                name: "Microsoft Edge",
                defaultIds: ["microsoft-edge.desktop", "microsoft-edge-stable.desktop"],
                binaryPaths: ["microsoft-edge", "microsoft-edge-stable"],
            },
        ];
    }
    if (platform === "win32") {
        return [
            {
                name: "Google Chrome",
                defaultIds: ["chromehtml"],
                binaryPaths: [
                    "C:\\Program Files\\Google\\Chrome\\Application\\chrome.exe",
                    "C:\\Program Files (x86)\\Google\\Chrome\\Application\\chrome.exe",
                ],
            },
            {
                name: "Brave",
                defaultIds: ["bravehtml"],
                binaryPaths: [
                    "C:\\Program Files\\BraveSoftware\\Brave-Browser\\Application\\brave.exe",
                    "C:\\Program Files (x86)\\BraveSoftware\\Brave-Browser\\Application\\brave.exe",
                ],
            },
            {
                name: "Microsoft Edge",
                defaultIds: ["msedgehtm"],
                binaryPaths: [
                    "C:\\Program Files\\Microsoft\\Edge\\Application\\msedge.exe",
                    "C:\\Program Files (x86)\\Microsoft\\Edge\\Application\\msedge.exe",
                ],
            },
        ];
    }
    return [];
}
function findExistingBinary(paths) {
    for (const candidate of paths) {
        if (candidate.includes("/") || candidate.includes("\\")) {
            if (fs.existsSync(candidate)) {
                return candidate;
            }
        }
        else {
            try {
                execSync(`which ${candidate}`, { stdio: "ignore" });
                return candidate;
            }
            catch {
                // not found on PATH
            }
        }
    }
    return null;
}
function getDefaultBrowserId() {
    const platform = os.platform();
    try {
        if (platform === "darwin") {
            const plistPath = `${os.homedir()}/Library/Preferences/com.apple.LaunchServices/com.apple.launchservices.secure.plist`;
            const json = execSync(`plutil -convert json -o - "${plistPath}"`, {
                stdio: ["ignore", "pipe", "ignore"],
                encoding: "utf-8",
                timeout: 2000,
            });
            const data = JSON.parse(json);
            const handler = data.LSHandlers?.find((h) => h.LSHandlerURLScheme === "http");
            return handler?.LSHandlerRoleAll?.toLowerCase() ?? null;
        }
        if (platform === "linux") {
            const result = execSync("xdg-settings get default-web-browser", {
                stdio: ["ignore", "pipe", "ignore"],
                encoding: "utf-8",
                timeout: 2000,
            });
            return result.trim().toLowerCase() || null;
        }
        if (platform === "win32") {
            const result = execSync(`reg query "HKCU\\Software\\Microsoft\\Windows\\Shell\\Associations\\UrlAssociations\\http\\UserChoice" /v ProgId`, {
                stdio: ["ignore", "pipe", "ignore"],
                encoding: "utf-8",
                timeout: 2000,
            });
            const match = result.match(/ProgId\s+REG_SZ\s+(\S+)/);
            return match?.[1]?.toLowerCase() ?? null;
        }
    }
    catch {
        return null;
    }
    return null;
}
export function matchChromiumBrowser(defaultId, browsers) {
    if (!defaultId) {
        return null;
    }
    return (browsers.find((b) => b.defaultIds.some((id) => defaultId.includes(id))) ??
        null);
}
const SCREEN_FRACTION = 0.95;
// Chrome's --window-size / --window-position use logical (CSS) pixels in a
// top-left origin coordinate system. Return the work area (screen minus
// menubar/dock/taskbar), not the raw frame, so centering and sizing stay
// inside the usable region.
function getWorkArea() {
    try {
        const platform = os.platform();
        if (platform === "darwin") {
            // NSScreen.visibleFrame is the usable area in Cocoa's bottom-left
            // origin. Convert to Chrome's top-left origin using the full frame.
            const out = execSync(`osascript -l JavaScript -e 'ObjC.import("AppKit"); const s = $.NSScreen.mainScreen; const f = s.frame; const v = s.visibleFrame; const topY = f.size.height - (v.origin.y + v.size.height); JSON.stringify([v.origin.x, topY, v.size.width, v.size.height])'`, {
                stdio: ["ignore", "pipe", "ignore"],
                encoding: "utf-8",
                timeout: 3000,
            });
            const parsed = JSON.parse(out.trim());
            if (parsed.length === 4) {
                return {
                    x: parsed[0],
                    y: parsed[1],
                    width: parsed[2],
                    height: parsed[3],
                };
            }
        }
        else if (platform === "linux") {
            // xdpyinfo reports the full X11 screen dimensions — panels/docks on
            // Linux are window-manager specific and not easily queried, so treat
            // the whole screen as the work area.
            const out = execSync("xdpyinfo | grep dimensions", {
                stdio: ["ignore", "pipe", "ignore"],
                encoding: "utf-8",
                timeout: 3000,
            });
            const match = out.match(/(\d+)x(\d+)/);
            if (match) {
                return { x: 0, y: 0, width: Number(match[1]), height: Number(match[2]) };
            }
        }
        else if (platform === "win32") {
            // WorkingArea excludes the taskbar; Bounds includes it. We want the
            // working area so the window doesn't overlap the taskbar.
            const out = execSync(`powershell -NoProfile -Command "Add-Type -AssemblyName System.Windows.Forms; $a=[System.Windows.Forms.Screen]::PrimaryScreen.WorkingArea; $a.X.ToString() + ',' + $a.Y.ToString() + ',' + $a.Width.ToString() + ',' + $a.Height.ToString()"`, {
                stdio: ["ignore", "pipe", "ignore"],
                encoding: "utf-8",
                timeout: 3000,
            });
            const match = out.match(/(-?\d+),(-?\d+),(\d+),(\d+)/);
            if (match) {
                return {
                    x: Number(match[1]),
                    y: Number(match[2]),
                    width: Number(match[3]),
                    height: Number(match[4]),
                };
            }
        }
    }
    catch {
        // detection unavailable
    }
    return null;
}
export function computeWindowArgs(area) {
    if (!area) {
        return [];
    }
    const w = Math.floor(area.width * SCREEN_FRACTION);
    const h = Math.floor(area.height * SCREEN_FRACTION);
    const x = area.x + Math.floor((area.width - w) / 2);
    const y = area.y + Math.floor((area.height - h) / 2);
    return [`--window-size=${w},${h}`, `--window-position=${x},${y}`];
}
function getWindowSizeArgs() {
    return computeWindowArgs(getWorkArea());
}
// Spawn without a shell so URLs/paths containing spaces, quotes, or other
// shell metacharacters can't be reinterpreted as command syntax. Detach +
// unref so the child outlives this process and doesn't hold the event loop.
function spawnDetached(command, args) {
    spawn(command, [...args], { detached: true, stdio: "ignore" }).unref();
}
function openOsDefault(url) {
    const platform = os.platform();
    if (platform === "darwin") {
        spawnDetached("open", [url]);
    }
    else if (platform === "win32") {
        // `start` is a cmd builtin, not an executable. The empty "" is the window
        // title placeholder — without it, `start` interprets the URL as the title.
        spawnDetached("cmd", ["/c", "start", "", url]);
    }
    else {
        spawnDetached("xdg-open", [url]);
    }
}
/**
 * Open a URL in a chromeless window (no address bar, tabs, or menu) via
 * Chrome/Chromium/Brave/Edge's `--app=` flag.
 *
 * Order of preference:
 *   1. If the user's default browser is Chromium-family and installed, open
 *      it with `--app=` — chromeless AND uses the profile where the user is
 *      already signed in.
 *   2. If the default is a non-Chromium browser (Firefox, Safari, etc.), open
 *      the OS default — tabbed, but respects the user's sign-in state.
 *   3. If default detection fails, try any Chromium-family binary for
 *      chromeless; otherwise fall back to the OS default.
 */
export function openChromelessWindow(url) {
    const browsers = getChromiumBrowsers();
    const defaultId = getDefaultBrowserId();
    const defaultChromium = matchChromiumBrowser(defaultId, browsers);
    const sizeArgs = getWindowSizeArgs();
    if (defaultChromium) {
        const binary = findExistingBinary(defaultChromium.binaryPaths);
        if (binary) {
            spawnDetached(binary, [`--app=${url}`, ...sizeArgs]);
            return;
        }
    }
    if (defaultId !== null && defaultChromium === null) {
        openOsDefault(url);
        return;
    }
    for (const browser of browsers) {
        const binary = findExistingBinary(browser.binaryPaths);
        if (binary) {
            spawnDetached(binary, [`--app=${url}`, ...sizeArgs]);
            return;
        }
    }
    openOsDefault(url);
}
