// bitfab-plugin-lib/dist/index.js
export { closeStudio, navigateStudio, pollAgentSessionEvents, pushAgentSessionEvent, } from "./agentSessionChannel.js";
export { findOpenPort, runBrowserHandoff, startBrowserHandoff, } from "./browserHandoff.js";
export { formatDevBuildBannerLines, formatDevBuildStatusLine, formatRelativeTime, getDevBuildSummary, hashSrcDir, isDevBuildActive, readBuildInfo, } from "./buildInfo.js";
export { captureFromHook } from "./chatSessions/index.js";
export { buildBitfabClientHeaders, buildBitfabRequestHeaders, } from "./clientHeaders.js";
export { runLogin } from "./commands/login.js";
export { runLogout } from "./commands/logout.js";
export { runOpenExperiments } from "./commands/openExperiments.js";
export { openStudio, runOpenStudio } from "./commands/openStudio.js";
export { runOpenTracePlan } from "./commands/openTracePlan.js";
export { runPersistReplayLabels } from "./commands/persistReplayLabels.js";
export { runPushActivity } from "./commands/pushActivity.js";
export { runStartDataset } from "./commands/startDataset.js";
export { runStartTemplatePreview } from "./commands/startTemplatePreview.js";
export { runStatus } from "./commands/status.js";
export { runTrackSkill } from "./commands/trackSkill.js";
export { parseUpdateMode, runUpdate } from "./commands/update.js";
export { runWaitForTrace } from "./commands/waitForTrace.js";
export { getConfig, getSessionLogConsent, hasCredentials, setSessionLogConsent, } from "./config.js";
export { startHandoffTicketChannel } from "./handoffTicketChannel.js";
export { createHandoffTicket, pollHandoffTicket } from "./handoffTickets.js";
export { runCaptureHook } from "./hooks/captureHook.js";
export { runSessionStart } from "./hooks/sessionStart.js";
export { extractSkillFromPrompt, runUserPromptSubmit, } from "./hooks/userPromptSubmit.js";
export { getCurrentVersion, getInstalledSdks } from "./installedSdk.js";
export { detectClaudeInstallScopes } from "./installScope.js";
export { startMcpServer } from "./mcpServer.js";
export { generatePkce } from "./pkce.js";
export { checkForSdkUpdates, checkForSdkUpdatesOffline, formatSdkUpdateLine, getBakedSdkVersion, getLatestSdkVersion, isSdkVersionNewer, } from "./sdkUpdates.js";
export { BITFAB_SKILLS } from "./skills.js";
export { checkForUpdate, detectLegacyInstall, formatUpdateMessage, legacyMigrationMessage, } from "./updates.js";
export { detectWorkspaces } from "./workspaces.js";
