// bitfab-plugin-lib/dist/installedSdk.js
import fs from "node:fs";
import path from "node:path";
import { detectWorkspaces } from "./workspaces.js";
const TS_PACKAGE = "bitfab";
const PY_PACKAGE = "bitfab-py";
const RB_PACKAGE = "bitfab";
const GO_MODULE = "github.com/Project-White-Rabbit/bitfab-go";
function readFileSafe(filePath) {
    try {
        return fs.readFileSync(filePath, "utf-8");
    }
    catch {
        return null;
    }
}
function stripRange(raw) {
    // Extract the first version-shaped token, preserving pre-release and build
    // metadata (e.g. "^0.11.0-beta.1" → "0.11.0-beta.1"). Previously this just
    // stripped non-digit/dot chars, which corrupted "-beta.1" into "0.110.1".
    // Accepts 2+ components so Ruby's `~> 0.10` also parses.
    const match = raw.trim().match(/\d+(?:\.\d+){1,}(?:[-+][\w.+-]+)?/);
    return match ? match[0] : "";
}
function readTsDeclared(cwd) {
    const manifestPath = path.join(cwd, "package.json");
    const content = readFileSafe(manifestPath);
    if (!content) {
        return null;
    }
    try {
        const pkg = JSON.parse(content);
        const raw = pkg.dependencies?.[TS_PACKAGE] ??
            pkg.devDependencies?.[TS_PACKAGE] ??
            pkg.peerDependencies?.[TS_PACKAGE];
        if (!raw) {
            return null;
        }
        return { version: stripRange(raw) || null, path: manifestPath };
    }
    catch {
        return null;
    }
}
function readTsFromPnpm(cwd) {
    const p = path.join(cwd, "pnpm-lock.yaml");
    const content = readFileSafe(p);
    if (!content) {
        return null;
    }
    const match = content.match(/['"/\s]bitfab@(\d+\.\d+\.\d+(?:[-+][\w.]+)?)/);
    return match ? { version: match[1], path: p } : null;
}
function readTsFromNpm(cwd) {
    const p = path.join(cwd, "package-lock.json");
    const content = readFileSafe(p);
    if (!content) {
        return null;
    }
    try {
        const pkg = JSON.parse(content);
        const v = pkg.packages?.[`node_modules/${TS_PACKAGE}`]?.version ??
            pkg.dependencies?.[TS_PACKAGE]?.version;
        return v ? { version: v, path: p } : null;
    }
    catch {
        return null;
    }
}
function readTsFromYarn(cwd) {
    const p = path.join(cwd, "yarn.lock");
    const content = readFileSafe(p);
    if (!content) {
        return null;
    }
    const match = content.match(/^(?:"?bitfab@[^",:\n]+"?(?:,\s*"?bitfab@[^",:\n]+"?)*):\s*\n(?:[^\n]*\n)*?\s+version\s+"([^"]+)"/m);
    return match ? { version: match[1], path: p } : null;
}
function readTsFromBun(cwd) {
    const p = path.join(cwd, "bun.lock");
    const content = readFileSafe(p);
    if (!content) {
        return null;
    }
    const match = content.match(/"bitfab@(\d+\.\d+\.\d+(?:[-+][\w.]+)?)"/);
    return match ? { version: match[1], path: p } : null;
}
function readTsFromNodeModules(cwd) {
    const p = path.join(cwd, "node_modules", TS_PACKAGE, "package.json");
    const content = readFileSafe(p);
    if (!content) {
        return null;
    }
    try {
        const pkg = JSON.parse(content);
        return pkg.version ? { version: pkg.version, path: p } : null;
    }
    catch {
        return null;
    }
}
function readTsResolved(dir, monorepoRoot) {
    const local = readTsFromPnpm(dir) ??
        readTsFromNpm(dir) ??
        readTsFromYarn(dir) ??
        readTsFromBun(dir) ??
        readTsFromNodeModules(dir);
    if (local || dir === monorepoRoot) {
        return local;
    }
    return (readTsFromPnpm(monorepoRoot) ??
        readTsFromNpm(monorepoRoot) ??
        readTsFromYarn(monorepoRoot) ??
        readTsFromBun(monorepoRoot) ??
        readTsFromNodeModules(monorepoRoot));
}
function readPyDeclared(cwd) {
    const pyproject = path.join(cwd, "pyproject.toml");
    const pypContent = readFileSafe(pyproject);
    if (pypContent?.includes(PY_PACKAGE)) {
        const m = pypContent.match(new RegExp(`${PY_PACKAGE}\\s*[=~><^]*\\s*["']?([0-9][^"'\\s,\\]]*)?`));
        return {
            version: m?.[1] ? stripRange(m[1]) || null : null,
            path: pyproject,
        };
    }
    const reqPath = path.join(cwd, "requirements.txt");
    const reqContent = readFileSafe(reqPath);
    if (reqContent) {
        for (const rawLine of reqContent.split("\n")) {
            const line = rawLine.trim();
            if (!line || line.startsWith("#")) {
                continue;
            }
            const m = line.match(new RegExp(`^${PY_PACKAGE}\\s*([=~><!]+)\\s*([0-9][^\\s;]*)`));
            if (m) {
                return { version: stripRange(m[2]) || null, path: reqPath };
            }
            if (line === PY_PACKAGE) {
                return { version: null, path: reqPath };
            }
        }
    }
    return null;
}
function readPyTomlLockVersion(lockPath) {
    const content = readFileSafe(lockPath);
    if (!content) {
        return null;
    }
    const blocks = content.split(/^\[\[package\]\]\s*$/m);
    for (const block of blocks) {
        if (new RegExp(`^\\s*name\\s*=\\s*["']${PY_PACKAGE}["']\\s*$`, "m").test(block)) {
            const m = block.match(/^\s*version\s*=\s*["']([^"']+)["']/m);
            if (m) {
                return { version: m[1], path: lockPath };
            }
        }
    }
    return null;
}
function readPyFromPipfileLock(cwd) {
    const p = path.join(cwd, "Pipfile.lock");
    const content = readFileSafe(p);
    if (!content) {
        return null;
    }
    try {
        const data = JSON.parse(content);
        const raw = data.default?.[PY_PACKAGE]?.version ?? data.develop?.[PY_PACKAGE]?.version;
        return raw ? { version: stripRange(raw) || null, path: p } : null;
    }
    catch {
        return null;
    }
}
function readPyResolved(dir, monorepoRoot) {
    const local = readPyTomlLockVersion(path.join(dir, "uv.lock")) ??
        readPyTomlLockVersion(path.join(dir, "poetry.lock")) ??
        readPyFromPipfileLock(dir);
    if (local || dir === monorepoRoot) {
        return local;
    }
    return (readPyTomlLockVersion(path.join(monorepoRoot, "uv.lock")) ??
        readPyTomlLockVersion(path.join(monorepoRoot, "poetry.lock")) ??
        readPyFromPipfileLock(monorepoRoot));
}
function readRbDeclared(cwd) {
    const gemfilePath = path.join(cwd, "Gemfile");
    const content = readFileSafe(gemfilePath);
    if (!content) {
        return null;
    }
    const m = content.match(new RegExp(`gem\\s+["']${RB_PACKAGE}["'](?:\\s*,\\s*["']([^"']+)["'])?`));
    if (!m) {
        return null;
    }
    return {
        version: m[1] ? stripRange(m[1]) || null : null,
        path: gemfilePath,
    };
}
function readRbResolved(cwd) {
    const p = path.join(cwd, "Gemfile.lock");
    const content = readFileSafe(p);
    if (!content) {
        return null;
    }
    const m = content.match(new RegExp(`^\\s{4}${RB_PACKAGE}\\s+\\(([^)]+)\\)`, "m"));
    return m ? { version: stripRange(m[1]) || null, path: p } : null;
}
function readGoResolved(cwd) {
    const p = path.join(cwd, "go.mod");
    const content = readFileSafe(p);
    if (!content) {
        return null;
    }
    const m = content.match(new RegExp(`${GO_MODULE.replace(/\./g, "\\.")}\\s+v(\\d[\\w.\\-+]*)`));
    return m ? { version: m[1], path: p } : null;
}
function detectAtDir(dir, monorepoRoot, workspacePath) {
    const result = [];
    const tsDeclared = readTsDeclared(dir);
    const tsResolved = tsDeclared ? readTsResolved(dir, monorepoRoot) : null;
    if (tsDeclared) {
        result.push({
            language: "typescript",
            packageName: TS_PACKAGE,
            declaredVersion: tsDeclared.version,
            resolvedVersion: tsResolved?.version ?? null,
            manifestPath: tsDeclared.path,
            lockfilePath: tsResolved?.path ?? null,
            workspacePath,
        });
    }
    const pyDeclared = readPyDeclared(dir);
    const pyResolved = pyDeclared ? readPyResolved(dir, monorepoRoot) : null;
    if (pyDeclared) {
        result.push({
            language: "python",
            packageName: PY_PACKAGE,
            declaredVersion: pyDeclared.version,
            resolvedVersion: pyResolved?.version ?? null,
            manifestPath: pyDeclared.path,
            lockfilePath: pyResolved?.path ?? null,
            workspacePath,
        });
    }
    const rbDeclared = readRbDeclared(dir);
    const rbResolved = rbDeclared ? readRbResolved(dir) : null;
    if (rbDeclared) {
        result.push({
            language: "ruby",
            packageName: RB_PACKAGE,
            declaredVersion: rbDeclared.version,
            resolvedVersion: rbResolved?.version ?? null,
            manifestPath: rbDeclared.path,
            lockfilePath: rbResolved?.path ?? null,
            workspacePath,
        });
    }
    const goResolved = readGoResolved(dir);
    if (goResolved) {
        result.push({
            language: "go",
            packageName: GO_MODULE,
            declaredVersion: null,
            resolvedVersion: goResolved.version,
            manifestPath: null,
            lockfilePath: goResolved.path,
            workspacePath,
        });
    }
    return result;
}
export function getInstalledSdks(rootDir) {
    const workspaces = detectWorkspaces(rootDir);
    const dirs = [
        { dir: rootDir, workspacePath: "." },
        ...workspaces.map((d) => ({
            dir: d,
            workspacePath: path.relative(rootDir, d) || ".",
        })),
    ];
    const seen = new Set();
    const results = [];
    for (const { dir, workspacePath } of dirs) {
        for (const sdk of detectAtDir(dir, rootDir, workspacePath)) {
            const key = `${workspacePath}\0${sdk.language}`;
            if (seen.has(key)) {
                continue;
            }
            seen.add(key);
            results.push(sdk);
        }
    }
    return results;
}
export function getCurrentVersion(sdk) {
    return sdk.resolvedVersion ?? sdk.declaredVersion;
}
