// bitfab-plugin-lib/dist/mcpProxy.js
export function errorResult(message) {
    return {
        content: [{ type: "text", text: message }],
        isError: true,
    };
}
export function parseSSEData(text) {
    const dataLine = text.split("\n").find((line) => line.startsWith("data: "));
    return dataLine ? dataLine.slice(6) : null;
}
export async function parseResponse(response) {
    const contentType = response.headers.get("content-type") ?? "";
    if (contentType.includes("text/event-stream")) {
        const text = await response.text();
        const data = parseSSEData(text);
        if (!data) {
            throw new Error("Empty SSE response");
        }
        return JSON.parse(data);
    }
    return response.json();
}
export class McpProxy {
    sessionId = null;
    loginHint;
    constructor(platform) {
        this.loginHint = platform.loginHint;
    }
    async fetch(config, method, params) {
        const headers = {
            "Content-Type": "application/json",
            Accept: "application/json, text/event-stream",
            Authorization: `Bearer ${config.apiKey}`,
        };
        if (this.sessionId) {
            headers["mcp-session-id"] = this.sessionId;
        }
        const response = await fetch(`${config.serviceUrl}/mcp`, {
            method: "POST",
            headers,
            body: JSON.stringify({ jsonrpc: "2.0", id: 1, method, params }),
        });
        const newSessionId = response.headers.get("mcp-session-id");
        if (newSessionId) {
            this.sessionId = newSessionId;
        }
        return response;
    }
    async ensureSession(config) {
        if (this.sessionId) {
            return;
        }
        const response = await this.fetch(config, "initialize", {
            protocolVersion: "2025-03-26",
            capabilities: {},
            clientInfo: { name: "bitfab-plugin", version: "1.0.0" },
        });
        if (!response.ok) {
            throw new Error(`Initialize failed (${response.status})`);
        }
        await parseResponse(response);
    }
    async toolCall(config, toolName, args) {
        if (!config.apiKey) {
            return errorResult(`Not authenticated. Run ${this.loginHint} to connect your Bitfab account.`);
        }
        try {
            await this.ensureSession(config);
        }
        catch (err) {
            const message = err instanceof Error ? err.message : String(err);
            return errorResult(`Failed to initialize Bitfab session: ${message}`);
        }
        let response;
        try {
            response = await this.fetch(config, "tools/call", {
                name: toolName,
                arguments: args,
            });
        }
        catch (err) {
            const message = err instanceof Error ? err.message : String(err);
            return errorResult(`Network error connecting to Bitfab: ${message}`);
        }
        // Stale session — clear and retry once
        if (response.status === 404) {
            this.sessionId = null;
            try {
                await this.ensureSession(config);
                response = await this.fetch(config, "tools/call", {
                    name: toolName,
                    arguments: args,
                });
            }
            catch (err) {
                const message = err instanceof Error ? err.message : String(err);
                return errorResult(`Failed to reconnect to Bitfab: ${message}`);
            }
        }
        if (!response.ok) {
            return errorResult(`Bitfab API error (${response.status}): ${await response.text().catch(() => "unknown")}`);
        }
        let data;
        try {
            data = await parseResponse(response);
        }
        catch {
            return errorResult("Failed to parse response from Bitfab API");
        }
        if (data.error) {
            return errorResult(`Bitfab error: ${data.error.message}`);
        }
        return data.result ?? errorResult("Empty response from Bitfab API");
    }
    getSessionId() {
        return this.sessionId;
    }
    clearSession() {
        this.sessionId = null;
    }
}
