// bitfab-plugin-lib/dist/mcpServer.js
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { StdioServerTransport } from "@modelcontextprotocol/sdk/server/stdio.js";
import { z } from "zod";
import { McpProxy } from "./mcpProxy.js";
const ENABLE_GRADER_TOOLS = false;
const DATE_PATTERN = /^\d{4}-\d{2}-\d{2}$/;
// Several MCP clients (Claude Code, Claude Desktop, some SSE clients)
// stringify non-string tool arguments before sending them in `tools/call` —
// numbers, booleans, and arrays all arrive as raw strings even though the
// advertised JSON Schema declares the right types. This is a known upstream
// bug tracked in:
//   - anthropics/claude-code#3966
//   - anthropics/claude-code#5504
//   - anthropics/claude-code#11359
//   - anthropics/claude-code#26094
// Most production MCP servers (Notion, Azure DevOps, FastMCP, ...) defend
// themselves with z.coerce / z.preprocess — we do the same. The advertised
// schema stays strict so the LLM still sees the right types. Use it as
// `z.preprocess(parseJsonString, z.number()...)`. Malformed JSON is left
// as-is so zod can emit a clear "Invalid arguments" error.
//
// This helper is duplicated in bitfab-web/src/app/mcp/route.ts so each MCP
// server in the chain defends itself independently — the plugin handles
// args from stdio clients, the server handles args from any direct HTTP
// callers. Keep them in sync.
function parseJsonString(val) {
    if (typeof val !== "string") {
        return val;
    }
    try {
        return JSON.parse(val);
    }
    catch {
        return val;
    }
}
export function createMcpServer(platform, getConfig, pluginVersion) {
    const proxy = new McpProxy(platform, pluginVersion);
    function proxyToolCall(toolName, args) {
        return proxy.toolCall(getConfig(), toolName, args);
    }
    const server = new McpServer({
        name: "Bitfab",
        version: pluginVersion,
    });
    server.registerTool("get_bitfab_api_key", {
        description: "Get your Bitfab API key for SDK initialization. Returns the plaintext key to use in environment variables. The SDK setup reference — install commands, initialization, instrumentation patterns — lives at https://docs.bitfab.ai (TypeScript: /typescript-sdk, Python: /python-sdk, Ruby: /ruby-sdk, Go: /go-sdk); read the relevant page there when setting up.",
        inputSchema: {},
    }, async () => proxyToolCall("get_bitfab_api_key", {}));
    server.registerTool("list_trace_functions", {
        description: "List all traced functions in your organization. Use this to discover which functions are instrumented with the Bitfab SDK.",
        inputSchema: {},
    }, async () => proxyToolCall("list_trace_functions", {}));
    if (ENABLE_GRADER_TOOLS) {
        server.registerTool("get_trace_function_diagnostics", {
            description: "Get detailed failure diagnostics for a specific traced function — includes evaluation criteria, failure summaries, and recent individual failures with trace IDs. Use this to understand exactly why traces are failing so you can make targeted code fixes. You can call this directly if you already know the traceFunctionKey from the codebase, or use list_trace_functions first to discover available keys.",
            inputSchema: {
                traceFunctionKey: z
                    .string()
                    .describe("The function key to get diagnostics for"),
                graderId: z
                    .string()
                    .optional()
                    .describe("Optional: scope to a single grader ID"),
            },
        }, async ({ traceFunctionKey, graderId }) => proxyToolCall("get_trace_function_diagnostics", {
            traceFunctionKey,
            graderId,
        }));
        server.registerTool("create_grader", {
            description: "Create a failure-mode grader specification for a traced function. Use this after identifying a failure pattern in the user's AI agent — the traceFunctionKey must match a function instrumented with the Bitfab SDK. The grader records what to evaluate (evaluationFocus) and optionally how to distinguish pass/fail cases. Call list_graders first to check if a similar grader already exists.",
            inputSchema: {
                traceFunctionKey: z
                    .string()
                    .describe("The trace function key from the instrumented source code"),
                name: z
                    .string()
                    .describe("A short, descriptive name for this grader (e.g. 'Hallucination check')"),
                evaluationFocus: z
                    .string()
                    .describe("What this grader evaluates — describe the failure mode or quality dimension"),
                passCriteria: z
                    .string()
                    .optional()
                    .describe("What a passing result looks like"),
                failCriteria: z
                    .string()
                    .optional()
                    .describe("What a failing result looks like"),
                archived: z
                    .preprocess(parseJsonString, z.boolean())
                    .optional()
                    .describe("Set to true to archive (soft-delete) this grader"),
            },
        }, async ({ traceFunctionKey, name, evaluationFocus, passCriteria, failCriteria, archived, }) => proxyToolCall("create_grader", {
            traceFunctionKey,
            name,
            evaluationFocus,
            passCriteria,
            failCriteria,
            archived,
        }));
        server.registerTool("list_graders", {
            description: 'List all graders and their criteria for a traced function. Call this at the start of a session to understand existing failure-mode specifications before making improvements, or before creating a new grader to avoid duplicates. IMPORTANT: When working on a user\'s AI agent codebase, always specify a status filter — use status="active" to see graders currently evaluating traces, status="live" for production-ready graders, or status="specification" for grader specs that haven\'t been trained yet. Without a status filter, you\'ll get all graders including archived and candidate ones, which adds noise.',
            inputSchema: {
                traceFunctionKey: z
                    .string()
                    .describe("The trace function key to list graders for"),
                status: z
                    .string()
                    .optional()
                    .describe("Filter by grader status: candidate, active, training, live, archived, or specification"),
                type: z
                    .string()
                    .optional()
                    .describe("Filter by grader type: code_javascript or simforge_llm_as_judge"),
                verbose: z
                    .preprocess(parseJsonString, z.boolean())
                    .optional()
                    .describe("Show full criteria details for all graders. Specification graders always show full details regardless of this flag."),
            },
        }, async ({ traceFunctionKey, status, type, verbose }) => proxyToolCall("list_graders", {
            traceFunctionKey,
            status,
            type,
            verbose,
        }));
    }
    server.registerTool("search_traces", {
        description: 'Search and filter traces for a traced function. Returns matching traces with IDs, status, timestamps, total duration (ms), total token usage (in/out/cached/total), model, and output preview — all when available. Replay traces (from test runs) carry the same totals. Use read_traces with scope="full" for per-span duration, tokens, time-to-first-token, and model. Supports keyword search, date range, status filter, and regex matching. Use drillDown to refine previous results.',
        inputSchema: {
            traceFunctionKey: z
                .string()
                .describe("The trace function key to search"),
            searchQuery: z.string().optional().describe("Full-text keyword search"),
            status: z
                .enum(["pending", "completed", "failed"])
                .optional()
                .describe("Filter by trace status"),
            createdAfter: z
                .string()
                .regex(DATE_PATTERN, "Must be YYYY-MM-DD format")
                .optional()
                .describe("Only traces after this date (YYYY-MM-DD)"),
            createdBefore: z
                .string()
                .regex(DATE_PATTERN, "Must be YYYY-MM-DD format")
                .optional()
                .describe("Only traces before this date (YYYY-MM-DD)"),
            traceFilterRegex: z
                .string()
                .optional()
                .describe("Regex to filter trace output"),
            limit: z
                .preprocess(parseJsonString, z.number().min(1).max(50))
                .optional()
                .describe("Max traces to return (default 10, max 50)"),
            drillDown: z
                .preprocess(parseJsonString, z.boolean())
                .optional()
                .describe("If true, merge with previous search filters to narrow results"),
            traceIds: z
                .preprocess(parseJsonString, z.array(z.uuid()))
                .optional()
                .describe("Specific trace IDs to fetch"),
            environment: z
                .string()
                .optional()
                .describe("Filter by environment (e.g., production, staging, development, preview). Omit to search across all environments."),
            labelSource: z
                .enum(["human", "agent"])
                .optional()
                .describe('Filter by who authored the label: "human" for human-authored labels only, "agent" for agent-authored labels only. Omit to include any source. This is a pure authorship filter — for trust/validation, use the separate `validated` parameter.'),
            validated: z
                .preprocess(parseJsonString, z.boolean())
                .optional()
                .describe('If true, only return traces with a validated label — a label that is either human-authored OR an agent-authored label that a human has approved (`approvedAt` is set). Use this when you want trustworthy/reviewed labels. Combine with `labelSource: "agent"` to get only approved-agent labels (excluding pure-human labels). Omit (or pass false) to include unapproved labels too.'),
            labelResult: z
                .preprocess(parseJsonString, z.boolean())
                .optional()
                .describe("Filter by label result: true for passing labels, false for failing labels"),
        },
    }, async (args) => proxyToolCall("search_traces", args));
    server.registerTool("read_traces", {
        description: 'Read one or more traces by ID. Includes inputs, outputs, status, trace-level duration (ms), token usage (in/out/cached/total), and model (when available). Use these to compare latency and cost between a trace and its replay. By default returns span summaries; set scope="full" to get full span details (input, output, reasoning, context, errors) plus per-span duration, tokens, time-to-first-token, and model when available.',
        inputSchema: {
            traceIds: z
                .preprocess(parseJsonString, z.array(z.uuid()).min(1).max(10))
                .describe("Trace IDs to read (1-10)"),
            scope: z
                .enum(["summary", "full"])
                .optional()
                .describe('Level of span detail: "summary" (default) for one-line summaries, "full" for complete input/output/reasoning/content/errors'),
        },
    }, async ({ traceIds, scope }) => proxyToolCall("read_traces", { traceIds, scope }));
    server.registerTool("update_agent_labels", {
        description: 'Set or archive the agent\'s pass/fail verdict on one or more traces (`labelSource="agent"`). Use this AFTER you have read the traces with read_traces and decided yourself whether each one looks like a pass or a fail. To set a verdict, pass `label` (true=PASS, false=FAIL) and `annotation` (your reasoning, shown to the human reviewer in the labeling UI). To clear a previously-set agent verdict (e.g., you changed your mind or labeled the wrong trace), pass `archive: true` and omit label/annotation. New verdicts start unapproved; once a human approves one in the UI, the label joins the validated dataset (queryable via `search_traces` with `validated: true`). Archiving is non-destructive — the row is hidden from all reads but kept for audit, and you can immediately re-label the trace from scratch.',
        inputSchema: {
            labels: z
                .preprocess(parseJsonString, z
                .array(z.object({
                traceId: z.uuid().describe("The trace ID to update"),
                archive: z
                    .boolean()
                    .optional()
                    .describe("If true, archive the existing agent label for this trace and ignore label/annotation. Default false."),
                label: z
                    .boolean()
                    .optional()
                    .describe("true for PASS, false for FAIL. Required when archive is false/omitted."),
                annotation: z
                    .string()
                    .min(1)
                    .optional()
                    .describe("Your reasoning for the verdict — shown to the human reviewer in the labeling UI. Required when archive is false/omitted."),
            }))
                .min(1)
                .max(50))
                .describe("One update per trace (1-50)"),
        },
    }, async ({ labels }) => proxyToolCall("update_agent_labels", { labels }));
    server.registerTool("create_trace_plan", {
        description: "Post a tracing instrumentation plan and get back a URL the user opens to confirm or adjust. Send the full call tree (root + children + framework spans + ~10 surrounding callers above the root and ~10 surrounding callees below each leaf as `pure` context nodes) plus initial recommended captured node ids and any sample input/output stored on the relevant nodes. Pass `traceFunctionKey` so future Modify cycles for this same key can bootstrap from this plan via get_trace_plan. Returns a URL the user opens in the browser; after they confirm or abort there, call get_trace_plan to read the final captured set and act on it.",
        inputSchema: {
            language: z
                .enum(["python", "typescript", "ruby", "go"])
                .describe("Source language of the user's code"),
            tree: z
                .preprocess(parseJsonString, z.unknown())
                .describe("TracePlanTree: { rootId, nodes: { [id]: TraceNode } }. Each TraceNode has id, name, kind ('manual' | 'auto' | 'pure'), file, line, signature, parentId, childIds, plus optional framework, fields, sampleInput, sampleOutput. Include ~10 surrounding callers above the root and ~10 surrounding callees below each leaf as `pure` (uncaptured) context nodes so the user can see what's adjacent in the codebase when they edit in the UI."),
            capturedNodeIds: z
                .preprocess(parseJsonString, z.array(z.string()))
                .describe("Initial recommended captured set. Must reference ids present in tree.nodes. Must form a connected sub-tree (selecting any descendant implies its ancestors). Surrounding `pure` context nodes are not captured by default."),
            stats: z
                .preprocess(parseJsonString, z.unknown())
                .optional()
                .describe("Optional sample-run stats: { durationMs?, tokens?, cost? }"),
            traceFunctionKey: z
                .string()
                .min(1)
                .optional()
                .describe("Optional trace function key (the string passed to getFunction / get_function / bitfab_function / WithFunctionName). Persisted on the plan so Modify cycles can fetch this plan as the `before` tree via get_trace_plan with the same key."),
        },
    }, async ({ language, tree, capturedNodeIds, stats, traceFunctionKey }) => proxyToolCall("create_trace_plan", {
        language,
        tree,
        capturedNodeIds,
        stats,
        traceFunctionKey,
    }));
    server.registerTool("get_trace_plan", {
        description: "Read a trace plan. Two modes: pass `planId` to read a specific plan (use this after create_trace_plan + browser confirmation to learn whether the user confirmed, cancelled, or is still pending); or pass `traceFunctionKey` to fetch the latest *confirmed* plan for that key (use this at the start of a Modify cycle to bootstrap the `before` tree from the prior plan, so you don't have to re-derive it from the code). Returns the full plan including the tree as JSON.",
        inputSchema: {
            planId: z
                .string()
                .optional()
                .describe("The trace plan id returned by create_trace_plan."),
            traceFunctionKey: z
                .string()
                .optional()
                .describe("Trace function key. Returns the latest confirmed plan for this key in the caller's organization, or a 'no prior plan' message if none exists."),
        },
    }, async ({ planId, traceFunctionKey }) => proxyToolCall("get_trace_plan", { planId, traceFunctionKey }));
    return { server, proxy };
}
export async function startMcpServer(platform, getConfig, pluginVersion) {
    const { server } = createMcpServer(platform, getConfig, pluginVersion);
    const transport = new StdioServerTransport();
    await server.connect(transport);
}
