// bitfab-plugin-lib/dist/sdkUpdates.js
import { BAKED_SDK_VERSIONS } from "./bakedSdkVersions.js";
import { getCurrentVersion, getInstalledSdks, } from "./installedSdk.js";
const TS_PACKAGE = "bitfab";
const PY_PACKAGE = "bitfab-py";
const RB_PACKAGE = "bitfab";
const GO_MODULE = "github.com/Project-White-Rabbit/bitfab-go";
async function fetchJson(url) {
    try {
        const response = await fetch(url, { signal: AbortSignal.timeout(3000) });
        if (!response.ok) {
            return null;
        }
        return await response.json();
    }
    catch {
        return null;
    }
}
function encodeGoModulePath(modulePath) {
    return modulePath.replace(/[A-Z]/g, (ch) => `!${ch.toLowerCase()}`);
}
export async function getLatestSdkVersion(language) {
    if (language === "typescript") {
        const data = (await fetchJson(`https://registry.npmjs.org/${TS_PACKAGE}/latest`));
        return data?.version ?? null;
    }
    if (language === "python") {
        const data = (await fetchJson(`https://pypi.org/pypi/${PY_PACKAGE}/json`));
        return data?.info?.version ?? null;
    }
    if (language === "ruby") {
        const data = (await fetchJson(`https://rubygems.org/api/v1/gems/${RB_PACKAGE}.json`));
        return data?.version ?? null;
    }
    if (language === "go") {
        const data = (await fetchJson(`https://proxy.golang.org/${encodeGoModulePath(GO_MODULE)}/@latest`));
        return data?.Version ? data.Version.replace(/^v/, "") : null;
    }
    return null;
}
export function getBakedSdkVersion(language) {
    return BAKED_SDK_VERSIONS[language];
}
function parseVersion(v) {
    // Per semver: strip build metadata (after `+`) entirely — it doesn't
    // affect precedence. Then split off the pre-release tag (after first `-`).
    // parseInt("0-beta") returns 0 (not NaN), so we must strip these *before*
    // splitting on "." to do the numeric comparison.
    const noBuild = v.split("+")[0];
    const dash = noBuild.indexOf("-");
    if (dash === -1) {
        return { base: noBuild, prerelease: null };
    }
    return {
        base: noBuild.slice(0, dash),
        prerelease: noBuild.slice(dash + 1),
    };
}
export function isSdkVersionNewer(latest, current) {
    if (!current) {
        return false;
    }
    const l = parseVersion(latest);
    const c = parseVersion(current);
    const latestParts = l.base.split(".").map((n) => Number.parseInt(n, 10));
    const currentParts = c.base.split(".").map((n) => Number.parseInt(n, 10));
    for (let i = 0; i < 3; i++) {
        const a = latestParts[i] ?? 0;
        const b = currentParts[i] ?? 0;
        if (Number.isNaN(a) || Number.isNaN(b)) {
            return false;
        }
        if (a !== b) {
            return a > b;
        }
    }
    // Numeric parts equal: stable > pre-release per semver.
    // 0.11.0 is newer than 0.11.0-beta.1; 0.11.0-rc.2 is not newer than 0.11.0.
    // Build metadata (0.11.0+build.5) doesn't change precedence.
    return c.prerelease !== null && l.prerelease === null;
}
function toStatus(sdk, latest, source, remoteCheckFailed) {
    const current = getCurrentVersion(sdk);
    return {
        language: sdk.language,
        packageName: sdk.packageName,
        declaredVersion: sdk.declaredVersion,
        resolvedVersion: sdk.resolvedVersion,
        current,
        latest,
        latestSource: source,
        remoteCheckFailed,
        updateAvailable: latest !== null && isSdkVersionNewer(latest, current),
        manifestPath: sdk.manifestPath,
        lockfilePath: sdk.lockfilePath,
        workspacePath: sdk.workspacePath,
    };
}
/**
 * Synchronous, offline SDK update check. Compares the resolved version
 * (from lockfiles) against the version that was baked into the plugin at
 * build time. No network, no process spawning — safe to call on every
 * session start.
 */
export function checkForSdkUpdatesOffline(cwd) {
    const installed = getInstalledSdks(cwd);
    if (installed.length === 0) {
        return [];
    }
    return installed.map((sdk) => toStatus(sdk, getBakedSdkVersion(sdk.language), "baked", false));
}
/**
 * Full SDK update check: offline first, then a remote registry fetch.
 * - If the remote fetch returns a version, it replaces the baked value.
 * - If the remote fetch fails, remoteCheckFailed is set so the caller
 *   can punt to the agent (which can run the package manager's native
 *   outdated command to determine the latest).
 */
export async function checkForSdkUpdates(cwd) {
    const offline = checkForSdkUpdatesOffline(cwd);
    if (offline.length === 0) {
        return [];
    }
    return Promise.all(offline.map(async (status) => {
        const installed = {
            language: status.language,
            packageName: status.packageName,
            declaredVersion: status.declaredVersion,
            resolvedVersion: status.resolvedVersion,
            manifestPath: status.manifestPath,
            lockfilePath: status.lockfilePath,
            workspacePath: status.workspacePath,
        };
        const remote = await getLatestSdkVersion(status.language);
        if (remote === null) {
            return toStatus(installed, status.latest, "baked", true);
        }
        return toStatus(installed, remote, "remote", false);
    }));
}
export function formatSdkUpdateLine(status) {
    const current = status.current ?? "unknown";
    const prefix = status.workspacePath && status.workspacePath !== "."
        ? `${status.workspacePath} `
        : "";
    return `${prefix}${status.language} (${status.packageName}) ${current} → ${status.latest}`;
}
