// bitfab-plugin-lib/dist/workspaces.js
import fs from "node:fs";
import path from "node:path";
function exists(p) {
    try {
        fs.accessSync(p);
        return true;
    }
    catch {
        return false;
    }
}
function readFile(p) {
    try {
        return fs.readFileSync(p, "utf-8");
    }
    catch {
        return null;
    }
}
/**
 * Expand a simple glob pattern relative to rootDir. Supports single-segment
 * "*" matching; does not support "**" — patterns containing "**" return
 * empty, and the caller (detectWorkspaces) logs a warning so the user
 * knows we skipped it instead of silently getting no matches.
 */
function expandGlob(rootDir, pattern) {
    const trimmed = pattern.replace(/\/$/, "");
    if (trimmed.split("/").includes("**")) {
        console.warn(`[bitfab] Skipping workspace pattern "${pattern}" — "**" globs are not supported. Please list nested paths explicitly.`);
        return [];
    }
    if (!trimmed.includes("*")) {
        const abs = path.resolve(rootDir, trimmed);
        return exists(abs) && fs.statSync(abs).isDirectory() ? [abs] : [];
    }
    let currentDirs = [rootDir];
    for (const segment of trimmed.split("/")) {
        const nextDirs = [];
        for (const dir of currentDirs) {
            if (segment === "*") {
                let entries;
                try {
                    entries = fs.readdirSync(dir, { withFileTypes: true });
                }
                catch {
                    continue;
                }
                for (const entry of entries) {
                    if (entry.isDirectory() &&
                        !entry.name.startsWith(".") &&
                        entry.name !== "node_modules") {
                        nextDirs.push(path.join(dir, entry.name));
                    }
                }
            }
            else {
                const abs = path.join(dir, segment);
                if (exists(abs)) {
                    nextDirs.push(abs);
                }
            }
        }
        currentDirs = nextDirs;
    }
    return currentDirs.filter((d) => {
        try {
            return fs.statSync(d).isDirectory();
        }
        catch {
            return false;
        }
    });
}
function readPnpmWorkspacePackages(rootDir) {
    const content = readFile(path.join(rootDir, "pnpm-workspace.yaml"));
    if (!content) {
        return [];
    }
    const match = content.match(/^packages:\s*\n((?:\s*-\s+.*\n?)+)/m);
    if (!match) {
        return [];
    }
    const patterns = [];
    for (const line of match[1].split("\n")) {
        const m = line.match(/^\s*-\s*['"]?([^'"#\s]+)['"]?/);
        if (m) {
            patterns.push(m[1]);
        }
    }
    return patterns.flatMap((p) => expandGlob(rootDir, p));
}
function readPackageJsonWorkspaces(rootDir) {
    const content = readFile(path.join(rootDir, "package.json"));
    if (!content) {
        return [];
    }
    try {
        const pkg = JSON.parse(content);
        let patterns = [];
        if (Array.isArray(pkg.workspaces)) {
            patterns = pkg.workspaces;
        }
        else if (pkg.workspaces && Array.isArray(pkg.workspaces.packages)) {
            patterns = pkg.workspaces.packages;
        }
        return patterns.flatMap((p) => expandGlob(rootDir, p));
    }
    catch {
        return [];
    }
}
function readUvWorkspaceMembers(rootDir) {
    const content = readFile(path.join(rootDir, "pyproject.toml"));
    if (!content) {
        return [];
    }
    const section = content.match(/\[tool\.uv\.workspace\]\s*\n([\s\S]*?)(?=\n\[|\n$)/);
    if (!section) {
        return [];
    }
    const members = section[1].match(/members\s*=\s*\[([\s\S]*?)\]/);
    if (!members) {
        return [];
    }
    const patterns = [];
    for (const m of members[1].matchAll(/["']([^"']+)["']/g)) {
        patterns.push(m[1]);
    }
    return patterns.flatMap((p) => expandGlob(rootDir, p));
}
function readGoWorkModules(rootDir) {
    const content = readFile(path.join(rootDir, "go.work"));
    if (!content) {
        return [];
    }
    const dirs = [];
    const blockMatch = content.match(/use\s*\(([\s\S]*?)\)/);
    if (blockMatch) {
        for (const line of blockMatch[1].split("\n")) {
            const trimmed = line.trim().replace(/\s*\/\/.*$/, "");
            if (!trimmed) {
                continue;
            }
            const abs = path.resolve(rootDir, trimmed);
            if (exists(abs)) {
                dirs.push(abs);
            }
        }
    }
    for (const m of content.matchAll(/^use\s+([^\s(][^\s]*)/gm)) {
        const abs = path.resolve(rootDir, m[1]);
        if (exists(abs)) {
            dirs.push(abs);
        }
    }
    return dirs;
}
/**
 * Detect workspace member directories for each supported ecosystem. Returns
 * absolute paths; the caller runs SDK detection on each. An empty result
 * means the project isn't a monorepo — the caller should scan rootDir alone.
 */
export function detectWorkspaces(rootDir) {
    const unique = new Set();
    for (const d of readPnpmWorkspacePackages(rootDir)) {
        unique.add(d);
    }
    for (const d of readPackageJsonWorkspaces(rootDir)) {
        unique.add(d);
    }
    for (const d of readUvWorkspaceMembers(rootDir)) {
        unique.add(d);
    }
    for (const d of readGoWorkModules(rootDir)) {
        unique.add(d);
    }
    return [...unique];
}
